<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>patRoon handbook • patRoon</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="patRoon handbook">
<meta property="og:description" content="patRoon">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">patRoon</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Reference
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../reference/index.html">Online</a>
    </li>
    <li>
      <a href="../reference/patRoon.pdf">PDF</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial.html">Online HTML page</a>
    </li>
    <li>
      <a href="../articles/tutorial.pdf">PDF</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Handbook
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../handbook_bd/index.html">Book format</a>
    </li>
    <li>
      <a href="../articles/handbook.html">Single page</a>
    </li>
    <li>
      <a href="../handbook_bd/handbook.pdf">PDF</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rickhelmus/patRoon">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="handbook_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="handbook_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="handbook_files/viz-1.8.2/viz.js"></script><link href="handbook_files/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="handbook_files/grViz-binding-1.0.6.1/grViz.js"></script><script src="handbook_files/kePrint-0.0.1/kePrint.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>patRoon handbook</h1>
                        <h4 class="author">Rick Helmus</h4>
            
            <h4 class="date">2020-10-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rickhelmus/patRoon/blob/master/vignettes/handbook.Rmd"><code>vignettes/handbook.Rmd</code></a></small>
      <div class="hidden name"><code>handbook.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre, code {
    white-space: pre !important;
    overflow-x: auto !important;
    max-height: 275px;
    overflow-y: auto;
}

blockquote {
    background-color: aliceblue;
    padding: 0.25em 0.75em;
}

p.caption {
  font-size: 0.8em;
}
</style>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Nowadays there are various software tools available to process data from non-target analysis (NTA) experiments. Individual tools such as <a href="http://proteowizard.sourceforge.net/">ProteoWizard</a>, <a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a>, <a href="http://openms.de/">OpenMS</a>, <a href="http://ipb-halle.github.io/MetFrag/">MetFrag</a> and mass spectrometry vendor tools are often combined to perform a complete data processing workflow. During this workflow, raw data files may undergo pre-treatment (e.g. conversion), chromatographic and mass spectral data are combined to extract so called <em>features</em> (or ‘peaks’) and finally annotation is performed to elucidate chemical identities. The aim of <code>patRoon</code> is to harmonize the many available tools in order to provide a consistent user interface without the need to know all the details of each individual software tool and remove the need for tedious conversion of data when multiple tools are used. The name is derived from a Dutch word that means <em>pattern</em> and may also be an acronym for <em>hyPhenated mAss specTROmetry nOn-target aNalysis</em>. The workflow of non-target analysis is typically highly dependent on several factors such as the analytical instrumentation used and requirements of the study. For this reason, <code>patRoon</code> does not enforce a certain workflow. Instead, most workflow steps are optional, are highly configurable and algorithms can easily be mixed or even combined. Furthermore, <code>patRoon</code> supplies a straightforward interface to easily inspect, select, visualize and report all data that is generated during the workflow.</p>
<p>The documentation of <code>patRoon</code> consists of three parts:</p>
<ol style="list-style-type: decimal">
<li>A tutorial (accessible at <a href="https://rickhelmus.github.io/patRoon/articles/tutorial.html">here</a>)</li>
<li>This handbook</li>
<li>The reference manual (accessible in <code>R</code> with <code><a href="../reference/patRoon-package.html">?`patRoon-package`</a></code> or online <a href="https://rickhelmus.github.io/patRoon/reference/index.html">here</a>)</li>
</ol>
<p>New users are highly recommended to start with the tutorial: this document provides an interactive introduction in performing a basic NTA processing workflow with <code>patRoon</code>. The handbook provides a more thorough overview of all concepts, functionalities and provides instructions and many examples on working with <code>patRoon</code>. Finally, the reference manual provides all the gritty details for all functionalities, and is meant if you want to know more details or need a quick reminder how a function should be used.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p><code>patRoon</code> depends on various other software tools to perform the non-target analysis workflow steps and to implement various other functionality. Most of these dependencies are automatically installed when you install the <code>patRoon</code> R package, however, some may need to be manually installed and/or configured.</p>
<blockquote>
<p><strong><em>NOTE</em></strong> It is highly recommended to perform installation steps in a ‘clean’ <code>R</code> session to avoid errors when installing or upgrading packages. As such it is recommended to close all open (R Studio) sessions and open a plain R console to perform the installation.</p>
</blockquote>
<div id="automatic-installation-windows-only" class="section level2">
<h2 class="hasAnchor">
<a href="#automatic-installation-windows-only" class="anchor"></a>Automatic installation (Windows only)</h2>
<p>An installation script is provided that automatically installs and configures all dependencies and finally installs <code>patRoon</code> itself. At this moment, this script only works with Microsoft Windows. You don’t have to install anything else to use it, simply open <code>R</code> and execute these commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/rickhelmus/patRoon/master/install_patRoon.R"</span><span class="op">)</span>
<span class="fu">installPatRoon</span><span class="op">(</span><span class="op">)</span></pre></div>
<p>A simple text based wizard will start and asks you what to install and how to do it. You can re-run this installer at any time, for instance, if something went wrong or you want to install additional dependencies.</p>
</div>
<div id="docker-image-experimental" class="section level2">
<h2 class="hasAnchor">
<a href="#docker-image-experimental" class="anchor"></a>Docker image (experimental)</h2>
<p>Docker images are provided to easily install a reproducible environment with <code>R</code>, <code>patRoon</code> and nearly all of its dependencies. This section assumes you have a basic understanding of <a href="https://www.docker.com/">Docker</a> and have it installed. If not, please refer to the many guides available on the Internet. The Docker images of <code>patRoon</code> were originally only used for automated testing, however, since these contain a complete working environment of <code>patRoon</code> they are also suitable for using the software. They come with all external dependencies (except ProteoWizard), R dependencies and MetFrag libraries. Furthermore, as of recently the Docker image now also contains <a href="https://rstudio.com/">RStudio</a> server, which makes using <code>patRoon</code> even easier. This feature is still experimental and may change over time. If you find bugs or suggestion please <a href="https://github.com/rickhelmus/patRoon/issues">file a bug report</a>!</p>
<p>Below are some example shell commands on how to run the image.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># run an interactive R console session</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="ex">docker</span> run --rm -it patroonorg/patroon</span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co"># run a linux shell, from which R can be launched</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="ex">docker</span> run --rm -it patroonorg/patroon bash</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co"># run rstudio server, acessible from localhost:8787</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co"># login with rstudio/yourpasswordhere</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="ex">docker</span> run --rm -p 8787:8787 -u 0 -e PASSWORD=yourpasswordhere -v ~/myvolume:/home/rstudio patroonorg/patroon /init</span></code></pre></div>
<p>Note that the first two commands run as the default user <code>patRoon</code>, while the last as <code>root</code>. The last command launches <a href="https://rstudio.com/">RStudio</a> server. You can access it by browsing to <code>localhost:8787</code> and logging with user <code>rstudio</code> and the password defined by the <code>PASSWORD</code> variable from the command (<code>yourpasswordhere</code> in the above example). This command also links a local volume to the <code>rstudio</code> user in order to obtain persistence of files in the container’s home directory. The Docker image is based on the excellent <a href="https://hub.docker.com/r/rocker/rstudio">RStudio image</a> from the <a href="https://www.rocker-project.org/">rocker project</a>. For more information on RStudio related options see the documentation for the <a href="https://hub.docker.com/r/rocker/rstudio">RStudio image</a>.</p>
</div>
<div id="manual-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#manual-installation" class="anchor"></a>Manual installation</h2>
<p>The manual installation is for users who don’t use Windows or Docker, prefer to do a manual installation or simply want to know what happens behind the scenes. The manual installation consists of three phases:</p>
<ol style="list-style-type: decimal">
<li>Installing some prerequisite <code>R</code> packages</li>
<li>Install and configure (non-<code>R</code>) dependencies</li>
<li>Install <code>patRoon</code>
</li>
</ol>
<div id="r-prerequisites" class="section level3">
<h3 class="hasAnchor">
<a href="#r-prerequisites" class="anchor"></a>R prerequisites</h3>
<p>When installing <code>patRoon</code> Windows users have the option to install from a customized (<a href="https://github.com/andrie/miniCRAN">miniCRAN</a>) repository (<code>patRoonDeps</code>). This repository provides a central repository for <code>patRoon</code> and all its R packages. An advantage is that installation will be faster and you will not need to install <a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a>. Note that you will need to have the latest <code>R</code> version installed in order to use this repository.</p>
<p>When you decide to use the <code>patRoonDeps</code> repository you can simply <em>skip</em> this step. <strong>Otherwise</strong> (i.e. you will use regular repositories instead), execute the following:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, <span class="st">"remotes"</span><span class="op">)</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"CAMERA"</span><span class="op">)</span>

<span class="co"># only needed for Bruker DataAnalysis integration</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"RDCOMClient"</span>, repos <span class="op">=</span> <span class="st">"http://www.omegahat.net/R"</span><span class="op">)</span>

<span class="co"># only when using the R interface (not recommended by default)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"c-ruttkies/MetFragR/metfRag"</span><span class="op">)</span></pre></div>
<p>Note that the latter two commands concern installation of <em>optional</em> packages. If you are unsure then you probably don’t need them.</p>
</div>
<div id="other-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#other-dependencies" class="anchor"></a>Other dependencies</h3>
<p>Depending on which functionality is used, the following external dependencies may need to be installed:</p>
<table class="table">
<colgroup>
<col width="39%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>Software</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://www.oracle.com/technetwork/java/javase/downloads/index.html">Java JDK</a></td>
<td>
<strong>Mandatory</strong> for e.g. plotting structures and using MetFrag.</td>
</tr>
<tr class="even">
<td><a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a></td>
<td>Necessary on Window and when <code>patRoon</code> is <em>not</em> installed from <code>patRoonDeps</code>.</td>
</tr>
<tr class="odd">
<td><a href="http://proteowizard.sourceforge.net/">ProteoWizard</a></td>
<td>Needed for automatic data-pretreatment (e.g. data file conversion and centroiding, Bruker users may use DataAnalysis integration instead).</td>
</tr>
<tr class="even">
<td><a href="http://openms.de/">OpenMS</a></td>
<td>Recommended. Used for e.g. finding and grouping features.</td>
</tr>
<tr class="odd">
<td><a href="http://ipb-halle.github.io/MetFrag/projects/metfragcl/">MetFrag CL</a></td>
<td>Recommended. Used for annotation with MetFrag.</td>
</tr>
<tr class="even">
<td><a href="ftp://newftp.epa.gov/COMPTOX/Sustainable_Chemistry_Data/Chemistry_Dashboard/MetFrag_metadata_files">MetFrag CompTox DB</a></td>
<td>Database files necessary for usage of the <a href="https://comptox.epa.gov/dashboard">CompTox</a> database with MetFrag. Note that a recent version of MetFrag (&gt;=2.4.5) is required. Note that the lists with additions for <a href="https://zenodo.org/record/3364464#.XnjM-XLvKUk">smoking metadata</a> and <a href="https://zenodo.org/record/3472781#.XnjMAHLvKUk">wastewater metadata</a> are also supported.</td>
</tr>
<tr class="odd">
<td><a href="https://zenodo.org/record/3548654">MetFrag PubChemLite DB</a></td>
<td>Database files needed to use PubChemLite with MetFrag (currently tested with tier0 and tier1 November 2019 versions).</td>
</tr>
<tr class="even">
<td><a href="https://bio.informatik.uni-jena.de/software/sirius/">SIRIUS</a></td>
<td>For formula and/or compound annotation.</td>
</tr>
<tr class="odd">
<td><a href="http://openbabel.org/wiki/Main_Page">OpenBabel</a></td>
<td>Used in some cases for suspect screening (e.g. to calculate molecular masses for suspects with only InChI information). Otherwise optional.</td>
</tr>
<tr class="even">
<td><a href="https://pngquant.org/">pngquant</a></td>
<td>Used to reduce size of HTML reports, definitely optional.</td>
</tr>
</tbody>
</table>
<p>After installation you may need to configure the file path to ProteoWizard, OpenMS, SIRIUS, MetFrag, the MetFrag CompTox DB and/or pngquant (normally ProteoWizard and OpenMS should be automatically found). To configure their file locations you should set some global package options with the <code><a href="https://rdrr.io/r/base/options.html">options()</a></code> R function, for instance:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.pwiz <span class="op">=</span> <span class="st">"C:/ProteoWizard"</span><span class="op">)</span> <span class="co"># location of ProteoWizard installation folder</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.SIRIUS <span class="op">=</span> <span class="st">"C:/sirius-win64-3.5.1"</span><span class="op">)</span> <span class="co"># location where SIRIUS was extracted</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.OpenMS <span class="op">=</span> <span class="st">"/usr/local/bin"</span><span class="op">)</span> <span class="co"># directory with the OpenMS binaries</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.pngquant <span class="op">=</span> <span class="st">"~/pngquant"</span><span class="op">)</span> <span class="co"># directory containing pngquant binary</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.MetFragCL <span class="op">=</span> <span class="st">"~/MetFrag2.4.5-CL.jar"</span><span class="op">)</span> <span class="co"># full location to the jar file</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.MetFragCompTox <span class="op">=</span> <span class="st">"C:/CompTox_17March2019_SelectMetaData.csv"</span><span class="op">)</span> <span class="co"># full location to desired CompTox CSV file</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.MetFragPubChemLite <span class="op">=</span> <span class="st">"~/PubChemLite_14Jan2020_tier0.csv"</span><span class="op">)</span> <span class="co"># full location to desired PubChemLite CSV file</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.obabel <span class="op">=</span> <span class="st">"C:/Program Files/OpenBabel-3.0.0"</span><span class="op">)</span> <span class="co"># directory with OpenBabel binaries</span></pre></div>
<p>These commands have to be executed everytime you start a new R session (e.g. as part of your script). However, it is probably easier to add them to your <code>~/.Rprofile</code> file so that they are executed automatically when you start R. If you don’t have this file yet you can simply create it yourself (for more information see e.g. <a href="https://stackoverflow.com/a/46819910">this SO answer</a>).</p>
</div>
<div id="patroon-installation" class="section level3">
<h3 class="hasAnchor">
<a href="#patroon-installation" class="anchor"></a>patRoon installation</h3>
<p>Finally, it is time to install <code>patRoon</code> itself. As mentioned before, Windows users (who have the latest <code>R</code> version) can install <code>patRoon</code> and all its package dependencies from the <code>patRoonDeps</code> repository:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># install from patRoonDeps (only Windows with latest R version)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"patRoon"</span>, repos <span class="op">=</span> <span class="st">"https://rickhelmus.github.io/patRoonDeps/"</span>, type <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span>

<span class="co"># optional, data for tutorial</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"patRoonData"</span>, repos <span class="op">=</span> <span class="st">"https://rickhelmus.github.io/patRoonDeps/"</span>, type <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span></pre></div>
<p><strong>Otherwise</strong>, installation occurs directly from GitHub:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"rickhelmus/patRoon"</span><span class="op">)</span>

<span class="co"># optional, data for tutorial</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"rickhelmus/patRoonData"</span><span class="op">)</span> <span class="co"># example data used by tutorial</span></pre></div>
<p>Afterwards, you can run the <code><a href="../reference/verifyDependencies.html">verifyDependencies()</a></code> function to see if <code>patRoon</code> can find all its dependencies (you may need to restart R beforehand)</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu">patRoon</span><span class="fu">::</span><span class="fu"><a href="../reference/verifyDependencies.html">verifyDependencies</a></span><span class="op">(</span><span class="op">)</span></pre></div>
</div>
</div>
</div>
<div id="workflow-concepts" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-concepts" class="anchor"></a>Workflow concepts</h1>
<!-- UNDONE: -->
<!-- - usage of algorithms: algorithm arg vs specific fun -->
<!-- - S4 classes?: common output, hierarchy, generics, ... -->
<!-- - show images of feature definition etc? -->
<p>In a non-target workflow both chromatographic and mass spectral data is automatically processed in order to provide a comprehensive chemical characterization of your samples. While the exact workflow is typically dependent on the type of study, it generally involves of the following steps:</p>
<!-- UNDONE: include other data processing steps here? -->
<div id="htmlwidget-f4382f3a145661702be2" style="width:750px;height:250px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-f4382f3a145661702be2">{"x":{"diagram":"\ndigraph workflow {\n  graph [ rankdir = LR, compound = true, style = invis ]\n  node [ shape = box,\n         fixedsize = true,\n         width = 2.8,\n         height = 1,\n         fontsize = 20,\n         fillcolor = darkseagreen1,\n         style = filled ]\n\n  subgraph cluster1 {\n    \"Data pre-treatment\" -> \"Find features\"\n    \"Find features\" -> \"Group features\"\n  }\n  \n  subgraph cluster2 {\n    \"Suspect screening\" -> \"Group features\" [minlen=3, style=dashed]\n    \"Find features\" -> \"Suspect screening\" [style=invis]\n    \"Group features\" -> \"Suspect screening\" [style=dashed]\n  }\n  \n  subgraph cluster3 {\n    graph [style = dashed ]\n    color=blue\n    \"MS peak lists\" \"Formula annotation\"  \"Compound Annotation\"\n  }\n\n  subgraph cluster4 {\n    Componentization\n  }\n  \n  \"Group features\" -> \"Formula annotation\" [ltail=cluster1, lhead=cluster3, style=dashed]\n  \"Group features\" -> \"Componentization\" [ltail=cluster1, lhead=cluster4, style=dashed]\n  \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p>Note that <code>patRoon</code> supports flexible composition of workflows. In the scheme above you can recognize optional steps by a <em>dashed line</em>. The inclusion of each step is only necessary if a further steps depends on its data. For instance, annotation and componentization do not depend on each other and can therefore be executed in any order or simply be omitted. A brief descripton of all steps is given below.</p>
<p>During <strong>data pre-treatment</strong> raw MS data is prepared for further analysis. A common need for this step is to convert the data to an open format so that other tools are able to process it. Other pre-treatment steps may involve re-calibration of <em>m/z</em> data or performing advanced filtering operations.</p>
<p>The next step is to extract <strong>features</strong> from the data. While different terminologies are used, a feature in <code>patRoon</code> refers to a single chromatographic peak in an extracted ion chromatogram for a single <em>m/z</em> value (within a defined tolerance). Hence, a feature contains both chromatographic data (e.g. retention time and peak height) and mass spectral data (e.g. the accurate <em>m/z</em>). Note that with mass spectrometry multiple <em>m/z</em> values may be detected for a single compound as a result of adduct formation, natural isotopes and/or in-source fragments. Some algorithms may try to combine these different masses in a single feature. However, in <code>patRoon</code> we generally assume this is not the case (and may optionally be done afterwards during the componentization step described below). Features are sometimes simply referred to as ‘peaks’.</p>
<p>Features are found per analysis. Hence, in order to compare a feature across analyses, the next step is to group them. This step is essential as it finds equal features even if their retention time or <em>m/z</em> values slightly differ due to analytical variability. The resulting <strong>feature groups</strong> are crucial input for subsequent workflow steps. Prior to grouping, <em>retention time alignment</em> between analyses may be performed to improve grouping of features, especially when processing multiple analysis batches at once. Outside <code>patRoon</code> feature groups may also be defined as <em>profiles</em>, <em>aligned</em> or <em>grouped features</em> or <em>buckets</em>.</p>
<p>Depending on the study type, <strong>suspect screening</strong> is then performed to limit the features that should be considered for further processing. As its name suggests, with suspect screening only those features which are suspected to be present are considered for further processing. These suspects are retrieved from a suspect list which contains the <em>m/z</em> and (optionally) retention times for each suspect. Typical suspect lists may be composed from databases with known pollutants or from predicted transformation products. Note that for a ‘full’ non-target analysis no suspect screening is performed, hence, this step is simply omitted and all features are to be considered.</p>
<p>After features have been collected the next step typically involves <strong>annotation</strong>. During this step MS and MS/MS data are collected in so called <strong>MS peak lists</strong>, which are then used as input for formula and compound annotation. Formula annotation involves automatic calculation of possible formulae for each feature based on its <em>m/z</em>, isotopic pattern and MS/MS fragments, whereas compound annotation (or identification) involves the assignment of actual chemical structures to each feature. Note that during formula and compound annotation typically multiple candidates are assigned to a single feature. To assist interpretation of this data each candidate is therefore ranked on characteristics such as isotopic fit, number of explained MS/MS fragments and metadata from an online database such as number of scientific references or presence in common suspect lists.</p>
<p>Besides annotation, another step to perform after extraction of features is <strong>componentization</strong>. A <strong>component</strong> is defined as a collection of multiple feature groups that are somehow related to each other. Typical exmples are features that belong to the same chemical compound (i.e. with different <em>m/z</em> values but equal retention time), such as adducts, isotopes and in-source fragments. Other examples are homologues series and features that display a similar intensity trend across samples.</p>
<p>To summarize:</p>
<ul>
<li>
<strong>Data-pretreatment</strong> involves preparing raw MS data for further processing (e.g. conversion to an open format)</li>
<li>
<strong>Features</strong> describe chromatographic and <em>m/z</em> information (or ‘peaks’) in all analyses.</li>
<li>A <strong>feature group</strong> consists of equal features across analyses.</li>
<li>With <strong>suspect screening</strong> only features that are considered to be on a suspect list are considered further in the workflow.</li>
<li>
<strong>MS peak lists</strong> Summarizes all MS and MS/MS data that will be used for subsequent annotation.</li>
<li>During <strong>formula</strong> and <strong>compound annotation</strong> candidate formulae/structures will be assigned and ranked for each feature.</li>
<li>
<strong>Componentization</strong> involves consolidating different feature groups that have a relationship to each other in to a single component.</li>
</ul>
<p>The next chapters will discuss how to generate this data and process it.</p>
</div>
<div id="generating-workflow-data" class="section level1">
<h1 class="hasAnchor">
<a href="#generating-workflow-data" class="anchor"></a>Generating workflow data</h1>
<p>Each step in the non-target workflow is performed by a function that performs the heavy lifting of a workflow step behind the scenes and finally return the results. An important goal of <code>patRoon</code> is to support multiple algorithms for each workflow step, hence, when such a function is called you have to specify which algorithm you want to use. The available algorithms and their characteristics will be discussed in the next sections. An overview of all functions involved in generating workflow data is shown in the table below.</p>
<table class="table">
<colgroup>
<col width="25%">
<col width="59%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th>Workflow step</th>
<th>Function</th>
<th>Output S4 class</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Data pre-treatment</td>
<td>
<code><a href="../reference/convertMSFiles.html">convertMSFiles()</a></code>, <code><a href="../reference/bruker-utils.html">recalibrarateDAFiles()</a></code>
</td>
<td>-</td>
</tr>
<tr class="even">
<td>Finding features</td>
<td><code><a href="../reference/feature-finding.html">findFeatures()</a></code></td>
<td><code>features</code></td>
</tr>
<tr class="odd">
<td>Grouping features</td>
<td><code><a href="../reference/feature-grouping.html">groupFeatures()</a></code></td>
<td><code>featureGroups</code></td>
</tr>
<tr class="even">
<td>Suspect screening</td>
<td>
<code><a href="../reference/suspect-screening.html">screenSuspects()</a></code> + <code><a href="../reference/suspect-screening.html">groupFeaturesScreening()</a></code>
</td>
<td><code>featureGroups</code></td>
</tr>
<tr class="odd">
<td>MS peak lists</td>
<td><code><a href="../reference/MSPeakLists-generation.html">generateMSPeakLists()</a></code></td>
<td><code>MSPeakLists</code></td>
</tr>
<tr class="even">
<td>Formula annotation</td>
<td><code><a href="../reference/formula-generation.html">generateFormulas()</a></code></td>
<td><code>formulas</code></td>
</tr>
<tr class="odd">
<td>Compound annotation</td>
<td><code><a href="../reference/compound-generation.html">generateCompounds()</a></code></td>
<td><code>compounds</code></td>
</tr>
<tr class="even">
<td>Componentization</td>
<td><code><a href="../reference/component-generation.html">generateComponents()</a></code></td>
<td><code>components</code></td>
</tr>
</tbody>
</table>
<p>All of these functions store their output in objects derived from so called S4 classes. Knowing the details about the S4 class system of <code>R</code> is generally not important when using <code>patRoon</code> (and well written resources are available if you want to know more). In brief, usage of this class system allows a general data format that is used irrespective of the algorithm that was used to generate the data. For instance, when features have been found by <a href="http://openms.de/">OpenMS</a> or <a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a> they both return the same data format.</p>
<p>Another advantage of the S4 class system is the usage of so called <em>generic functions</em>. To put simply: a generic function performs a certain task for different types of data objects. A good example is the <code><a href="../reference/generics.html">plotSpec()</a></code> function which plots an (annotated) spectrum from data of MS peak lists or from formula or compound annotation:</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co"># mslists, formulas, compounds contain results for MS peak lists and</span>
<span class="co"># formula/compound annotations, respectively.</span>

<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">mslists</span>, <span class="va">...</span><span class="op">)</span> <span class="co"># plot raw MS spectrum</span>
<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">formulas</span>, <span class="va">...</span><span class="op">)</span> <span class="co"># plot annotated spectrum from formula annotation data</span>
<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">compounds</span>, <span class="va">...</span><span class="op">)</span> <span class="co"># likewise but for compound annotation.</span></pre></div>
<p>The next sections will further detail on how to actually perform the non-target workflow steps to generate data.</p>
<div id="preparations" class="section level2">
<h2 class="hasAnchor">
<a href="#preparations" class="anchor"></a>Preparations</h2>
<div id="data-pre-treatment" class="section level3">
<h3 class="hasAnchor">
<a href="#data-pre-treatment" class="anchor"></a>Data pre-treatment</h3>
<p>Prior to performing the actual non-target data processing workflow some preparations often need to be made. Often data has to be pre-treated, for instance, by converting it to an open format that is usable for subsequent workflow steps or to perform mass re-calibration. Some common functions are listed below.</p>
<table class="table">
<colgroup>
<col width="34%">
<col width="22%">
<col width="36%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th>Task</th>
<th>Function</th>
<th>Algorithms</th>
<th>Supported file formats</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Conversion</td>
<td><code><a href="../reference/convertMSFiles.html">convertMSFiles()</a></code></td>
<td>
<a href="http://openms.de/">OpenMS</a>, <a href="http://proteowizard.sourceforge.net/">ProteoWizard</a>, DataAnalysis</td>
<td>All common (algorithm dependent)</td>
</tr>
<tr class="even">
<td>Advanced (e.g. spectral filtering)</td>
<td><code><a href="../reference/convertMSFiles.html">convertMSFiles()</a></code></td>
<td><a href="http://proteowizard.sourceforge.net/">ProteoWizard</a></td>
<td>All common</td>
</tr>
<tr class="odd">
<td>Mass re-calibration</td>
<td><code><a href="../reference/bruker-utils.html">recalibrarateDAFiles()</a></code></td>
<td>DataAnalysis</td>
<td>Bruker</td>
</tr>
</tbody>
</table>
<p>The <code><a href="../reference/convertMSFiles.html">convertMSFiles()</a></code> function supports conversion between many different file formats typically used in non-target analysis. Furthermore, other pre-treatment steps are available (e.g. centroiding, filtering) when the <a href="http://proteowizard.sourceforge.net/">ProteoWizard</a> algorithm is used. For an overview of these functionalities see the <a href="http://proteowizard.sourceforge.net/tools/msconvert.html">MsConvert documentation</a>. Some examples:</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co"># Converts a single mzXML file to mzML format</span>
<span class="fu"><a href="../reference/convertMSFiles.html">convertMSFiles</a></span><span class="op">(</span><span class="st">"standard-1.mzXML"</span>, to <span class="op">=</span> <span class="st">"mzML"</span>, algorithm <span class="op">=</span> <span class="st">"openms"</span><span class="op">)</span>

<span class="co"># Converts all Thermo files with ProteoWizard (the default) in the analyses/raw</span>
<span class="co"># directory and stores the mzML files in analyses/raw. Afterwards, only MS1</span>
<span class="co"># spectra are retained.</span>
<span class="fu"><a href="../reference/convertMSFiles.html">convertMSFiles</a></span><span class="op">(</span><span class="st">"analyses/raw"</span>, <span class="st">"analyses/mzml"</span>, from <span class="op">=</span> <span class="st">"thermo"</span>,
               centroid <span class="op">=</span> <span class="st">"vendor"</span>, filters <span class="op">=</span> <span class="st">"msLevel 1"</span><span class="op">)</span></pre></div>
<blockquote>
<p><strong><em>NOTE</em></strong> Most algorithms further down the workflow require the <em>mzML</em> or <em>mzXML</em> file format and additionally require that mass peaks have been centroided. When using the ProteoWizard algorithm (the default), centroiding by vendor algorithms is generally recommended (i.e. by setting <code>centroid="vendor"</code> as shown in the above example).</p>
</blockquote>
<p>When Bruker MS data is used it can be automatically re-calibrated to improve its mass accuracy. Often this is preceeded by calling the <code><a href="../reference/bruker-utils.html">setDAMethod()</a></code> function to set a DataAnalysis method to all files in order to configure automatic re-calibration. The <code><a href="../reference/bruker-utils.html">recalibrarateDAFiles()</a></code> function performs the actual re-calibration. The <code><a href="../reference/bruker-utils.html">getDACalibrationError()</a></code> function can be used at anytime to request the current calibration error of each analysis. An example of these functions is shown below.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co"># anaInfo is a data.frame with information on analyses (see next section)</span>
<span class="fu"><a href="../reference/bruker-utils.html">setDAMethod</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"path/to/DAMethod.m"</span><span class="op">)</span> <span class="co"># configure Bruker files with given method that has automatic calibration setup</span>
<span class="fu"><a href="../reference/bruker-utils.html">recalibrarateDAFiles</a></span><span class="op">(</span><span class="va">anaInfo</span><span class="op">)</span> <span class="co"># trigger re-calibration for each analysis</span>
<span class="fu"><a href="../reference/bruker-utils.html">getDACalibrationError</a></span><span class="op">(</span><span class="va">anaInfo</span><span class="op">)</span> <span class="co"># get calibration error for each analysis (NOTE: also shown when previous function is finished)</span></pre></div>
</div>
<div id="anaInfo" class="section level3">
<h3 class="hasAnchor">
<a href="#anaInfo" class="anchor"></a>Analysis information</h3>
<p>The final bits of preparation is constructing the information for the analyses that need to be processed. In <code>patRoon</code> this is referred to as the <em>analysis information</em> and often stored in a variable <code>anaInfo</code> (of course you are free to choose a different name!). The analysis information should be a <code>data.frame</code> with the following columns:</p>
<ul>
<li>
<strong>path</strong>: the directory path of the file containing the analysis data</li>
<li>
<strong>analysis</strong>: the name of the analysis. This should be the file name <em>without</em> file extension.</li>
<li>
<strong>group</strong>: to which <em>replicate group</em> the analysis belongs. All analysis which are replicates of each other get the same name.</li>
<li>
<strong>blank</strong>: which replicate group should be used for blank subtraction.</li>
<li>
<strong>conc</strong> (optional, advanced) A numeric value describing the concentration or any other value for which the intensity in this sample may correlate, for instance, dilution factor, sampling time etc. This column is only required when you want to obtain quantitative information (e.g. concentrations) using the <code><a href="../reference/generics.html">as.data.table()</a></code> method function (see <code><a href="../reference/featureGroups-class.html">?featureGroups</a></code> for more information).</li>
</ul>
<p>The <code><a href="../reference/analysis-information.html">generateAnalysisInfo()</a></code> function can be used to (semi-)automatically generate a suitable <code>data.frame</code> that contains all the required information for a set of analysis. For, instance, the following line was used in the tutorial:</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co"># Take example data from patRoonData package (triplicate solvent blank + triplicate standard)</span>
<span class="fu"><a href="../reference/analysis-information.html">generateAnalysisInfo</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="fu">patRoonData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/patRoonData/man/patRoonData.html">exampleDataPath</a></span><span class="op">(</span><span class="op">)</span>,
                     groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"standard"</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,
                     blanks <span class="op">=</span> <span class="st">"solvent"</span><span class="op">)</span></pre></div>
<pre><code>#&gt;                                                path   analysis    group   blank
#&gt; 1 /usr/local/lib/R/site-library/patRoonData/extdata  solvent-1  solvent solvent
#&gt; 2 /usr/local/lib/R/site-library/patRoonData/extdata  solvent-2  solvent solvent
#&gt; 3 /usr/local/lib/R/site-library/patRoonData/extdata  solvent-3  solvent solvent
#&gt; 4 /usr/local/lib/R/site-library/patRoonData/extdata standard-1 standard solvent
#&gt; 5 /usr/local/lib/R/site-library/patRoonData/extdata standard-2 standard solvent
#&gt; 6 /usr/local/lib/R/site-library/patRoonData/extdata standard-3 standard solvent</code></pre>
<p>Alternatively, the <code><a href="../reference/GUI-utils.html">newProject()</a></code> function discussed in the next section can be used to interactively construct this information.</p>
</div>
<div id="automatic-project-generation-with-newproject" class="section level3">
<h3 class="hasAnchor">
<a href="#automatic-project-generation-with-newproject" class="anchor"></a>Automatic project generation with newProject()</h3>
<p>The previous sections already highlighted some steps that have to be performed prior to starting a new non-target analysis workflow: data pre-treatment and gathering information on the analysis. Most of the times you will put this and other <code>R</code> code a script file so you can re-call what you have done before (i.e. reproducible research).</p>
<p>The <code><a href="../reference/GUI-utils.html">newProject()</a></code> function can be used to setup a new project. When you run this function it will launch a small tool (see screenshot below) where you can select your analyses and configure the various workflow steps which you want to execute (e.g. data pre-treatment, finding features, annotation etc). After setting everything up the function will generate a template script which can easily be edited afterwards. In addition, you have the option to create a new RStudio project, which is advantegeous as it neatly seperates your data processing work from the rest.</p>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAzQAAAKTCAIAAABAftrCAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgAADsIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMS41ZEdYUgAAOPhJREFUeF7t3e2vHNd9J3j9RcFitMCO3+xOAsN+MwmCAHkRIMEC+2LWgGkEi8xEEoPVIDIsA/KFMoplz9gA1/A48soKAwvXNkIqNPRAaElbFCmTNEXrWaSkiUXrYRSZQ0qGtaeqTlWdeurbfW/fvqfZnw9+tqpOnXPqdN1b3V/2fbrtEwDgVvffF3fphZeu/4+bwwrtcdJFPP3/nf7Vr9798MPrN258dPPmx2mFltAejoY+sfdS/fcz/0taHz7/f/76pb/qNcauA71uC42ddurLt335VGcj2Q7/Df///vvvP/744w899NARAAAOyH996KEnnnjitnfeeeexx/7ptdeuvn7lTaWUUkopdYAVItlt//RPJ3qtSimllFLqoOq21173nplSSimlVC51W29fKaWUUkodYAlnSimllFIZlXCmlFJKKZVRCWdKKaWUUhmVcKaUUkoptbq69PwLv3jhpbQl7IbGZlc4U0oppZRaUYUQ9p/+099+9cGvN/ksbITd0NjkswXC2TMPfva22h2P9o/uVOe++pm7vt9vzKTC2uZ8RNt3fObrz/QblVJKKaXmqhDF/varX7vni1+q8lmVzMLuHXfe1cS1ecPZ9++87bY7t3uNi1TO4Wz+Es6UUkoptadK89kwmYWaL5yd+fpnhqHk0bs+8+DX77gthrbmfbXqLaj2bbbiaPHWVKV3tPd+1ffv/OxXH7yrd6jTOawkZsTtO2777FfP1KPKjaqK3Ue/Hk9YdS5Gfb1YQ/ko2rXVD+r7d8bgOFhYs/JwinDGZrs6qpRSSim1cDX5bJjMQs0VzkJk+cyD53qNIZzddlv9ZlgR1KoO/XfI6uSUtM/qnMSpKjn1O4eEVA4pzl5FqP67WcUksaWIVjHSNYkqDGwz2W3V5DGcDRbWdKjLO2dKKaWU2muFNFa9Z1a9f7bMcNY0FnmoVcWg5n2mKkK1OWysc6zkPbBzX72ziEHDzlWfZx68q3h7LCS59r204ST1ypM+YcL6XbH2fbgqnA3OVQfBtoQzpZRSSu2pmrfN7rjzrvT7z5oO831Z89G7Rr7hrBvO2sRTVIhiMSGFeDQMZ93ObY2Gs37n4rzb5dGiz/fj/G3tJZx1pxLOlFJKKbXMSpNZ2G5203w25w8EFF8fbN88e/SuIsQk4SxkoOZrhWU1saYYWCaeNpwNOrc1DGcjnUOi+sxn61D12c/U0zYVMla9sLCMcsIknBVBs56w6VmFs+G5kqmqEs6UUkoptfu69PwLf/M3D6TfZ1bls3//H/5yF79Ko4hZURV0knAWKuSYqIwvRdApfPaOO+M7W1VLtd3rnEzSD2djnZvA10laTYVJwkmrEbFbGs7SCevG5gcCBudqvjgbF1Z2aN+ZU0oppZRaqEIIa5JZVWG3SWah5g9na1PplzXnK++HKaWUUiqX2vRwVr1VFt9gU0oppZQ66LoFw5lSSiml1PqWcKaUUkoplVEJZ0oppZRSGZVwppRSSimVUQlnSimllFIZ1W2XXnhJKaWUUkplUrf9y4cfKqWUUkqpTEo4U0oppZTKqIQzpZRSSqmMSjhTSimllMqohDOllFJKqYxKOFNKKaWUyqiEM6WUUkqpjEo4U0oppZTKqIQzpZRSSqmMSjhTSimllMqohDOllFJKqYxKOFNKKaWUyqhmhbN/AgBgH/RCV1o7hLMPAABYKuEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZGR/w9mVK1cuXLhwJm9hhWGdccUAwHr62g+e/d2/+off+cLfZVtheQ/+4Nm43Gn7GM5C4nnxxRevX7/+Sd7CCsM6X3/99bhuAGDdhNDzP//7v/9Xd/zg9sP/mG2F5YVFfnV7h3y2j+HswoUL+SezSlhnWG1cNwCwbn738NHMk1lVYZH/5vDRuOgJ+xjOzpw5E7PPOgirjesGANbN73zh73oxKNsKS42LniCcRcIZAKwv4awo4QwAyIRwVpRwBgBkQjgrSjgDADIhnBUlnMGmOH3v7YnPPfJabF+2cJ57T8ftA/DaI5+7fa8LmPUQwvxLuHbFB2P/PgSwvoSzooQz2ABlYEnDwKBhaZaRjQ7YzIewpFQlnME44awo4QxueUUQGIaNMoIsPx8IZ3MRzmCccFbU0sLZ+a3iKyWlQ9vXYuNyhVPsNLVwBn2TIey106dfC8rtIivc+0jRsxCDSTmy1gsrnWPN9GUMjJpTjjYOVGnldDtt2nWR5Q2i1ayHka6tOjRrteMzzZp/crrq4U5dDNhc84azZ/8lvvB/8skHL1/qH12g3nj1k4/O/LDXOFdlH86KZNbkpmvbh/YnnwlnsAtlPOhnhr4qQ/TTQ9vQnaSXfzoHu8eq4NI5NhFJqhU0R7td51pe3OsuYKRfZzXNXnKoO0NP0S9ZyMj8zV73BMPTJdMA0VzhrEhmTaK6dOb6XvLZLRvOBmmsaNg632xGVZfy2Fb9NlvsFTTvvNVN7ci6RTiDxZWZYMcYkEST0nBU0mMQLNI0k24Ppp2xmvGusWF0eWnnpMdgMRMPo9gaHBrMMNAZNmsdwxOkqxkcBEpzhLNBGvvhOx988i/bYSOEtusffRDSwPV3tqrdSrVbVIhitbffqKYqVfms2f3k1Wfryacr73A2yGatInHFaFVvlpmr6l40JYmt7FdvtSPbXmFLOIMFDfPJmF5WGBs0FifKKBLFjJLmlfJwN+UM40w0nD4Z3T847Nw2JWeY9TAmFzJjjYXOqedcRy1pGhkJBDuHsyKKTbzXlb6j1nYrI1cdxWKqa4+275xtv111670zN1n5h7OYwGaoexX/rSNWnbaGM7TZLDkqnMEuFDFgMmrUelmhTBEjYpfk8CCHpNvlqUeMrabo222fjjLlkV60abokw8rNEUW/cm1jC+k+hIF0KTPXMXKCZOJ0GqA1Xzgr3ycbVhGq4qGtlz9q3zAr3lFr3jyrqslk6UbzhlkR43Z882xt3zkrj7XqcFbHrjptpVGsUrSk6m7CGSxsMgck8aHXZ2Y8GUaStHu6vUgCGfadXt7YxE1TsoBZDyOZvW/WsO6p51xHLWkaGQkEe33nrA5hRTjrqL/u2RoJZ6kdv48t73A2ks5CjKqDWH2gDmVj4Sxtq4zmMOEMdmU8h5RJoW7uZ4VhsmhbBtN1ZuqMHEaQyVAyPmtsGF1eZ5p2eDqus5hS2zJYSXNoOCrRGVb2nFhHullJew/ODpR2DmfpVydjhVzVfM9ZEs6qr1G2lb43lmay3sa8lXk4K2NTm5uKqFVlrSScFT0mw1l5tGqsh3Rbqs2m+zThDMaUsSANA1VDGx2GWaGMFk1Tmiuqsc3QOFXT0Jmq13eQV1rloe4s7TSdOUvD453ldc/YXU69111bemh4ulZ3+hnrGD1B5/FNnQI22RzhrHoDrAlSzbeUdb98mbzBVnwzWdHehrPyfbV+Jmu/56wYu+5f1qyU6Suqw1fSunU+RqvRcBYjWOxZNiQtdR/hDPagCgeNbjIYzQqdEW0eCfpHylCS5o7Y3tkvTOeRagWPtH3TE+64vORg2ZwOnn4Y3bUlh/oPIVVP15xyYh2ViROMPyJgvnAWKv0CZfMOWRrOOn2Sb0QrffDyG/V3lZXZLuazarswz+/mWIdwlgfhDNbU0tLKIJwBa2TecJZBCWfzEs5gTS0tnHlPCtaZcFaUcAbkYBmZqv4Co/fNYG0JZ0UJZwBAJoSzooQzACATwllRwhkAkAnhrCjhDADIhHBWlHAGAGRCOCtqx3B24cKF69frX8qWt7DOsNq4bgBg3fzu4aP/6o4f9GJQhhUW+W8OH42LnrCP4ezKlSsvvvhi/vksrDCsM6w2rhsAWDcPPHrmf/q/Hsk8n4XlhUV+5R9+Ghc9YR/DWRASz4ULF87kLaxQMgOAdfc33z/zu3/1D7/zhb/LtsLy7v/+M3G50/Y3nAEAsBDhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAysqdw9jIAAEu1p3AWf4srAABLIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkJFVhrPzW7f3Hdq+Fg/OrZpl63zcXVAYvePQ8gy7WFklOcPelvrJte1D9Sr2OBMAsDYOOJwFu05Bi5sz4pTddres5YWokMx2vQoAYH2tPpwlySVpiJtb5X9inyqfVNpRvVmq3VIyczo2BpykY7drpTm8db7cbGPR+BnSxcX2wRmqhvJg3NxuRrVTdWaqTjtoSmYqJadqF1qNOrS9XR9sDwEA6yKvcBYVLZ18UopRI52lMyqIXQZj23wTJYsoDE/WPVtronn8DFXD8PFVxh9COdPscDacqzow/SgAgLWx+nDW18koMX7U+3W2iLGjPJrklHKz26do74ytmqudZGhH1V4fSIdPnKHbnuieIdmrNuORamc4PB2drrtzJL0WzZGyXzxSjel1AwDWxAGHsyY7pLlkuBsbytjRHor5oyN06caa1GDWSn9Ae66JM9R9au2M3TMkeyMH0hWmJ6o6dRfVDh88uqpheKhtBwDWyOrD2Xhc6B0bJIs2zvRzSs/W+W5GSU0soD+gPdfEGWK/zuGqtXuGZG/kQHXC9EC63V1Uf6bk0VUdi0PdIW07ALBGMg1ncb8bNJJwkmwmOSWqetST9fsnJ6l1BsRzVdNOnKFr8gyzD5TTpu3po+wlraRbp1d9pOzXHVLt1d0AgDWRazirM0gipo60Z7WdqPoMxnaGFuoEEw1PNhwSle0j/fsLKvolS002671ypsH8dafkFKGhM3w4Jh1SP7Rqrz4jALAmsg1nQRqB2iO9nmlSSYaPj21a6wTTauaJv4qi7TF+hvQESefOGZKldldd7sVByYnjt8vFXs2BsN8d3llT/9z1fmcuAGBdrDKcLUcabAAAbjFrFc7ad6uEMwDg1rRW4Sz9Qh8AwK1orcIZAMCtTjgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGVlpOLt58+abb7558eLFMwAAiwspImSJkChitljEinPIrpe6unAWFnfp0qXXXnvt7bfffh8AYHEhRYQsERLFoqEn9H/++edDWrp+/Xps2mfhROF04aSLLnV14SysL1zNsNCwxI8BABYXUkTIEq+//nrIFTFhzCf0X3TIUuzivKsLZxcvXnzvvfd+85vf/BYAYLdClgiJIuSKmDDmE/qv7D2zVDjpoktdXTg7c+bMRx99FK5p3AcAWFzIEiFRhFwR9+ezaP8lWvTUKw1nH3/8sXAGAOxFyBIhUQhnI7WLcFZ9TTPuAwAsLmSJkCiEs5ESzgCA1RPO+k1NCWcAwOoJZ/2mpoQzAGD1hLN+U1PCGQCwesJZv6kp4QwAWL19DmfXLoeurfOvfhAP7FqYJW7NZ+3C2S+PHz58/JdxBwDYNPsazi5/69/e3vNHW2f2ls9u9XAmmwHAZtvXcHZ+K0ayjv/tj/4k8e/uOnr5Ruw/j1s8nMlmALDhDiCcDRw6di0OmMOtHc5kMwDYdKsKZ5/6THynbMR/fHzDwtnFb/5x5Zvx74T+8vjxaiscqdsAgI20gnD2qUPbr9648cHO5vrq5vqHs4vfrPNXE9Lqd8tkMwDYePsezj71pac/uPzQn3yqfP9spk/9x6fn+FmB9Q9n02QzAGCfw9mn7nr82pvbh2L8mul/P/pqHDbTLRzOZDMAYJ/D2fb2m9cev2uOd81u/6NvXY6DdnDrhjM/DAAA7HM4u3Hj2tNfmieafepLT8/76zRu2XAmmwEAwb6Gs09unNnaOZt96k/+9vz8v+nslg1nAADB/oazfSCcAQC3MuGs39SUcAYArJ5w1m9qSjgDAFZPOOs3NSWcAQCrJ5z1m5oSzgCA1RPO+k1NCWcAwOoJZ/2mpoQzAGD1hLN+U1OLhrOLFy9++OGHwhkAsBchS4REEXJF3J9P6H/9+vW4s0LhpIsudXXh7M0337x69apwBgDsRcgSb7zxRsgVcX8+of+iQ5ZiF+ddXTi7efPmpUuXwtX89a9/HS4rAMCiQooIWSIkipArYsKYT+j//PPPh5y0svfPwonC6cJJF13q6sJZEBYXVnnx4sUzAACLCykiZIlF405lxTlk10tdaTgDAGA24QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGRkleHs/Nbtt99+aPta3N034Txb5+P20iy2+Gvbh+brHjqu4IoAjfJejhZ+ptjzrV1NUAiHq6WUi0g2gY13y4Wz/XqK24/Fz/08DyxHm41q+5GHpm/t6sjIOYUzoHVQ4Sw+E203z5Ttk1Ln2TN9rqrGlJrmOM9WdejOO8v/VIZPc8nMnafN9IydUc0Zt86ni0+PjJ2n+9xcb2/XQ9LmqJ54ciXAEqR3Zn27NXdacvs1t3r3GWbrfDpBvT3PrV1KnjaqI3H2YgHJZmGHZxjg1naw4ayjegoaHJgYMNa89f3OXvcprfOMWYgzDNqnD+ywlkT9rF20T8zTaS57Tq4EWJJ4807ds4mqS/cZ5vxOt3bbJeqcqDNZcaRqKJ+qks1uv2C4WuDWdsDhLH0qqo6knRKd5urJL3lGi/MUqoZ2v5aeI32C7bTXT6vF8O7Mabdye2QtieQE9Zydnap/2mvGSoCliTd2VN9v3dsvuTNj9+ZGTG/aeJN2BlU9015d1Xz1dMlef7MeXE3VnB/YCAcbzjpPUZ0jUd0hPu91lP278wSDhsrUk+X4+NCvP6Bun7GWRDq6O1O1V52wc2RyJXEXWI7eLVzcdN27NNG/L9OO6c1c6D9JjN293fmSvXazt7ySJwLYLPmFs1L69DT1dFUe6M4TDBoqU0+WVXt/fOjXH9AucXItiXR0d6b0hJ0j6YHS4KIAy1Tdc+Vd171LE/0nlLRjf1D/SWLs7u3Ol+y1m/WyOtoVABsg03AWtb127hENGqKqvT7Qm7iZOT4ttgfigNhedSuPjKwlkT43d5+nq7101vrI5EqAJWnv/N5u534ftLf34fDWru/Z9Jbt3tqp7nzJXn9zbDCwKXILZ/EJLlV1qgYkqom68wRtv96T22DmeHx4xskD9ZGJtSTS5+Z0u96rFpycoWgYnnAwL7Ang5t35G6sVHdf/xkmvZ2Hd+zErZ3ozpfspQcGi/RMABsmw3fOOs946ZNS+ow18uQWNcOHz2fpzKNDgrS9PWP8afl2yrG1JKoZq+7pdr3XXX7bMLkSYEnSm7dzm43dfv1nmLFbu/erNEr9W7vWnS/Z652oGR/0pgBufasMZwC3jjSoASyRcAawG8IZsE+EM4DdEM6AfSKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOgE1x8+bNN9988+LFi2eAlQi3W7jpwq0Xb0LmI5wBGyG8PFy6dOm11157++233wdWItxu4aYLt558thDhDNgI4Z/v4UXi+vXr4UXiY2Alwu0WbrrXX3893IDxVmQOwhmwES5evPjee+/95je/+S2wQuGmC7deuAHjrcgcDiCcXbly5dSpU8dKYePq1avxAMC+OXPmzEcffRReKuI+sBLhpgu3XrgB4z5zWGk4Cx+hc+fO/fSnP3355Zf/W+mll146ffr0c889t9MzZvknhnt/Yfj81u23b52PO3tWzFY49N3vHopnWu4JgIMUXhs+/vhj4QxWLNx04dYTzhay0nD27LPPhmR2bSDks7Nnz8ZO48pw1otnS81OYbJe9hPO4FYSXhuqr2nGfWAlwk0Xbj3hbCGrC2evvvrqk08++Ubphz/84f333/+lL33pRz/6UdXy+OOPhw6x64ginG1thbCUJKhOdorprRC7FC1N785OMTINXcnY0N52TU8wnB9YJ8IZHAjhbBdWF85C/3Pnzr3yyiuPPvrov0uE3dB49uzZmRMW2SjkpCIuNdmozU6d6NU0xzGFKlrVQauXzUqhLc7QTtaZaeS0wPoQzuBACGe7sLpwdvTo0V/84hchh33xi1+Muaz0la98JTRevnw5dIhdRzRBq4hGMSc1KakNVpV6v01noaV4363dGYSrdo5hOGuPVfr7QP6EMzgQwtkurF04qyJTmY3q7FQc62vSVd2z/HJlu1POlKq7joSz6fmBtSGcwYEQznZhdeHs+PHjZ8+eDTls6suaoUPsOiIJZ1VoCukoZqckTQ1UkSvGsuo/4f9HstlO4SweA9aVcAYHQjjbhdWFs5///OcnTpwIOazKZ1/5yle++MUvVskseOyxx0KH2HVEEZCSTFXGs0OHqnBWh6gx4dDWdnzLrN4Z7RqOTYWzWfMDa+JWCWe/PH74j/sOH/9lPArZEc52YXXh7MaNG08++eSPf/zjKo2lQmh76qmnQofYdUQvnFWBKajaiqPt4TRLlUc63brTNGaFsxnzA2viVgpn3+z8rvUyrnWbIB/C2S6sLpwF7733Xshnjz322NmzZy+XwkbYDcksHIqdxhXxqBeIynjWSUy1tGMnV7W5a2BmOIvbtd5CgDUwdzi7+M0/Pnz8Yv3+VPOmVIhAh48f/2bZWCWh0DHqvHPVNieJKXm/q+08f2OiON5PYsUppTMyJZztwkrDWfgIvf/++z/72c+OHz9+tBQ2wm5oXP9/zgJZWyScBTHrFDtVRipDUxuX2gOd7aJX3dxsp41tkGo2dmzsEM5YM8LZLqw0nFVu3LgR0tivSmFj5lczAZZjoXDWJp3xhNXZCeoxoykpNKZ96/2pvjvErGE4K1p2GAQHRzjbhQMIZwCrt0g4G8tdnTzWz1AxMfUzW6nMTj1lp/ZAMtVoY2psuomukAPhbBeEM2AjHGg4GzQmirkKnT6jjaV4qnZn5uRw4ISzXRDOgI2wSDhLglcTrToZqx+46jH90FYabezpT1iabExnK2PcTtPDwRHOdkE4AzbCQuGsyURtsOolpeJA2ilud3rVg4vGNj/Vre3UybDRxo6itZmrVIwZ6Qh5EM52QTgDNsJi75x9swg8hSb0DKNSGYpKnfa2OWlPGrsxrTazMT33MJxVbckUkBPhbBeEM2AjLBbO5BxYEuFsF4QzYCMIZ3AghLNdEM6AjSCcwYEQznZBOAM2wtzhDFgm4WwXhDNgIwhncCCEs10QzoCNIJzBgRDOdkE4AzaCcAYHQjjbhQMIZ1euXDl16tSxUti4evVqPACwb4QzOBDC2S6sNJyFj9C5c+d++tOfvvzyy/+t9NJLL50+ffq5556b7xnz2vah2xuHtq/F5lU4vzV+wmZJWyfDVtmlaFrt4oAdXbx48cMPPxTOYMXCTRduvXADxn3msNJw9uyzz4Zkdm0g5LOzZ8/GTlPKFJRknpCWQiI6H/f23UQ4K1bVX4RwBhl68803r169KpzBioWb7o033gg3YNxnDqsLZ6+++uqTTz4ZPkKjHn/88dAhdh0xFnjatiKoNREp6Rrat7bDbnBo+xfbhw5tbxeRru5c9IyaucsQFnsFVccyCJY6QaxtLsY35+2uNenUT3HA6ty8efPSpUvh2ebXv/51eLUAViDcbuGmC7deuAHjrcgcVhfOQv9z5869MuHs2bOzJuzmnYEiAE2Es/ZAGcXaSTpTthOUQ2J7OSIOD+1jCyi6xB7NhMnM6cI6JwRWL7w8hH++X7x48QywEuF2CzedZLao1YWzo0eP/uIXv4hZbODy5cuhQ+w6lIacEZ3DE9moF476YavenzFVp39U9Ijdm86djXTV/X0AgL5NDWfFTl95cJnhrJirb8ajAABYYTg7fvz42bNnYxYbCIdCh9h1KElJY3YXzkYnXHY4E8YAgEWsLpz9/Oc/P3HiRMxiA4899ljoELuOGA1TTfiZkahG24Op4DRjqsH5g6JH7N50Hm4AAMxpdeHsxo0bTz755I9//OMYxxIhtD311FOhQ+w6qkg6adQpUlS9Xx6LGSltn0pahXRM0nNqSGhPBjeKHrF707k7qh2WtAMAjFtdOAvee++9kM8ee+yxs2fPXi6FjbAbklk4FDvNUmSdRifnFMGnsnW+zVFTSStKp6u7TQ6pOvfDVdFjRjirdmr9wQAAfSsNZ7/97W/ff//9n/3sZ8ePHz9aChthNzSGQ7ETAMAGW2k4q9y4cSOksV+VwsYOX80EANgkBxDOAACYIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGbISbN2/+5Cc/+dGPfrQ9EBpPnz4dOsSuAAdKOAM2wqlTp0ICe2dCOBQ6xK6Trm0fur1xaPtabN6bYtKt83FnXue39n7+4sTLehBzCsueeKjhyEC3a+fiJ8cWvn7FqcoBK7sAuzvRwg+MW4hwBmyEH/zgB9cGLl68+PnPf/7nP/952A4dYtdRZTZIXmHLNLGM185dvAZPZ5xFrCybpKaWHtp7aykucNPUW2v6wVj4+hUTL+MDt9928YnBLUM4AzbC3//937/R9eKLL/7pn/7ppz/96d///d+/fPly6BC7jhgLMm1bkTi2i7zQJIYiAETJ62vSmrx5U6lnb1sGJ4xCj+TIcM6q8dD2dnMkfYlv+h/a2poIZ+2UzeFiVVvnmwOdzJCsoGmfXkA1U9xplQO6ayl6xrZkM4otxX8qxeFut2JdzZnajltbdXune/souieqlQs8X8+SDpv80DedJk7UuQ695u4DY+MIZ8BG+O53v/tK4uWXX/7zP//z3ys9+OCDoSV0iF2Huq/6A+ULa/tKW+zWe+3IorX7Cl31KXoMewfpNInQ3OkyNme5GQ+UL/NxnqR/2dwObqRTttud3v0ZR1Y/tYCg2G33amFAfdZaMl0133BUIZkvHZEuLV1BOVO103YvGtNT9ZcSJOPSPmVzbA3GJ+qeqO6drDZdeLNdbLQzs1mEM2AjVOHs+eef/8IXvvD0008/8MADVTL7i7/4izKtzQxn6WvqiM7h/mvq+GtsOyQ5Hhrrl+hSf78w4yW7nXOwoGqaTnPS3uq31SN67WlzO2GyP7GAUncVUWjsraXoljQVkzTSrskauidqzjN1/rq9d3xygUljM6TTPDVReqJ04np/9ISD3mwS4QzYCCF7vfjii5///OerTFb5sz/7s8uXLy83nBU7fe3Y4hW3VrW2r8HpsVr6Wl8K0/fbhnN2FlQcLoc0G9HIVJ2BQb24+r9RPdPkYx1fQKW/jNLITIM+laZnPJ4srTtzs4bOYpIBdffe8XTGVuiULqge0xk7NVF6or5wvD7cM7oMNoRwBmyEKpz95V/+Zcxlv/d7f/AHf3DmzJkqmQWzwtnEy2dt5it0rZgjvhoHba/2NXiHs5S6fabm7KyhGdKfP/Tqn66/+GJIsV//N6pn6ndvjC8gGjlvt60YPT5vqzlDsrTuiZoezUalGTD1KJIZW6FT7zGUfTpjpyaaOlFUH+4ZXQYbQjgDNkLIXiGBvfzyy3fffXdIZp/+9KePHz9exbLKrHA2/vrZvNZ2XnTne6lthyQHJl68O0KfZvbJOScW1J1/bKH9tnpEr32iuTWxgNLooDAgbSv6tA1jI5q2YiOeqtuvWUNv+KC9d7y7+Fq3sRnSaZ6aaOpEtdETlr2HrWwG4QzYCI888khMYa+88uUvf/nIkSNxp/a9730vdh1VvFSmr6zFC2q933txLXabnmMvzOVc9ZD0Nbg80Mw0+prd7z46Z2dk0qu32XlAUbr4drvqHadMZhl/rNMLKHSO1UJj0iXonLEc0xnUTlJ0TLvFaTrjk1OWM1U73eXWp0+3E+W4tE81X7tVGp9o/ERJc7sV1FMWjcnUbBThDNgIj5ViEBsIh+Z4TiteNhvNi2n7ctooXldrndfiaOt8uVONie1xgvQcoy/NoUN76ok5OwsqmtsRzfzlb4BIHkSrXUJn7eXv3qikC0uW0PafsYBib/jAOo8qKtfRWUIr6RwPVHM2a69+70VnDbG9+A0fZXtnXSMPuqN8RMVv4ej2KZs7j2ZkovRE6cNIz9QO63cdXixufcIZsBH++Z//OTxrfe973/t/Bx555JEQzn75y1/GrpkrXrNX/YLdyTF7EmZazkQrNkxh8zuIjxjrTTgDNsLNmzfffffdtye88847a/S3NUNQWPFr/dLC2eqXviR7CWd7GctmEs4A1k54uV/p+09LCmdrG812H7CKcVNfKoUpwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI8IZAEBGhDMAgIwIZwAAGRHOAAAyIpwBAGREOAMAyIhwBgCQEeEMACAjwhkAQEaEMwCAjAhnAAAZEc4AADIinAEAZEQ4AwDIiHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkRDgDAMiIcAYAkBHhDAAgI3sKZy8DALBUewpnHwAAsFTCGQBARoQzAICMCGcAABkRzgAAMiKcAVPefeHY/YcP33/shXdjA0AO3j330D13zHDkmQ8+eOXYfXd/4+SVOGK9bEw4u/rCU6fPz1eXL771fhwFm+vdFx69Pz7P3XH/o/IZkI8QvOKz07gYzsLWeuazjQhnv3ru5B8e/sfb56+/furEVfmMTZYms4p8BmRj/nAWrGE+24hwdvo7g/i1Y8lny/f2uYfvO1zdKgNr+97zLWmYzCry2T67cixe9/uPuR0y8PzDd5cfjrsffj62zOutJ75Wjrzja0+8FZtYroXCWbBurzGbEc6+FSPXH/6Xn5zofxGzWz8++afy2f5499Q34k0yQT7Lw1Qyq8hn+6l5Lbnv2CuxKXr7hXMX3B6r9syR6sNRvtCPuHLhVGn4oZn+SLIki4azYK1eYzYrnN17JrbM0HSWz5Zrp1upIJ8dtNnJrCKf7Zupl/R3nzlyd2h81Ov8au0QzmYksBmHWI6dXlFGwlmwPu9JC2d9nXBW5LOTp70SLcNOt9JO7vnaE57l9tc8yawin+2PiZf0Cw9VX1274+4jz7wd21gB4SxjO72iHDkVO64p4ayvH87mG8WO9hrOgnuPvhAnY/leqb/faT73e9FZvtGX9FceTe6cQTx794UnHj52QVTeFwuFs7dfOFd+jfPcC+EjJJztux1fUe6+/9vHTpYfksS5V9blXhHO+oSzfbKEcDbxFMlSPH905m8N6rvn6KLfI82ORl7S33rigaqp0cbity4cvb/8ERsJYH8sFM46nYWzfbe7V5T1+XgIZ33C2T4RztZL81LTcPX33eAl/e2TYz9GE+LZu1dOHknC9N0PnSv7s1TCWcaEs6lax3C2809rPv2Tw3/dxrKqhLOlGNxKh+/7xre/ffREfLd5Ls/7qfSVEc4OQO8lvfo5gPncf8LNsXTDcDbja5fC2WpNhLO7H7pw5cT0d2gIZ1kZvhm2UAlnS5HeSnd/7YRvKM+ccHYAui/p9W/Zmk94UYqzsCzDcDYjgQlnqzUezh4of6/cufpHaAaEs6wIZzlob6V7y+9WunLqoclfSTuPex46J9/tI+HsAEy82s+pel1ieYSz/DS/qHlM8y+Ut6bePBPOsjIjnP2vW09tfefpre889X/83/1DTQlnS1E/W1W3T/PLAXbPr97eV8LZAdhbOIv/7GFphLP8vPtK9YXljpNHv1a8oJT/PHnh0YdPhX+3178g+NSpEw/f377aCGdZmQpnn3vkxV/FLh988PaL/89Wv0NVwtlSxGerb5x8u/jBwHvLG2UPvA7tM+HsAOwxnIXbq3hZYlmG4azzB526f6NpxiH23VtPPHBP8e/+8i/R9H7pUvrXaYSzrEyEs5On4/HamZP/ut+nKOFsKcrXnfJfNsNfDrAw31+z74SzA7DncOZ3AS7VMJyRjeZmif8muXKleNtsp3/3C2dZGQ9nW8+8Fo/XXn3mc70+ZQlnSxFupfLdrvFfDrCY8u03luDtF04de/jbPY+eC89y4+HsrXOPxk6Nh4+dKn5yjSXohrO3no9fl1nEM36keXmEs4wl4az+N8mOf8BZOMuMd85y8MoTTxRvOp/79p6/2cz3PS/NuW/Ha9pR/n7ZV05Uv9+0dvj+E+FZbeI31X7bb9laim4446AJZxlLw1n15tk83y4jnGXF95xlYwnfbOavPy/R1BfOvnFyNP++NfWup9eu5ei+3uze+vx157xN3R8LKr8LjSXr3Sx33//APC8uwllWpsJZKD+tuUrNN8nunm82W6rpF5/D9zxwJH7dMjrywD2Tv/pEOFuOZYUzb7wtx5LCmftjP+zuZhHOsnL6O/28tVBtPRfnYY/ePffQQn++sefwfQ+d8obAUu3ixefwPXdXGa3ZCLz4LMfM3+G0AO+cLce7F47u6ZcxRt452w/C2VStUTj71XMn/3AQueatvz552g+nc4taOJx9rfxNKA/f026UhDNgpYSzqVqjcFa4+kL/z2jOVS+8Jplx61o4nN39jWOnTj58f7JREs6AlRLOpmrNwhkw0P+RzN2pfpATYGV29V0ya/RX/4QzAICMCGcAABkRzgAAMrJW4az/55VOPPJqPDKfS/fGIVcf2Vrst5ed/tY/fu741WLrzMmRv/tER/mNmrv4FvFi3IZ/Z3n/G/RnX47QeyXf3rq38yz7w7qqR70fBj+AsdAjcYMsS/FxWNon0exPyPboOn/i7re93RjzaK7+2txFaxfOkkAWctJi+awJZwtrwxk7CnfBkSPhfwt//nvtKZ+j2ktQXJBZV2RVz/Z7O49w1up+fBflBlmOcB3vO3LkvmV9Fs3+hFznT9fV6d0YxWf6kq/a+n0c1jmcVb/6/1uXys3izbDqHbXmLbH2DwMUfZoOYYb6nbNiwpP31t2aga8dPxEHlqdrdot81r5zNjjjxGybJtwE4S7r3ArFTngy7P2TqLgfo+q2jK89nZegarbqWKUzba+pnbKZYL0UDyBdevok1V6CsqnZrY73jnaEWe87dixem2r6cuL7iiG9yzu4cOl55hjVfgiqxmb/yDNF9yPH4pgwW32oGTycr1h55zMnXU3VZa0UD7m9Vq3dXMnkEpXXotgoPzbNbqW5Uu2EY0vYHOHqhWuSXsPy+hf/okwvz1hboXthi17tXvKBLBs6R5uPVNIrTjt1ro1RXNT0YZcXJO4PrlZ1JbtPaMHoVa1uif/8nfbjUE9d/nd4zetZyuedpvUgrHc4a6JSm9Kat9M6KaoaNfiyZjFhfEusmKHqX8wQ/yZ6M23Y6H1Zs9u/nHZ0to0T7pvyMzp8jsdnovh0VrdWW8VG8kxVbhYb9bF4V8TZhi3JBMVW2VbesvWUzcnXSnGl4gOtNI88OdJstg947GiraIo966PNRauPV9vF1vDKhdaqcadRxfE4vN1sHkI5umorRvSHjMxX9WsHl1uhrTq4htrHmBp55O1lSTaLreZi1NM0l6MZG3THlH2bfkVT3W8TNQ8/XJH6IpZXqWmttsba2suZbLVXtugXp2w2O0fLreJQb97Rc22S4lHXH4yguB7xIrRHutct2ayvc/8Kth+iqrU6XLQWjaPXvD1x0ZYuafVujXAWUlfzTlUdvNpw1hgNZ/WEI/3L98zGw1k9Q6HYLo7uNNtGCJ/S8RM6fJpXn+XV53nbOviMr2+b5lgztJ5tOKpoqaevJ6jnWWPJlaqMXa+mbfQBj4xIZ42H0+s3crirOc8iozqDyoPJ6GRA3W90vqQxbavXsHaKx5OoHtroI080D7g5mPZqjibTFMebaxQ7NP02XLg29XUIl2TkgtUXaqytcxXDdjk+bWs0H6L2aNxqjgT1OUbPtUmKa9lRX6HiQHq16mvab9zpqraXte6ZHq0P9meptw/EeoezIjzV4az6YmJVVZDqfDmyMBrO4ptkSZwqjjZTTYSzNA7WR8dn2yy9eyx+7jf3RfoZ3+nauzPCVtESujR3R9O96RH3o7K5aa1Pt26KR9k85ELb0H3EZVs4mF7XVmeKTr+yYzgc/1NqLm0U+rZNRadm/A6j+q29QcnoZLOefHS+5szJkKRt7RQPsr6AjdFH3m2tmsauZHs5kutSHO9oRpXW9vItQf/SVJdx7IJOtTVXr+4weeHL0e3RuBX+08xbt42da6MkF6VzjcsDqfJIeonipUsmGL2q7Zi6deyap7N0Rh+E9Q5nxVcPi/DUpK5Rzbtcc4WzOvAVFn/nbLPDWfHp3Nw0yWd3c1+0beldUB9O7oawWX5TwfDmiJ2TviNmH81YelkKxQNprk3/GrbXdfRoq3exi+20VzPNlM55Zo3qnag3KBmdbNb9xuZLG5shox3XRFh7c4EaYw8o7dgcn7qS1dFkmuT4iNlHb229D0B9zdJLMqstvcjNXE1b+7FIRrcj4lY6bz1i9FybpL6WpeJq1HujVyPtHbd3uqrtRHVrerQ+2J+lGX0Q1jmchQBU79YprepTxKYmV5Xhqeq2aDgr3h6bCGflGatuzTI2PpwVn82dG6m+h5r7ov2Mrw9VLeN3RucGrCcIm2VrcrJirrCZDA+bdf/10l6WQnkRqt3ew61am8syerSVtNUnSC5W29g5Y6Jznlmjek29QcnoZLOZfGwVzcFkSNK2dtrHmBp55L2mkSuZbDXXr7kuRXPcKWYKm83Y8mDTccP0r399nZKr01zHsbayMbny5VbnYLyyxYl6R5ut4ljV1GyNnmuTFFeivgBB70pWRzpXvG5sDhcbM65q0VodrltHr3l74mKWdvRBWLtwlnzBMQ1qZd6q2mOKmmwJo2aFs5jJijp5Ok1s1Txtt3b++Bbapoez9vO6EW+T5lM/uSOKjdJ9x56Jbend0t2Jt0qpbRy0NXP217E22odUGbsCR56pr2f1eMvNkaOtoqX+mcd4qHd5Z1+55jw7jmpa2o9qvbbOzxgmm8lyh/P1DpZDql5183opLkVyARvDR960jF3J9nD5/nI1JrlWhapvIZ5weI6NM7z85UUZ//HXsba43b2M1YUt95pjYzdpO0nVVkg+NmPn2hjFhet8ZIqGwdWqr0txiXpPaMHMqxqPht51a3o0ueb1LOXPlXeWtGprFc7YHJ37ir3ZyKd74Na0iie04hXoYJ81hTPyU/yrSTRbHuEMuGXs3xNa+dITHfQrkHAGAJAR4QwAICPCGQBARoQzAICMCGcAABkRzgAAMiKcAQBkZE/hDACApeuFrrRmhTOllFJKKbXiEs6UUkoppTIq4UwppZRSKqMSzpRSSimlMirhTCmllFIqoxLOlFJKKaUyKuFMKaWUUiqjEs6UUkoppTIq4UwppZRSKqMSzpRSSimlMirhTCmllFIqoxLOlFJKKaUyKuFMKaWUUiqjEs6UUkoppTIq4UwppZRSKqMSzpRSSimlMirhTCmllFIqoxLOlFJKKaWyqQ8//P8B+0aVXmTVgyIAAAAASUVORK5CYII=" width="450px"></p>
<blockquote>
<p><strong><em>NOTE</em></strong> At the moment <code><a href="../reference/GUI-utils.html">newProject()</a></code> <em>only</em> works with (recent) versions of RStudio.</p>
</blockquote>
</div>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<p>Collecting features from the analyses consists of finding all features, grouping them across analyses (optionally after retention time alignment) and finally, if desired suspect screening:</p>
<div id="htmlwidget-59cc71c1539139298d3b" style="width:500px;height:70px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-59cc71c1539139298d3b">{"x":{"diagram":"\ndigraph Features {\n  graph [ rankdir = LR, compound = true ]\n  node [ shape = box,\n         fixedsize = true,\n         width = 2.2,\n         height = 1,\n         fontsize = 18,\n         fillcolor = darkseagreen1,\n         style = filled ]\n\n    \"Find features\" -> \"Group features\"\n    \"Suspect screening\"\n    \"Group features\" -> \"Suspect screening\" [dir=both, style=dashed]\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><div id="finding-and-grouping-features" class="section level3">
<h3 class="hasAnchor">
<a href="#finding-and-grouping-features" class="anchor"></a>Finding and grouping features</h3>
<p>Several algorithms are available for finding features. These are listed in the table below alongside their usage and general remarks.</p>
<table class="table">
<colgroup>
<col width="21%">
<col width="58%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th>Algorithm</th>
<th>Usage</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://openms.de/">OpenMS</a></td>
<td><code><a href="../reference/feature-finding.html">findFeatures(algorithm = "openms", ...)</a></code></td>
<td>Uses the <a href="https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FeatureFinderMetabo.html">FeatureFinderMetabo</a> algorithm</td>
</tr>
<tr class="even">
<td><a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a></td>
<td><code><a href="../reference/feature-finding.html">findFeatures(algorithm = "xcms", ...)</a></code></td>
<td>Uses <code><a href="https://rdrr.io/pkg/xcms/man/xcmsSet.html">xcms::xcmsSet()</a></code> function</td>
</tr>
<tr class="odd">
<td>
<a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a> (import)</td>
<td><code><a href="../reference/feature-finding.html">importFeatures(algorithm = "xcms", ...)</a></code></td>
<td>Imports an existing <code>xcmsSet</code> object.</td>
</tr>
<tr class="even">
<td><a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS3</a></td>
<td><code><a href="../reference/feature-finding.html">findFeatures(algorithm = "xcms3", ...)</a></code></td>
<td>Uses <code><a href="https://rdrr.io/pkg/xcms/man/chromatographic-peak-detection.html">xcms::findChromPeaks()</a></code> from the new XCMS3 interface</td>
</tr>
<tr class="odd">
<td>
<a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS3</a> (import)</td>
<td><code><a href="../reference/feature-finding.html">importFeatures(algorithm = "xcms3", ...)</a></code></td>
<td>Imports an existing <code>XCMSnExp</code> object.</td>
</tr>
<tr class="even">
<td><a href="https://cran.r-project.org/web/packages/enviPick/index.html">enviPick</a></td>
<td><code><a href="../reference/feature-finding.html">findFeatures(algorithm = "envipick", ...)</a></code></td>
<td>Uses <code><a href="https://rdrr.io/pkg/enviPick/man/enviPickwrap.html">enviPick::enviPickwrap()</a></code>
</td>
</tr>
<tr class="odd">
<td>DataAnalysis</td>
<td><code><a href="../reference/feature-finding.html">findFeatures(algorithm = "bruker", ...)</a></code></td>
<td>Uses Find Molecular Features from DataAnalysis (Bruker only)</td>
</tr>
</tbody>
</table>
<p>Most often the performance of these algorithms heavily depend on the data and parameter settings that are used. Since obtaining a good feature dataset is crucial for the rest of the workflow, it is highly recommended to experiment with different settings (this process can also be automated, see <a href="#fOpt">the feature optimization section</a> for more details). Some common parameters to look at are listed in the table below. However, there are many more (advanced) parameters that can be set, please see the reference documentation for these (e.g. <code><a href="../reference/feature-finding.html">?findFeatures</a></code>).</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="83%">
</colgroup>
<thead><tr class="header">
<th>Algorithm</th>
<th>Common parameters</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://openms.de/">OpenMS</a></td>
<td>
<code>noiseThrInt</code>, <code>chromSNR</code>, <code>chromFWHM</code>, <code>mzPPM</code>, <code>minFWHM</code>, <code>maxFWHM</code> (see <code><a href="../reference/feature-finding.html">?findFeatures</a></code>)</td>
</tr>
<tr class="even">
<td>
<a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a> / <a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS3</a>
</td>
<td>
<code>peakwidth</code>, <code>mzdiff</code>, <code>prefilter</code>, <code>noise</code> (assuming default <code>centWave</code> algorithm, see <code>?findPeaks.centWave</code> / <code>?CentWaveParam</code>)</td>
</tr>
<tr class="odd">
<td><a href="https://cran.r-project.org/web/packages/enviPick/index.html">enviPick</a></td>
<td>
<code>dmzgap</code>, <code>dmzdens</code>, <code>drtgap</code>, <code>drtsmall</code>, <code>drtdens</code>, <code>drtfill</code>, <code>drttotal</code>, <code>minpeak</code>, <code>minint</code>, <code>maxint</code> (see <code>?enviPickwrap</code>)</td>
</tr>
<tr class="even">
<td>DataAnalysis</td>
<td>See <em>Find</em> -&gt; <em>Parameters…</em> -&gt; <em>Molecular Features</em> in DataAnalysis.</td>
</tr>
</tbody>
</table>
<blockquote>
<p><strong><em>NOTE</em></strong> DataAnalysis feature settings have to be configured in DataAnalysis prior to calling <code><a href="../reference/feature-finding.html">findFeatures()</a></code>.</p>
</blockquote>
<p>Similarly, for grouping features across analyses several algorithms are supported.</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="51%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Algorithm</th>
<th>Usage</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://openms.de/">OpenMS</a></td>
<td><code><a href="../reference/feature-grouping.html">groupFeatures(algorithm = "openms", ...)</a></code></td>
<td>Uses the <a href="https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FeatureLinkerUnlabeled.html">FeatureLinkerUnlabeled</a> algorithm (and <a href="https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_MapAlignerPoseClustering.html">MapAlignerPoseClustering</a> for retention alignment)</td>
</tr>
<tr class="even">
<td><a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a></td>
<td><code><a href="../reference/feature-grouping.html">groupFeatures(algorithm = "xcms", ...)</a></code></td>
<td>Uses <code><a href="https://rdrr.io/pkg/xcms/man/group-methods.html">xcms::group()</a></code> <code><a href="https://rdrr.io/pkg/xcms/man/retcor-methods.html">xcms::retcor()</a></code> functions</td>
</tr>
<tr class="odd">
<td>
<a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a> (import)</td>
<td><code><a href="../reference/feature-grouping.html">importFeatureGroupsXCMS(...)</a></code></td>
<td>Imports an existing <code>xcmsSet</code> object.</td>
</tr>
<tr class="even">
<td><a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS3</a></td>
<td><code><a href="../reference/feature-grouping.html">groupFeatures(algorithm = "xcms3", ...)</a></code></td>
<td>Uses <code><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html">xcms::groupChromPeaks()</a></code> and <code><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html">xcms::adjustRtime()</a></code> functions</td>
</tr>
<tr class="odd">
<td>
<a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS3</a> (import)</td>
<td><code><a href="../reference/feature-grouping.html">importFeatureGroupsXCMS3(...)</a></code></td>
<td>Imports an existing <code>XCMSnExp</code> object.</td>
</tr>
<tr class="even">
<td>ProfileAnalysis</td>
<td><code><a href="../reference/feature-grouping.html">importFeatureGroups(algorithm = "brukerpa", ...)</a></code></td>
<td>Import <code>.csv</code> file exported from Bruker ProfileAnalysis</td>
</tr>
<tr class="odd">
<td>TASQ</td>
<td><code><a href="../reference/feature-grouping.html">importFeatureGroups(algorithm = "brukertasq", ...)</a></code></td>
<td>Imports a <em>Global result table</em> (exported to Excel file and then saved as <code>.csv</code> file)</td>
</tr>
</tbody>
</table>
<p>Just like finding features, each algorithm has their own set of parameters. Often the defaults are a good start but it is recommended to have look at them. See <code><a href="../reference/feature-grouping.html">?groupFeatures</a></code> for more details.</p>
<p>When using the <a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a> algorithms both the ‘classical’ interface and latest <code>XCMS3</code> interfaces are supported. Currently, both interfaces are mostly the same regarding functionalities and implementation. However, since future developments of XCMS are primarily focused the latter this interface is recommended.</p>
<p>Some examples of finding and grouping features are shown below.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="co"># The anaInfo variable contains analysis information, see the previous section</span>

<span class="co"># Finding features</span>
<span class="va">fListOMS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-finding.html">findFeatures</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"openms"</span><span class="op">)</span> <span class="co"># OpenMS, with default settings</span>
<span class="va">fListOMS2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-finding.html">findFeatures</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"openms"</span>, noiseThrInt <span class="op">=</span> <span class="fl">500</span>, chromSNR <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># OpenMS, adjusted minimum intensity and S/N</span>
<span class="va">fListXCMS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-finding.html">findFeatures</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"xcms"</span>, ppm <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># XCMS</span>
<span class="va">fListXCMSImp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-finding.html">importFeatures</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"xcms"</span>, <span class="va">xset</span><span class="op">)</span> <span class="co"># import XCMS xcmsSet object</span>
<span class="va">fListXCMS3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-finding.html">findFeatures</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"xcms3"</span>, <span class="fu">CentWaveParam</span><span class="op">(</span>peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">15</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># XCMS3</span>
<span class="va">fListEP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-finding.html">findFeatures</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"envipick"</span>, minint <span class="op">=</span> <span class="fl">1E3</span><span class="op">)</span> <span class="co"># enviPick</span>

<span class="co"># Grouping features</span>
<span class="va">fGroupsOMS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-grouping.html">groupFeatures</a></span><span class="op">(</span><span class="va">fListOMS</span>, <span class="st">"openms"</span><span class="op">)</span> <span class="co"># OpenMS grouping, default settings</span>
<span class="va">fGroupsOMS2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-grouping.html">groupFeatures</a></span><span class="op">(</span><span class="va">fListOMS2</span>, <span class="st">"openms"</span>, rtalign <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># OpenMS grouping, no RT alignment</span>
<span class="va">fGroupsOMS3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-grouping.html">groupFeatures</a></span><span class="op">(</span><span class="va">fListXCMS</span>, <span class="st">"openms"</span>, maxGroupRT <span class="op">=</span> <span class="fl">6</span><span class="op">)</span> <span class="co"># group XCMS features with OpenMS, adjusted grouping parameter</span>
<span class="co"># group enviPick features with XCMS3, disable minFraction</span>
<span class="va">fGroupsXCMS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-grouping.html">groupFeatures</a></span><span class="op">(</span><span class="va">fListEP</span>, <span class="st">"xcms3"</span>,
                             <span class="fu">xcms</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks-density.html">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="va">analInfo</span><span class="op">$</span><span class="va">group</span>,
                                                    minFraction <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="suspect-screening" class="section level3">
<h3 class="hasAnchor">
<a href="#suspect-screening" class="anchor"></a>Suspect screening</h3>
<p>After features have been grouped suspect screening may be performed to eliminate any feature groups with <em>m/z</em> and (optionally) retention properties not present in a given suspect list. A typical workflow looks like this:</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="va">suspects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"susp1"</span>, <span class="st">"susp2"</span>, <span class="st">"susp3"</span><span class="op">)</span>,
                       mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">100.1034</span>, <span class="fl">250.0456</span>, <span class="fl">300.1234</span><span class="op">)</span>,
                       stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">scr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/suspect-screening.html">screenSuspects</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">suspects</span><span class="op">)</span>
<span class="va">fGroupsSusp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/suspect-screening.html">groupFeaturesScreening</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">scr</span><span class="op">)</span></pre></div>
<!-- UNDONE: make this interactive -->
<p>This will perform the following steps:</p>
<ol style="list-style-type: decimal">
<li>Create a suspect list. A suspect list is a <code>data.frame</code> with mandatory <code>name</code> and <code>mz</code> columns. Optionally, it may also contain an <code>rt</code> column for retention times (in seconds). In reality loading such a table is easiest from a <em>.csv</em> file with e.g. <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code>.</li>
<li>Screen a feature groups object for the suspects. The return value will be a <code>data.frame</code> with results.</li>
<li>Use the screening results to filter the original feature groups and store the results in <code>fGroupsSusp</code>.</li>
</ol>
<p>Alternatively, the <code>mz</code> column of the input suspect list can also be substituted by a <code>neutralMass</code>, <code>SMILES</code>, <code>InChI</code> or <code>formula</code> column, so that ion masses don’t need to be calculated in advance. This is especially advantageous for e.g. NORMAN suspect lists where such information is readily available. A requirement is that the adduct species should be specified for automatic ion mass calculation. This can either be done with an extra <code>adduct</code> column in the suspect list, or by using the <code>adduct</code> function argument to <code>screenSuspects</code>, e.g.:</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="va">suspects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"susp1"</span>, <span class="st">"susp2"</span>, <span class="st">"susp3"</span><span class="op">)</span>,
                       SMILES <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"C1=CC=C(C=C1)C(=O)O"</span>, <span class="st">"C1=CC=C2C(=C1)C=CC3=CC=CC=C3N2C(=O)N"</span>, <span class="st">"C1=CC2=NNN=C2C=C1"</span><span class="op">)</span>,
                       stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">scr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/suspect-screening.html">screenSuspects</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">suspects</span>, adduct <span class="op">=</span> <span class="st">"[M+H]+"</span><span class="op">)</span>
<span class="va">fGroupsSusp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/suspect-screening.html">groupFeaturesScreening</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">scr</span><span class="op">)</span></pre></div>
<p>Note that in some cases you may need to install <a href="http://openbabel.org/wiki/Main_Page">OpenBabel</a> (e.g. when only InChI data is available for mass calculation).</p>
</div>
</div>
<div id="annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#annotation" class="anchor"></a>Annotation</h2>
<p>The annotation consists of collecting MS peak lists and then formula and/or compound annotation:</p>
<div id="htmlwidget-047243cfa4c53a566038" style="width:500px;height:120px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-047243cfa4c53a566038">{"x":{"diagram":"\ndigraph Annotation {\n  graph [ rankdir = LR ]\n  node [ shape = box,\n         fixedsize = true,\n         width = 2.3,\n         height = 1,\n         fontsize = 18,\n         fillcolor = darkseagreen1,\n         style = filled ]\n\n    \"MS peak lists\" -> \"Formula annotation\"\n    \"MS peak lists\" -> \"Compound annotation\"\n    \"Formula annotation\":e -> \"Compound annotation\":e [style = dotted, constraint = false]\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p>Note that compound annotation is normally not dependent upon formula annotation. However, formula data can be used to improve ranking of candidates afterwards by the <code><a href="../reference/compounds-class.html">addFormulaScoring()</a></code> function, which will be discussed later in this section.</p>
<div id="ms-peak-lists" class="section level3">
<h3 class="hasAnchor">
<a href="#ms-peak-lists" class="anchor"></a>MS peak lists</h3>
<table class="table">
<colgroup>
<col width="10%">
<col width="45%">
<col width="44%">
</colgroup>
<thead><tr class="header">
<th>Algorithm</th>
<th>Usage</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://bioconductor.org/packages/release/bioc/html/mzR.html">mzR</a></td>
<td><code><a href="../reference/MSPeakLists-generation.html">generateMSPeakLists(algorithm = "mzr", ...)</a></code></td>
<td>Uses <a href="https://bioconductor.org/packages/release/bioc/html/mzR.html">mzR</a> for spectra retrieval. Recommended default.</td>
</tr>
<tr class="even">
<td>DataAnalysis</td>
<td><code><a href="../reference/MSPeakLists-generation.html">generateMSPeakLists(algorithm = "bruker", ...)</a></code></td>
<td>Loads data after automatically generating MS and MS/MS spectra in DataAnalysis</td>
</tr>
<tr class="odd">
<td>DataAnalysis FMF</td>
<td><code><a href="../reference/MSPeakLists-generation.html">generateMSPeakLists(algorithm = "brukerfmf", ...)</a></code></td>
<td>Uses spectra from the <em>find molecular features</em> algorithm.</td>
</tr>
</tbody>
</table>
<p>The recommended default algorithm is <code>mzr</code>: this algorithm is generally faster and is not limited to a vendor data format as it will read the open <code>mzML</code> and <code>mzXML</code> file formats. On the other hand, when DataAnalysis is used with Bruker data the spectra can be automatically background subtracted and there is no need for file conversion. Note that the <code>brukerfmf</code> algorithm only works when <code><a href="../reference/feature-finding.html">findFeatures()</a></code> was called with the <code>bruker</code> algorithm.</p>
<p>When <code>generateMSPeakists()</code> is called it will</p>
<ol style="list-style-type: decimal">
<li>Find all MS and MS/MS spectra that ‘belong’ to a feature. For MS spectra this means that all spectra close to the retention time of a feature will be collected. In addition, for MS/MS normally only spectra will be considered that have a precursor mass close to that of the feature (however, this can be disabled for data that was recorded with data independent acquisition (DIA, MS^E, bbCID, …)).</li>
<li>Average all MS and MS/MS spectra to produce peak lists for each feature.</li>
<li>Average all peak lists for features within the same group.</li>
</ol>
<p>Data from either (2) or (3) is used for subsequent annotation steps. Formula calculation can use either (as a trade-off between possibly more accurate results by outlier removal <em>vs</em> speed), whereas compound annotation will always use data from (3) since annotating single features (as opposed to their groups) would take a very long time.</p>
<p>There are several common function arguments to <code><a href="../reference/MSPeakLists-generation.html">generateMSPeakLists()</a></code> that can be used to optimize its behaviour:</p>
<!-- UNDONE: add topMost for Bruker when it's there -->
<table class="table">
<colgroup>
<col width="28%">
<col width="16%">
<col width="55%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Algorithm(s)</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>maxMSRtWindow</code></td>
<td>
<code>mzr</code>, <code>bruker</code>
</td>
<td>Maximum time window +/- the feature retention time (in seconds) to collect spectra for averaging. Higher values may significantly increase processing times.</td>
</tr>
<tr class="even">
<td><code>precursorMzWindow</code></td>
<td><code>mzr</code></td>
<td>Maximum precursor <em>m/z</em> search window to find MS/MS spectra. Set to <code>NULL</code> to disable (i.e. for DIA experiments).</td>
</tr>
<tr class="odd">
<td><code>topMost</code></td>
<td><code>mzr</code></td>
<td>Only retain feature data for no more than this amount analyses with highest intensity. For instance, a value of <em>1</em> will only keep peak lists for the feature with highest intensity in a feature group.</td>
</tr>
<tr class="even">
<td><code>bgsubtr</code></td>
<td><code>bruker</code></td>
<td>Perform background subtraction (if the spectra type supports this, e.g. MS and bbCID)</td>
</tr>
<tr class="odd">
<td>
<code>minMSIntensity</code>, <code>minMSMSIntensity</code>
</td>
<td>
<code>bruker</code>, <code>brukerfmf</code>
</td>
<td>Minimum MS and MS/MS intensity. Note that DataAnalysis reports many zero intensity peaks so a value of at least <em>1</em> is recommended.</td>
</tr>
<tr class="even">
<td><code>MSMSType</code></td>
<td><code>bruker</code></td>
<td>The type of spectra that should be used for MSMS: <code>"BBCID"</code> for bbCID experiments, otherwise <code>"MSMS"</code> (the default).</td>
</tr>
</tbody>
</table>
<p>In addition, several parameters can be set that affect spectral averaging. These parameters are passed as a <code>list</code> to the <code>avgFeatParams</code> (<code>mzr</code> algorithm only) and <code>avgFGroupParams</code> arguments, which affect averaging of feature and feature group data, respectively. Some typical parameters include:</p>
<ul>
<li>
<code>clusterMzWindow</code>: Maximum <em>m/z</em> window used to cluster mass peaks when averaging. The better the MS resolution, the lower this value should be.</li>
<li>
<code>topMost</code>: Retain no more than this amount of most intense mass peaks. Useful to filter out ‘noisy’ peaks.</li>
<li>
<code>minIntensityPre</code> / <code>minIntensityPost</code>: Mass peaks below this intensity will be removed before/after averaging.</li>
</ul>
<p>See <code><a href="../reference/MSPeakLists-generation.html">?generateMSPeakLists</a></code> for all possible parameters.</p>
<p>A suitable list object to set averaging parameters can be obtained with the <code><a href="../reference/MSPeakLists-generation.html">getDefAvgPListParams()</a></code> function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="co"># lower default clustering window, other settings remain default</span>
<span class="va">avgPListParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSPeakLists-generation.html">getDefAvgPListParams</a></span><span class="op">(</span>clusterMzWindow <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span>

<span class="co"># Apply to both feature and feature group averaging</span>
<span class="va">plists</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSPeakLists-generation.html">generateMSPeakLists</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"mzr"</span>, avgFeatParams <span class="op">=</span> <span class="va">avgPListParams</span>, avgFGroupParams <span class="op">=</span> <span class="va">avgPListParams</span><span class="op">)</span></pre></div>
</div>
<div id="formulae" class="section level3">
<h3 class="hasAnchor">
<a href="#formulae" class="anchor"></a>Formulae</h3>
<p>Formulae can be automatically calculated for all features using the <code><a href="../reference/formula-generation.html">generateFormulas()</a></code> function. The following algorithms are currently supported:</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="43%">
<col width="45%">
</colgroup>
<thead><tr class="header">
<th>Algorithm</th>
<th>Usage</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://sourceforge.net/projects/genform/">GenForm</a></td>
<td><code><a href="../reference/formula-generation.html">generateFormulas(algorithm = "genform", ...)</a></code></td>
<td>Bundled with <code>patRoon</code>. Reasonable default.</td>
</tr>
<tr class="even">
<td><a href="https://bio.informatik.uni-jena.de/software/sirius/">SIRIUS</a></td>
<td><code><a href="../reference/formula-generation.html">generateFormulas(algorithm = "sirius", ...)</a></code></td>
<td>Requires MS/MS data.</td>
</tr>
<tr class="odd">
<td>DataAnalysis</td>
<td><code><a href="../reference/formula-generation.html">generateFormulas(algorithm = "bruker", ...)</a></code></td>
<td>Requires FMF features (i.e. <code><a href="../reference/feature-finding.html">findFeatures(algorithm = "bruker", ...)</a></code>). MS peak lists are not needed. Uses <em>SmartFormula</em> algorithms.</td>
</tr>
</tbody>
</table>
<p>Calculation with <a href="https://sourceforge.net/projects/genform/">GenForm</a> is often a good default. It is fast and basic rules can be applied to filter out obvious non-existing formulae. A possible drawback of GenForm, however, is that may become slow when many candidates are calculated, for instance, due to a relative high feature <em>m/z</em> (e.g. &gt;600) or loose elemental restricitions. More thorough calculation is performed with <a href="https://bio.informatik.uni-jena.de/software/sirius/">SIRIUS</a>: this algorithm often yields fewer and often more plausible results. However, <a href="https://bio.informatik.uni-jena.de/software/sirius/">SIRIUS</a> requires MS/MS data (hence features without will not have results) and formula prediction may not work well for compounds that structurally deviate from the training sets used by <a href="https://bio.informatik.uni-jena.de/software/sirius/">SIRIUS</a>. Calculation with DataAnalysis is only possible when features are obtained with DataAnalysis as well. An advantage is that analysis files do not have to be converted and no MS peak generation is necessary, however, compared to other algorithms calculation is often relative slow.</p>
<p>There are two methods for formula assignment:</p>
<ol style="list-style-type: decimal">
<li>Formulae are first calculated for each individual feature within a feature group. These results are then pooled, outliers are removed and remaining formulae are assigned to the feature group (i.e. <code>calculateFeatures = TRUE</code>).</li>
<li>Formulae are directly calculated for each feature group by using group averaged peak lists (see previous section) (i.e. <code>calculateFeatures = FALSE</code>).</li>
</ol>
<p>The first method is more thorough and the possibility to remove outliers may sometimes result in better formula assignment. However, the second method is much faster and generally recommended for large number of analyses.</p>
<p>By default, formulae are either calculated by <em>only</em> MS/MS data (SIRIUS) or with both MS <em>and</em> MS/MS data (GenForm/Bruker). The latter also allows formula calculation when no MS/MS data is present. Furthermore, with Bruker algorithms, data from both MS and MS/MS formula data can be combined to allow inclusion of candidates that would otherwise be excluded by e.g. poor MS/MS data. However, a disadvantage is that formulae needs to be calculated twice. The <code>MSMode</code> argument (listed below) can be used to customize this behaviour.</p>
<p>An overview of common parameters that are typically set to customize formula calculation is listed below.</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="16%">
<col width="68%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Algorithm(s)</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>relMzDev</td>
<td>
<code>genform</code>, <code>sirius</code>
</td>
<td>The maximum relative <em>m/z</em> deviation for a formula to be considered (in <em>ppm</em>).</td>
</tr>
<tr class="even">
<td>elements</td>
<td>
<code>genform</code>, <code>sirius</code>
</td>
<td>Which elements to consider. By default <code>"CHNOP"</code>. Try to limit possible elements as much as possible.</td>
</tr>
<tr class="odd">
<td>calculateFeatures</td>
<td>
<code>genform</code>, <code>sirius</code>
</td>
<td>Whether formulae should be calculated first for all features (see discussion above) (always <code>TRUE</code> with DataAnalysis).</td>
</tr>
<tr class="even">
<td>featThreshold</td>
<td>All</td>
<td>Minimum relative amount (<em>0-1</em>) amongst all features within a feature group that a formula candidate should be present (e.g. <em>1</em> means that a candidate is only considered if it was assigned to all features).</td>
</tr>
<tr class="odd">
<td>adduct</td>
<td>
<code>genform</code>, <code>sirius</code>
</td>
<td>The adduct to consider for calculation (e.g. <code>"[M+H]+"</code>, <code>"[M-H]-"</code>, more details in the <a href="#adducts">adduct section</a>).</td>
</tr>
<tr class="even">
<td>MSMode</td>
<td>
<code>genform</code>, <code>bruker</code>
</td>
<td>Whether formulae should be generated only from MS data (<code>"ms"</code>), MS/MS data (<code>"msms"</code>) or both (<code>"both"</code>). The latter is default, see discussion above.</td>
</tr>
<tr class="odd">
<td>profile</td>
<td><code>sirius</code></td>
<td>Instrument profile, e.g. <code>"qtof"</code>, <code>"orbitrap"</code>, <code>"fticr"</code>.</td>
</tr>
</tbody>
</table>
<p>Some typical examples:</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="va">formulasGF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"genform"</span>, <span class="va">mslists</span><span class="op">)</span> <span class="co"># GenForm, default settings</span>
<span class="va">formulasGF2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"genform"</span>, <span class="va">mslists</span>, calculateFeatures <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># direct feature group assignment (faster)</span>
<span class="va">formulasSIR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"sirius"</span>, <span class="va">mslists</span>, elements <span class="op">=</span> <span class="st">"CHNOPSClBr"</span><span class="op">)</span> <span class="co"># SIRIUS, common elements for pollutant</span>
<span class="va">formulasSIR2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"sirius"</span>, adduct <span class="op">=</span> <span class="st">"[M-H]-"</span><span class="op">)</span> <span class="co"># SIRIUS, negative ionization</span>
<span class="va">formulasBr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"bruker"</span>, MSMode <span class="op">=</span> <span class="st">"MSMS"</span><span class="op">)</span> <span class="co"># Only consider MSMS data (SmartFormula3D)</span></pre></div>
</div>
<div id="compounds" class="section level3">
<h3 class="hasAnchor">
<a href="#compounds" class="anchor"></a>Compounds</h3>
<p>An important step in a typical non-target workflow is structural identification for features of interest. Afterall, this information may finally reveal <em>what</em> a feature is. The first step is to find all possible structures in a database that may be assigned to the feature (based on e.g. monoisotopic mass or formula). These candidates are then scored to rank likely candidates, for instance, on correspondence with in-silico or library MS/MS spectra and environmental relevance.</p>
<p>Structure assignment in <code>patRoon</code> is performed automatically for all feature groups with the <code><a href="../reference/compound-generation.html">generateCompounds()</a></code> function. Currently, this function supports two algorithms:</p>
<table class="table">
<colgroup>
<col width="22%">
<col width="38%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th>Algorithm</th>
<th>Usage</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://ipb-halle.github.io/MetFrag/">MetFrag</a></td>
<td><code><a href="../reference/compound-generation.html">generateCompounds(algorithm = "metfrag", ...)</a></code></td>
<td>Supports many databases (including custom) and scorings for candidate ranking.</td>
</tr>
<tr class="even">
<td>
<a href="https://bio.informatik.uni-jena.de/software/sirius/">SIRIUS</a> with <a href="https://www.csi-fingerid.uni-jena.de/">CSI:FingerID</a>
</td>
<td><code><a href="../reference/compound-generation.html">generateCompounds(algorithm = "sirius", ...)</a></code></td>
<td>Incorporates prior comprohensive formula calculations.</td>
</tr>
</tbody>
</table>
<p>Compound annotation is often a relative time and resource intensive procedure. For this reason, features are not annotated individually, but instead a feature group as a whole is annotated, which generally saves significant amounts of computational requirements. Nevertheless, it is not uncommon that this is the most time consuming step in the workflow. For this reason, prioritization of features is highly important, even more so to avoid ‘abusing’ servers when an online database is used for compound retrieval.</p>
<p>Selecting the right database is important for proper candidate assignment. Afterall, if the ‘right’ chemical compound is not present in the used database, it is impossible to assign the correct structure. Luckily, however, several large databases such as <a href="https://pubchem.ncbi.nlm.nih.gov/">PubChem</a> and <a href="http://www.chemspider.com/">ChemSpider</a> are openly available which contain tens of millions of compounds. On the other hand, these databases may also lead to many unlikely candidates and therefore more specialized (or custom databases) may be preferred. Which database will be used is dictated by the <code>database</code> argument to <code><a href="../reference/compound-generation.html">generateCompounds()</a></code>, currently the following options exist:</p>
<!-- UNDONE: add missing databases when support is added -->
<table class="table">
<colgroup>
<col width="29%">
<col width="35%">
<col width="35%">
</colgroup>
<thead><tr class="header">
<th>Database</th>
<th>Algorithm(s)</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>pubchem</code></td>
<td>
<code>"metfrag"</code>, <code>"sirius"</code>
</td>
<td>
<a href="https://pubchem.ncbi.nlm.nih.gov/">PubChem</a> is currently the largest compound database and is used by default.</td>
</tr>
<tr class="even">
<td><code>chemspider</code></td>
<td><code>"metfrag"</code></td>
<td>
<a href="http://www.chemspider.com/">ChemSpider</a> is another large database. Requires security token from <a href="http://www.chemspider.com/AboutServices.aspx">here</a> (see next section).</td>
</tr>
<tr class="odd">
<td><code>comptox</code></td>
<td><code>"metfrag"</code></td>
<td>The EPA <a href="https://comptox.epa.gov/dashboard">CompTox</a> contains many compounds and scorings relevant to environmental studies. Needs manual download (see next section).</td>
</tr>
<tr class="even">
<td><code>pubchemlite</code></td>
<td><code>"metfrag"</code></td>
<td>A specialized subset of the PubChem database. Needs manual download (see next section).</td>
</tr>
<tr class="odd">
<td><code>for-ident</code></td>
<td><code>"metfrag"</code></td>
<td>The <a href="https://water.for-ident.org">FOR-IDENT</a> (STOFF-IDENT) database for water related substances.</td>
</tr>
<tr class="even">
<td><code>kegg</code></td>
<td>
<code>"metfrag"</code>, <code>"sirius"</code>
</td>
<td>The <a href="https://www.genome.jp/kegg/">KEGG</a> database for biological compounds</td>
</tr>
<tr class="odd">
<td><code>hmdb</code></td>
<td>
<code>"metfrag"</code>, <code>"sirius"</code>
</td>
<td>The <a href="http://www.hmdb.ca/">HMDB</a> contains many human metabolites.</td>
</tr>
<tr class="even">
<td><code>bio</code></td>
<td><code>"sirius"</code></td>
<td>Selects all supports biological databases.</td>
</tr>
<tr class="odd">
<td>
<code>csv</code>, <code>psv</code>, <code>sdf</code>
</td>
<td><code>"metfrag"</code></td>
<td>Custom database (see next section). <a href="https://raw.githubusercontent.com/rickhelmus/patRoon/master/tests/testthat/test_data/test-mf-db-isomers.csv">CSV example</a>.</td>
</tr>
</tbody>
</table>
<div id="configuring-metfrag-databases-and-scoring" class="section level4">
<h4 class="hasAnchor">
<a href="#configuring-metfrag-databases-and-scoring" class="anchor"></a>Configuring MetFrag databases and scoring</h4>
<p>Some extra configuration may be necessary when using certain databases with MetFrag. In order to use the ChemSpider database a <a href="http://www.chemspider.com/AboutServices.aspx">security token</a> should be requested and set with the <code>chemSpiderToken</code> argument to <code><a href="../reference/compound-generation.html">generateCompounds()</a></code>. The CompTox and PubChemLite databases need to be manually downloaded from <a href="ftp://newftp.epa.gov/COMPTOX/Sustainable_Chemistry_Data/Chemistry_Dashboard/MetFrag_metadata_files">CompTox</a> (or variations with <a href="https://zenodo.org/record/3364464#.XnjM-XLvKUk">smoking</a> or <a href="https://zenodo.org/record/3472781#.XnjMAHLvKUk">wastewater</a> metadata) and <a href="https://zenodo.org/record/3548654">PUbChemLite</a>. The file location of this and other local databases (<code>csv</code>, <code>psv</code>, <code>sdf</code>) needs to be manually configured, see the examples below and/or <code><a href="../reference/compound-generation.html">?generateCompounds</a></code> for more information on how to do this.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="co"># PubChem: the default</span>
<span class="va">compsMF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compound-generation.html">generateCompounds</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">mslists</span>, <span class="st">"metfrag"</span>, adduct <span class="op">=</span> <span class="st">"[M+H]+"</span><span class="op">)</span>

<span class="co"># ChemSpider: needs security token</span>
<span class="va">compsMF2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compound-generation.html">generateCompounds</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">mslists</span>, <span class="st">"metfrag"</span>, database <span class="op">=</span> <span class="st">"chemspider"</span>,
                              chemSpiderToken <span class="op">=</span> <span class="st">"MY_TOKEN_HERE"</span>, adduct <span class="op">=</span> <span class="st">"[M+H]+"</span><span class="op">)</span>

<span class="co"># CompTox: set global option to database path</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.path.MetFragCompTox <span class="op">=</span> <span class="st">"~/CompTox_17March2019_SelectMetaData.csv"</span><span class="op">)</span>
<span class="va">compsMF3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compound-generation.html">generateCompounds</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">mslists</span>, <span class="st">"metfrag"</span>, database <span class="op">=</span> <span class="st">"comptox"</span>, adduct <span class="op">=</span> <span class="st">"[M+H]+"</span><span class="op">)</span>

<span class="co"># CompTox: set database location without global option</span>
<span class="va">compsMF4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compound-generation.html">generateCompounds</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">mslists</span>, <span class="st">"metfrag"</span>, database <span class="op">=</span> <span class="st">"comptox"</span>, adduct <span class="op">=</span> <span class="st">"[M+H]+"</span>,
                              extraOpts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>LocalDatabasePath <span class="op">=</span> <span class="st">"~/CompTox_17March2019_SelectMetaData.csv"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Same, but for custom database</span>
<span class="va">compsMF5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compound-generation.html">generateCompounds</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="va">mslists</span>, <span class="st">"metfrag"</span>, database <span class="op">=</span> <span class="st">"csv"</span>, adduct <span class="op">=</span> <span class="st">"[M+H]+"</span>,
                              extraOpts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>LocalDatabasePath <span class="op">=</span> <span class="st">"~/mydb.csv"</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>An example of a custom <em>.csv</em> database can be found <a href="https://raw.githubusercontent.com/rickhelmus/patRoon/master/tests/testthat/test_data/test-mf-db-isomers.csv">here</a>.</p>
<p>With MetFrag compound databases are not only used to retrieve candidate structures but are also used to obtain metadata for further ranking. Each database has its own scorings, a table with currently supported scorings can be obtained with the <code><a href="../reference/compound-generation.html">compoundScorings()</a></code> function (some columns omitted):</p>
<div style="border: 1px solid #ddd; padding: 5px; overflow-y: auto; height: 350px;">
<table class="table table">
<thead><tr>
<th style="text-align:left;">
name
</th>
<th style="text-align:left;">
metfrag
</th>
<th style="text-align:left;">
database
</th>
<th style="text-align:left;">
default
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
score
</td>
<td style="text-align:left;">
Score
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
fragScore
</td>
<td style="text-align:left;">
FragmenterScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
metFusionScore
</td>
<td style="text-align:left;">
OfflineMetFusionScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
individualMoNAScore
</td>
<td style="text-align:left;">
OfflineIndividualMoNAScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
numberPatents
</td>
<td style="text-align:left;">
PubChemNumberPatents
</td>
<td style="text-align:left;">
pubchem
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
numberPatents
</td>
<td style="text-align:left;">
Patent_Count
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
pubMedReferences
</td>
<td style="text-align:left;">
PubChemNumberPubMedReferences
</td>
<td style="text-align:left;">
pubchem
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
pubMedReferences
</td>
<td style="text-align:left;">
ChemSpiderNumberPubMedReferences
</td>
<td style="text-align:left;">
chemspider
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
pubMedReferences
</td>
<td style="text-align:left;">
NUMBER_OF_PUBMED_ARTICLES
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
pubMedReferences
</td>
<td style="text-align:left;">
PubMed_Count
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
extReferenceCount
</td>
<td style="text-align:left;">
ChemSpiderNumberExternalReferences
</td>
<td style="text-align:left;">
chemspider
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
dataSourceCount
</td>
<td style="text-align:left;">
ChemSpiderDataSourceCount
</td>
<td style="text-align:left;">
chemspider
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
referenceCount
</td>
<td style="text-align:left;">
ChemSpiderReferenceCount
</td>
<td style="text-align:left;">
chemspider
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
RSCCount
</td>
<td style="text-align:left;">
ChemSpiderRSCCount
</td>
<td style="text-align:left;">
chemspider
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
formulaScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
smartsInclusionScore
</td>
<td style="text-align:left;">
SmartsSubstructureInclusionScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
smartsExclusionScore
</td>
<td style="text-align:left;">
SmartsSubstructureExclusionScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
suspectListScore
</td>
<td style="text-align:left;">
SuspectListScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
retentionTimeScore
</td>
<td style="text-align:left;">
RetentionTimeScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
CPDATCount
</td>
<td style="text-align:left;">
CPDAT_COUNT
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
TOXCASTActive
</td>
<td style="text-align:left;">
TOXCAST_PERCENT_ACTIVE
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
dataSources
</td>
<td style="text-align:left;">
DATA_SOURCES
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
pubChemDataSources
</td>
<td style="text-align:left;">
PUBCHEM_DATA_SOURCES
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
EXPOCASTPredExpo
</td>
<td style="text-align:left;">
EXPOCAST_MEDIAN_EXPOSURE_PREDICTION_MG/KG-BW/DAY
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
ECOTOX
</td>
<td style="text-align:left;">
ECOTOX
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
NORMANSUSDAT
</td>
<td style="text-align:left;">
NORMANSUSDAT
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
MASSBANKEU
</td>
<td style="text-align:left;">
MASSBANKEU
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
TOX21SL
</td>
<td style="text-align:left;">
TOX21SL
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
TOXCAST
</td>
<td style="text-align:left;">
TOXCAST
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
KEMIMARKET
</td>
<td style="text-align:left;">
KEMIMARKET
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
MZCLOUD
</td>
<td style="text-align:left;">
MZCLOUD
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
pubMedNeuro
</td>
<td style="text-align:left;">
PubMedNeuro
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
CIGARETTES
</td>
<td style="text-align:left;">
CIGARETTES
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
INDOORCT16
</td>
<td style="text-align:left;">
INDOORCT16
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
SRM2585DUST
</td>
<td style="text-align:left;">
SRM2585DUST
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
SLTCHEMDB
</td>
<td style="text-align:left;">
SLTCHEMDB
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
THSMOKE
</td>
<td style="text-align:left;">
THSMOKE
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
ITNANTIBIOTIC
</td>
<td style="text-align:left;">
ITNANTIBIOTIC
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
STOFFIDENT
</td>
<td style="text-align:left;">
STOFFIDENT
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
KEMIMARKET_EXPO
</td>
<td style="text-align:left;">
KEMIMARKET_EXPO
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
KEMIMARKET_HAZ
</td>
<td style="text-align:left;">
KEMIMARKET_HAZ
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
REACH2017
</td>
<td style="text-align:left;">
REACH2017
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
KEMIWW_WDUIndex
</td>
<td style="text-align:left;">
KEMIWW_WDUIndex
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
KEMIWW_StpSE
</td>
<td style="text-align:left;">
KEMIWW_StpSE
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
KEMIWW_SEHitsOverDL
</td>
<td style="text-align:left;">
KEMIWW_SEHitsOverDL
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
ZINC15PHARMA
</td>
<td style="text-align:left;">
ZINC15PHARMA
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
PFASMASTER
</td>
<td style="text-align:left;">
PFASMASTER
</td>
<td style="text-align:left;">
comptox
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
peakFingerprintScore
</td>
<td style="text-align:left;">
AutomatedPeakFingerprintAnnotationScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
lossFingerprintScore
</td>
<td style="text-align:left;">
AutomatedLossFingerprintAnnotationScore
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
agroChemInfo
</td>
<td style="text-align:left;">
AgroChemInfo
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
bioPathway
</td>
<td style="text-align:left;">
BioPathway
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
drugMedicInfo
</td>
<td style="text-align:left;">
DrugMedicInfo
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
foodRelated
</td>
<td style="text-align:left;">
FoodRelated
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
pharmacoInfo
</td>
<td style="text-align:left;">
PharmacoInfo
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
safetyInfo
</td>
<td style="text-align:left;">
SafetyInfo
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
toxicityInfo
</td>
<td style="text-align:left;">
ToxicityInfo
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
knownUse
</td>
<td style="text-align:left;">
KnownUse
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
annoTypeCount
</td>
<td style="text-align:left;">
FPSum
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
annoTypeCount
</td>
<td style="text-align:left;">
AnnoTypeCount
</td>
<td style="text-align:left;">
pubchemlite
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
</tbody>
</table>
</div>
<p>The first two columns contain the generic and original MetFrag naming schemes for each scoring type. While both naming schemes can be used, the generic is often shorter and harmonized with other algorithms (e.g. SIRIUS). The <em>database</em> column specifies for which databases a particular scoring is available (empty if not database specific). Most scorings are selected by default (as specified by the <em>default</em> column), however, this behaviour can be customized by using the <code>scoreTypes</code> argument:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="co"># Only in-silico and PubChem number of patents scorings</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>compsMF1 &lt;-<span class="st"> </span><span class="kw">generateCompounds</span>(fGroups, mslists, <span class="st">"metfrag"</span>, <span class="dt">adduct =</span> <span class="st">"[M+H]+"</span>,</span>
<span id="cb19-3"><a href="#cb19-3"></a>                              <span class="dt">scoreTypes =</span> <span class="kw">c</span>(<span class="st">"fragScore"</span> <span class="st">"numberPatents"</span>))</span>
<span id="cb19-4"><a href="#cb19-4"></a></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="co"># Custom scoring in custom database</span></span>
<span id="cb19-6"><a href="#cb19-6"></a>compsMF2 &lt;-<span class="st"> </span><span class="kw">generateCompounds</span>(fGroups, mslists, <span class="st">"metfrag"</span>, <span class="dt">adduct =</span> <span class="st">"[M+H]+"</span>,</span>
<span id="cb19-7"><a href="#cb19-7"></a>                              <span class="dt">database =</span> <span class="st">"csv"</span>, <span class="dt">adduct =</span> <span class="st">"[M+H]+"</span>,</span>
<span id="cb19-8"><a href="#cb19-8"></a>                              <span class="dt">extraOpts =</span> <span class="kw">list</span>(<span class="dt">LocalDatabasePath =</span> <span class="st">"~/mydb.csv"</span>),</span>
<span id="cb19-9"><a href="#cb19-9"></a>                              <span class="dt">scoreTypes =</span> <span class="kw">c</span>(<span class="st">"fragScore"</span>, <span class="st">"myScore"</span>, <span class="st">"myScore2"</span>))</span></code></pre></div>
<p>By default ranking is performed with equal weight (i.e. <em>1</em>) for all scorings. This can be changed by the <code>scoreWeights</code> argument, which should be a <code>vector</code> containing the weights for all scorings following the order of <code>scoreTypes</code>, for instance:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>compsMF &lt;-<span class="st"> </span><span class="kw">generateCompounds</span>(fGroups, mslists, <span class="st">"metfrag"</span>, <span class="dt">adduct =</span> <span class="st">"[M+H]+"</span>,</span>
<span id="cb20-2"><a href="#cb20-2"></a>                             <span class="dt">scoreTypes =</span> <span class="kw">c</span>(<span class="st">"fragScore"</span> <span class="st">"numberPatents"</span>),</span>
<span id="cb20-3"><a href="#cb20-3"></a>                             <span class="dt">scoreWeights =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span></code></pre></div>
<p>Sometimes thousands or more structural candidates are found when annotating a feature group. In this situation processing all these candidates will too involving (especially when external databases are used). To avoid this a default cut-off is set: when the number of candidates exceed a certain amount the search will be aborted and no results will be reported for that feature group. The maximum number of candidates can be set with the <code>maxCandidatesToStop</code> argument. The default value is relative conservative, especially for local databases it may be useful to increase this number.</p>
</div>
<div id="timeout-and-error-handling" class="section level4">
<h4 class="hasAnchor">
<a href="#timeout-and-error-handling" class="anchor"></a>Timeout and error handling</h4>
<p>The use of online databases has the drawback that an error may occur, for instance, as a result of a connection error. Furthermore, MetFrag typically returns an error when too many candidates are found (as set by the <code>maxCandidatesToStop</code> argument). By default processing is restarted if an error has occurred (configured by the <code>errorRetries</code> argument). Similarly, the <code>timeoutRetries</code> and <code>timeout</code> arguments can be used to avoid being ‘stuck’ on obtaining results, for instance, due to an unstable internet connection.</p>
<p>If no compounds could be assigned due to an error a warning will be issued. In this case it is best to see what went wrong by manually checking the log files, which by default are stored in the <em>log/metfrag</em> folder.</p>
</div>
<div id="formula-scoring" class="section level4">
<h4 class="hasAnchor">
<a href="#formula-scoring" class="anchor"></a>Formula scoring</h4>
<p>Ranking of candidate structures may further be improved by incorporating formula information by using the <code><a href="../reference/compounds-class.html">addFormulaScoring()</a></code> function:</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="va">comps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compounds-class.html">addFormulaScoring</a></span><span class="op">(</span><span class="va">coms</span>, <span class="va">formulas</span>, updateScore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<p>Here, corresponding formula and explained fragments will be used to calculate a <em>formulaScore</em> for each candidate. Note that SIRIUS candidates are already based on calculated formulae, hence, running this function on SIRIUS results is less sensable unless scoring from another formula calculation algorithm is desired.</p>
</div>
<div id="further-options-and-parameters" class="section level4">
<h4 class="hasAnchor">
<a href="#further-options-and-parameters" class="anchor"></a>Further options and parameters</h4>
<p>There are <em>many</em> more options and parameters that affect compound annotation. For a full overview please have a look at the reference manual (e.g. by running <code><a href="../reference/compound-generation.html">?generateCompounds</a></code>).</p>
</div>
</div>
</div>
<div id="componentization" class="section level2">
<h2 class="hasAnchor">
<a href="#componentization" class="anchor"></a>Componentization</h2>
<p>In <code>patRoon</code> <em>componentization</em> refers to grouping related feature groups together in components. Currently there are three different methodologies to generate components:</p>
<ul>
<li>Similarity on chromatographic elution profiles: feature groups with similar chromatographic behaviour which are assuming to be the same chemical compound (e.g. adducts or isotopologues).</li>
<li>Homologous series: features with increasing <em>m/z</em> and retention time.</li>
<li>Intensity profiles: features that follow a similar intensity profile in the analyses.</li>
</ul>
<p>The following algorithms are currently supported:</p>
<table class="table">
<colgroup>
<col width="17%">
<col width="43%">
<col width="38%">
</colgroup>
<thead><tr class="header">
<th>Algorithm</th>
<th>Usage</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://msbi.ipb-halle.de/msbi/CAMERA/">CAMERA</a></td>
<td><code><a href="../reference/component-generation.html">generateComponents(algorithm = "camera", ...)</a></code></td>
<td>Clusters feature groups with similar chromatographic elution profiles and annotate by known chemical rules (adducts, isotopologues, in-source fragments).</td>
</tr>
<tr class="even">
<td><a href="https://github.com/sneumann/RAMClustR">RAMClustR</a></td>
<td><code><a href="../reference/component-generation.html">generateComponents(algorithm = "ramclustr", ...)</a></code></td>
<td>As above.</td>
</tr>
<tr class="odd">
<td><a href="https://cran.r-project.org/web/packages/nontarget/index.html">nontarget</a></td>
<td><code><a href="../reference/component-generation.html">generateComponents(algorithm = "nontarget", ...)</a></code></td>
<td>Uses the <a href="https://cran.r-project.org/web/packages/nontarget/index.html">nontarget</a> R package to perform unsupervised homologous series detection.</td>
</tr>
<tr class="even">
<td>Intensity clustering</td>
<td><code><a href="../reference/component-generation.html">generateComponents(algorithm = "intclust", ...)</a></code></td>
<td>Groups features with similar intensity profiles across analyses by hierarchical clustering.</td>
</tr>
</tbody>
</table>
<blockquote>
<p><strong><em>NOTE</em></strong> Componentization is a complex process and currently still in a relative young development phase. As such, its functionality and interface are planned to be further improved and results obtained now should always be manually checked, for instance, by using the <a href="#report">reporting functions</a>.</p>
</blockquote>
<div id="features-with-similar-chromatographic-behaviour" class="section level3">
<h3 class="hasAnchor">
<a href="#features-with-similar-chromatographic-behaviour" class="anchor"></a>Features with similar chromatographic behaviour</h3>
<p>Isotopes, adducts and in-source fragments typically result in detection of multiple mass peaks by the mass spectrometer for a single chemical compound. While some feature finding algorithms already try to collapse (some of) these in to a single feature, this process is often incomplete (if performed at all) and it is not uncommon that multiple features will describe the same compound. To overcome this complexity the algorithms from <a href="http://msbi.ipb-halle.de/msbi/CAMERA/">CAMERA</a> and <a href="https://github.com/sneumann/RAMClustR">RAMClustR</a> can be used to group features that undergo highly similar chromatographic behaviour but have different <em>m/z</em> values. Basic chemical rules are then applied to the resulting components to annotate adducts, in-source fragments and isotopologues, which may be highly useful for general identification purposes.</p>
<p>Some common function arguments to <code><a href="../reference/component-generation.html">generateComponents()</a></code> are listed below. Note that careful tuning for some of these is required to obtain useful results. In our experience the current default settings may significantly ‘over-cluster’ features that (clearly visibly) do not belong to each other. For this reason, you are advised to optimize and verify the various parameters supported by both algorithms. For a complete listing all arguments see the reference manual (e.g. <code><a href="../reference/component-generation.html">?generateComponents</a></code>).</p>
<table class="table">
<colgroup>
<col width="23%">
<col width="20%">
<col width="56%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Algorithm</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ionization</code></td>
<td>
<code>"camera"</code>, <code>"ramclustr"</code>
</td>
<td>Ionization mode: <code>"positive"</code> or <code>"negative"</code>
</td>
</tr>
<tr class="even">
<td><code>minSize</code></td>
<td>
<code>"camera"</code>, <code>"ramclustr"</code>
</td>
<td>Minimum component size. Smaller components will be removed.</td>
</tr>
<tr class="odd">
<td><code>relMinReplicates</code></td>
<td>
<code>"camera"</code>, <code>"ramclustr"</code>
</td>
<td>See below.</td>
</tr>
<tr class="even">
<td>
<code>st</code>, <code>sr</code>, <code>maxt</code>, <code>hmax</code>
</td>
<td><code>"ramclustr"</code></td>
<td>Common parameters to influence clustering of <a href="https://github.com/sneumann/RAMClustR">RAMClustR</a>. See <code>?ramclustR</code> for details.</td>
</tr>
<tr class="odd">
<td><code>extraOpts</code></td>
<td><code>"camera"</code></td>
<td>A list with extra argument passed to the <code>annotate()</code> function from <a href="http://msbi.ipb-halle.de/msbi/CAMERA/">CAMERA</a>.</td>
</tr>
<tr class="even">
<td>
<code>extraOptsRC</code>, <code>extraOptsFM</code>
</td>
<td><code>"ramclustr"</code></td>
<td>A list with extra arguments passed to the <code>ramclustR()</code> and <code>do.findmain()</code> functions from <a href="https://github.com/sneumann/RAMClustR">RAMClustR</a>.</td>
</tr>
</tbody>
</table>
<p>Note that both algorithms were primarily designed for datasets where features are generally present in the majority of the analyses (as is relatively common in metabolomics). For environmental analyses, however, this is often not the case. As a result, it may happen that not all features from the feature groups within a component share their presence in the same analyses. In reality, this situation would be fairly unusual, and it is likely that such features actually do not belong to the same component. An extra filter option was added to improve such scenarios: after componentization all features are checked to have a minimal presence across all analyses within the component. This is configured by the <code>relMinReplicates</code> argument of <code><a href="../reference/component-generation.html">generateComponents()</a></code>, which specifies the relative number of replicate groups in which a feature should be present. For instance, when this value is <em>0.5</em> (the default), a feature must be present in at least half of all replicate groups present in the component.</p>
<p>Some example usage is shown below.</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="co"># Use CAMERA with defaults</span>
<span class="va">componCAM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/component-generation.html">generateComponents</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"camera"</span>, ionization <span class="op">=</span> <span class="st">"positive"</span><span class="op">)</span>

<span class="co"># CAMERA with customized settings</span>
<span class="va">componCAM2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/component-generation.html">generateComponents</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"camera"</span>, ionization <span class="op">=</span> <span class="st">"positive"</span>,
                                 extraOpts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mzabs <span class="op">=</span> <span class="fl">0.001</span>, sigma <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Use RAMClustR with customized parameters</span>
<span class="va">componRC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/component-generation.html">generateComponents</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"ramclustr"</span>, ionization <span class="op">=</span> <span class="st">"positive"</span>, hmax <span class="op">=</span> <span class="fl">0.4</span>,
                               extraOptsRC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>cor.method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span>,
                               extraOptsFM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>ppm.error <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="homologues-series" class="section level3">
<h3 class="hasAnchor">
<a href="#homologues-series" class="anchor"></a>Homologues series</h3>
<p>Homologues series can be automatically detected by interfacing with the <a href="https://cran.r-project.org/web/packages/nontarget/index.html">nontarget</a> R package. Components are made from feature groups that show increasing <em>m/z</em> and retention time values. Series are first detected within each replicate group. Afterwards, series from all replicates are linked in case (partial) overlap occurs and this overlap consists of the <em>same</em> feature groups (see figure below). Linked series are then finally merged if this will not cause any conflicts with other series: such a conflict typically occurs when two series are not only linked to each other.</p>
<div class="figure">
<div id="htmlwidget-3d4f4484dc01f144805d" style="width:600px;height:250px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-3d4f4484dc01f144805d">{"x":{"diagram":"\ndigraph Homologs {\n  graph [ rankdir = LR, compound = true, style = invis ]\n  node [ shape = oval,\n         fixedsize = true,\n         width = 2.3,\n         height = 0.8,\n         fontsize = 25,\n         fillcolor = darkseagreen1,\n         style = filled ]\n\n  subgraph cluster3 {\n    I [fillcolor=skyblue]; J [fillcolor=skyblue]; K [fillcolor=skyblue]\n    G -> H -> I -> J -> K [style=invis]\n  }\n\n  subgraph cluster4 {\n    _I [label=I, fillcolor=skyblue]; X [fillcolor=indianred1]; _K [label=K, fillcolor=skyblue]\n    _I -> X -> _K -> L [style=invis]\n  }\n\n  H -> _I [ltail=cluster3, lhead=cluster4, style=invis]\n  \n  I -> _I [constraint=false, style=dashed, arrowhead=none]\n  K -> _K [constraint=false, style=dashed, arrowhead=none]\n  \n  subgraph cluster1 {\n    C [fillcolor=skyblue]; D [fillcolor=skyblue]\n    A -> B -> C -> D [style=invis]\n  }\n\n  subgraph cluster2 {\n    _C [label=C, fillcolor=skyblue]; _D [label=D, fillcolor=skyblue]\n    _C -> _D -> E -> F [style=invis]\n  }\n  \n  B -> _C [ltail=cluster1, lhead=cluster2, style=invis]\n\n  C -> _C [constraint=false, style=dashed, arrowhead=none]\n  D -> _D [constraint=false, style=dashed, arrowhead=none]\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p class="caption">
<strong>Linking of homologues series</strong> top: partial overlap and will be linked; bottom: no linkage due to different feature in overlapping series.
</p>
</div>
<p>The series that are linked can be interactively explored with the <code><a href="../reference/componentsNT-class.html">plotGraph()</a></code> function (discussed <a href="#vis_feat_ann">here</a>).</p>
<p>Common function arguments to <code><a href="../reference/component-generation.html">generateComponents()</a></code> are listed below.</p>
<table class="table">
<colgroup>
<col width="21%">
<col width="78%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ionization</code></td>
<td>Ionization mode: <code>"positive"</code> or <code>"negative"</code>
</td>
</tr>
<tr class="even">
<td>
<code>rtRange</code>, <code>mzRange</code>
</td>
<td>Retention and <em>m/z</em> increment range. Retention times can be negative to allow series with increasing <em>m/z</em> values and decreasing retention times.</td>
</tr>
<tr class="odd">
<td><code>elements</code></td>
<td>Vector with elements to consider.</td>
</tr>
<tr class="even">
<td>
<code>rtDev</code>, <code>absMzDev</code>
</td>
<td>Maximum retention time and <em>m/z</em> deviation.</td>
</tr>
<tr class="odd">
<td><code>extraOpts</code></td>
<td>List with extra arguments passed to the <code>homol.search()</code> function.</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="co"># default settings</span>
<span class="va">componNT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/component-generation.html">generateComponents</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"nontarget"</span>, ionization <span class="op">=</span> <span class="st">"positive"</span><span class="op">)</span>

<span class="co"># customized settings</span>
<span class="va">componNT2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/component-generation.html">generateComponents</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"nontarget"</span>, ionization <span class="op">=</span> <span class="st">"positive"</span>,
                                elements <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"C"</span>, <span class="st">"H"</span><span class="op">)</span>, rtRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">60</span>, <span class="fl">60</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="intclust" class="section level3">
<h3 class="hasAnchor">
<a href="#intclust" class="anchor"></a>Intensity clustering</h3>
<!-- UNDONE: refs? -->
<p>Whereas previous componentization methods utilized chemical properties to relate features, intensity clustering uses a statistical approach. This methodology is especially useful to find features that show similar trends in the analysed samples. Intensities for all features are first normalized and thereafter hierarchical clustering is performed to find features that show similar intensity profiles across analyses. Components are then formed from automatically assigned clusters (using the <a href="https://cran.r-project.org/package=dynamicTreeCut">dynamicTreeCut</a> R package, however, assignment can be changed afterwards).</p>
<p>Some common arguments to <code><a href="../reference/component-generation.html">generateComponents()</a></code> are listed below. It is recommended to test various settings (especially for <code>method</code>) to optimize the clustering results.</p>
<table class="table">
<colgroup>
<col width="36%">
<col width="12%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Default</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>method</code></td>
<td><code>"complete"</code></td>
<td>Clustering method. See <code><a href="https://rdrr.io/r/stats/hclust.html">?hclust</a></code>
</td>
</tr>
<tr class="even">
<td><code>metric</code></td>
<td><code>"euclidean"</code></td>
<td>Metric used to calculate the distance matrix. See <code>?daisy</code>.</td>
</tr>
<tr class="odd">
<td><code>normFunc</code></td>
<td><code>max</code></td>
<td>Function used to normalize data. Feature intensities within a feature group are divided by the result of when this function is called with their intensity values.</td>
</tr>
<tr class="even">
<td><code>average</code></td>
<td><code>TRUE</code></td>
<td>Whether intensities of replicates should first be averaged.</td>
</tr>
<tr class="odd">
<td>
<code>maxTreeHeight</code>, <code>deepSplit</code>, <code>minModuleSize</code>
</td>
<td>
<code>1</code>, <code>TRUE</code>, <code>1</code>
</td>
<td>Used for dynamic cluster assignment. See <code>?cutreeDynamicTree</code>.</td>
</tr>
</tbody>
</table>
<p>The resulting components are stored in an object from the <code>componentsIntClust</code> S4 class. Several methods are defined that can be used on such objects to re-assign clusters, perform plotting operations and so on. Below are some examples. For plotting see the relevant <a href="#plotClust">visualization section</a>. More info can be found in the reference manual (e.g. <code><a href="../reference/componentsIntClust-class.html">?componentsIntClust</a></code>).</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="co"># generate components with default settings</span>
<span class="va">componInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/component-generation.html">generateComponents</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"intclust"</span><span class="op">)</span>

<span class="co"># manually re-assign clusters</span>
<span class="va">componInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">treeCut</a></span><span class="op">(</span><span class="va">componInt</span>, k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>

<span class="co"># automatic re-assignment of clusters (adjusted max tree height)</span>
<span class="va">componInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">treeCutDynamic</a></span><span class="op">(</span><span class="va">componInt</span>, maxTreeHeight <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></pre></div>
</div>
</div>
</div>
<div id="processing-workflow-data" class="section level1">
<h1 class="hasAnchor">
<a href="#processing-workflow-data" class="anchor"></a>Processing workflow data</h1>
<p>The previous chapter mainly discussed how to create workflow data. This chapter will discuss how to <em>use</em> the data.</p>
<div id="inspecting-results" class="section level2">
<h2 class="hasAnchor">
<a href="#inspecting-results" class="anchor"></a>Inspecting results</h2>
<p>Several generic functions exist that can be used to inspect data that is stored in a particular object (e.g. features, compounds etc):</p>
<table class="table">
<colgroup>
<col width="41%">
<col width="23%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th>Generic</th>
<th>Classes</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/generics.html">length()</a></code></td>
<td>All</td>
<td>Returns the length of the object (e.g. number of features, compounds etc)</td>
</tr>
<tr class="even">
<td><code><a href="../reference/generics.html">algorithm()</a></code></td>
<td>All</td>
<td>Returns the name of the algorithm used to generate the object.</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/generics.html">groupNames()</a></code></td>
<td>All</td>
<td>Returns all the unique identitifiers (or names) of the feature groups for which this object contains results.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/generics.html">names()</a></code></td>
<td>
<code>featureGroups</code>, <code>components</code>
</td>
<td>Returns names of the feature groups (similar to <code><a href="../reference/generics.html">groupNames()</a></code>) or components</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/generics.html">show()</a></code></td>
<td>All</td>
<td>Prints general information.</td>
</tr>
<tr class="even">
<td>
<code>"[["</code> / <code>"$"</code> operators</td>
<td>All</td>
<td>Extract general information, see below.</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/generics.html">as.data.table()</a></code> / <code><a href="../reference/generics.html">as.data.frame()</a></code>
</td>
<td>All</td>
<td>Convert data to a <code>data.table</code> or <code>data.frame</code>, see below.</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/generics.html">analysisInfo()</a></code>, <code><a href="../reference/generics.html">analyses()</a></code>, <code><a href="../reference/generics.html">replicateGroups()</a></code>
</td>
<td>
<code>features</code>, <code>featureGroups</code>
</td>
<td>Returns the <a href="#anaInfo">analysis information</a>, analyses or replicate groups for which this object contains data.</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/featureGroups-class.html">groupInfo()</a></code></td>
<td><code>featureGroups</code></td>
<td>Returns feature group information (<em>m/z</em> and retention time values).</td>
</tr>
<tr class="even">
<td><code><a href="../reference/components-class.html">componentInfo()</a></code></td>
<td><code>components</code></td>
<td>Returns information for all components.</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/generics.html">annotatedPeakList()</a></code></td>
<td>
<code>formulas</code>, <code>compounds</code>
</td>
<td>Returns a table with annotated mass peaks (see below).</td>
</tr>
</tbody>
</table>
<p>The common <code>R</code> extraction operators <code>"[["</code>, <code>"$"</code> can be used to obtain data for a particular feature groups, analysis etc:</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="co"># Feature table (only first columns for readability)</span>
<span class="va">fList</span><span class="op">[[</span><span class="st">"standard-1"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>

<span class="co"># Feature group intensities</span>
<span class="va">fGroups</span><span class="op">$</span><span class="va">M120_R328_81</span>
<span class="va">fGroups</span><span class="op">[[</span><span class="fl">1</span>, <span class="st">"M120_R328_81"</span><span class="op">]</span><span class="op">]</span> <span class="co"># only first analysis</span>

<span class="co"># obtains list MS/MS peak list  (feature group averaged datas)</span>
<span class="va">mslists</span><span class="op">[[</span><span class="st">"M120_R328_81"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">MSMS</span>

<span class="co"># get all formula candidates for a feature group</span>
<span class="va">formulas</span><span class="op">[[</span><span class="st">"M120_R328_81"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span>

<span class="co"># get all compound candidates for a feature group</span>
<span class="va">compounds</span><span class="op">[[</span><span class="st">"M120_R328_81"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>

<span class="co"># get a table with information of a component</span>
<span class="va">components</span><span class="op">[[</span><span class="st">"CMP7"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></pre></div>
<p>A more sophisticated way to obtain data from a workflow object is to use <code><a href="../reference/generics.html">as.data.table()</a></code> or <code><a href="../reference/generics.html">as.data.frame()</a></code>. These functions will convert <em>all</em> information within the object to a table (<code>data.table</code> or <code>data.frame</code>) and allow various options to add extra information. An advantage is that this common data format can be used with many other functions within <code>R</code>. The output is in a <a href="https://r4ds.had.co.nz/tidy-data.html">tidy format</a>.</p>
<blockquote>
<p><strong><em>NOTE</em></strong> If you are not familiar with <code>data.table</code> and want to know more see <a href="https://github.com/Rdatatable/data.table/wiki">data.table</a>. Briefly, this is a more efficient and largely compatible alternative to the regular <code>data.frame</code>.</p>
</blockquote>
<blockquote>
<p><strong><em>NOTE</em></strong> The <code><a href="../reference/generics.html">as.data.frame()</a></code> methods defined in <code>patRoon</code> simply convert the results from <code><a href="../reference/generics.html">as.data.table()</a></code>, hence, both functions are equal in their usage and are defined for the same object classes.</p>
</blockquote>
<p>Some typical examples are shown below.</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="co"># obtain table with all features (only first columns for readability)</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">fList</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>

<span class="co"># Returns group info and intensity values for each feature group</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">fGroups</span>, average <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># average intensities for replicates</span>

<span class="co"># Returns all peak lists for each feature group</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">mslists</span><span class="op">)</span>

<span class="co"># Returns all formula candidates for each feature group with scoring</span>
<span class="co"># information, neutral loss etc</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">formulas</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>

<span class="co"># Returns all compound candidates for each feature group with scoring and other metadata</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">compounds</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>

<span class="co"># Returns table with all components (including feature group info, annotations etc)</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">components</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></pre></div>
<p>Finally, the <code><a href="../reference/generics.html">annotatedPeakList()</a></code> function is useful to inspect annotation results for a formula or compound candidate:</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="co"># formula annotations for for a formula candidate of feature group M120_R328_81</span>
<span class="fu"><a href="../reference/generics.html">annotatedPeakList</a></span><span class="op">(</span><span class="va">formulas</span>, precursor <span class="op">=</span> <span class="st">"C6H6N3"</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span>,
                  MSPeakLists <span class="op">=</span> <span class="va">mslists</span><span class="op">)</span>

<span class="co"># compound annotation for first candidate of feature group M120_R328_81</span>
<span class="fu"><a href="../reference/generics.html">annotatedPeakList</a></span><span class="op">(</span><span class="va">compounds</span>, index <span class="op">=</span> <span class="fl">1</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span>,
                  MSPeakLists <span class="op">=</span> <span class="va">mslists</span><span class="op">)</span></pre></div>
<p>More advanced examples for these functions are shown below.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="co"># Feature table, can also be accessed by numeric index</span>
<span class="va">fList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">mslists</span><span class="op">[[</span><span class="st">"standard-1"</span>, <span class="st">"M120_R328_81"</span><span class="op">]</span><span class="op">]</span> <span class="co"># feature data (instead of feature group averaged)</span>
<span class="va">formulas</span><span class="op">[[</span><span class="fl">1</span>, <span class="st">"M120_R328_81"</span><span class="op">]</span><span class="op">]</span> <span class="co"># feature data (if available, i.e. calculateFeatures=TRUE)</span>
<span class="va">components</span><span class="op">[[</span><span class="st">"CMP1"</span>, <span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co"># only for first feature group in component</span>

<span class="fu"><a href="../reference/generics.html">as.data.frame</a></span><span class="op">(</span><span class="va">fList</span><span class="op">)</span> <span class="co"># classic data.frame format, works for all objects</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span> <span class="co"># return non-averaged intensities (default)</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">fGroups</span>, features <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># include feature information</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">mslists</span>, averaged <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># peak lists each feature</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">mslists</span>, fGroups <span class="op">=</span> <span class="va">fGroups</span><span class="op">)</span> <span class="co"># add feature group information</span>

<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">formulas</span>, countElements <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"C"</span>, <span class="st">"H"</span><span class="op">)</span><span class="op">)</span> <span class="co"># include C/H counts (e.g. for van Krevelen plots)</span>
<span class="co"># report only top precursor and fragment formula. This yields in one row per feature group.</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">formulas</span>, maxFormulas <span class="op">=</span> <span class="fl">1</span>, maxFragFormulas <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co"># add various information for organic matter characterization (common elemental</span>
<span class="co"># counts/ratios, classifications etc)</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">formulas</span>, OM <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">compounds</span>, fGroups <span class="op">=</span> <span class="va">fGroups</span><span class="op">)</span> <span class="co"># add feature group informaion</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">compounds</span>, fragments <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># include information of all annotated fragments</span>

<span class="fu"><a href="../reference/generics.html">annotatedPeakList</a></span><span class="op">(</span><span class="va">formulas</span>, precursor <span class="op">=</span> <span class="st">"C6H6N3"</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span>,
                  MSPeakLists <span class="op">=</span> <span class="va">mslists</span>, onlyAnnotated <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># only include annotated peaks</span>
<span class="fu"><a href="../reference/generics.html">annotatedPeakList</a></span><span class="op">(</span><span class="va">compounds</span>, index <span class="op">=</span> <span class="fl">1</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span>,
                  MSPeakLists <span class="op">=</span> <span class="va">mslists</span>, formulas <span class="op">=</span> <span class="va">formulas</span><span class="op">)</span> <span class="co"># include formula annotations</span></pre></div>
</div>
<div id="filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering" class="anchor"></a>Filtering</h2>
<p>During a non-target workflow it is not uncommon that some kind of data-cleanup is necessary. Datasets are often highly complex, which makes separating data of interest from the rest highly important. Furthermore, general cleanup typically improves the quality of the dataset, for instance by removing low scoring annotation results or features that are unlikely to be ‘correct’ (e.g. noise or present in blanks). For this reason <code>patRoon</code> supports <em>many</em> different filters that easily clean data produced during the workflow in a highly customizable way.</p>
<p>All major workflow objects (e.g. <code>featureGroups</code>, <code>compounds</code>, <code>components</code> etc.) support filtering operations by the <code><a href="../reference/generics.html">filter()</a></code> generic. This function takes the object to be filtered as first argument and any remaining arguments describe the desired filter options. The <code><a href="../reference/generics.html">filter()</a></code> generic function then returns the modified object back. Some examples are shown below.</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="co"># remove low intensity (&lt;500) features</span>
<span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">features</span>, absMinIntensity <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>

<span class="co"># remove features with intensities lower than 5 times the blank</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fGroups</span>, blankThreshold <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="co"># only retain compounds with &gt;1 explained MS/MS peaks</span>
<span class="va">compounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">compounds</span>, minExplainedPeaks <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></pre></div>
<p>The following sections will provide a more detailed overview of available data filters.</p>
<blockquote>
<p><strong><em>NOTE</em></strong> Some other <code>R</code> packages (notably <code>dplyr</code>) also provide a <code><a href="../reference/generics.html">filter()</a></code> generic function. To use the <code><a href="../reference/generics.html">filter()</a></code> function from different packages you need explicitly specify which one to use in your script. This can be done by prefixing it with the package name, e.g. <code><a href="../reference/generics.html">patRoon::filter(...)</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr::filter(...)</a></code> etc.</p>
</blockquote>
<div id="features-1" class="section level3">
<h3 class="hasAnchor">
<a href="#features-1" class="anchor"></a>Features</h3>
<p>There are many filters available for feature data:</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="21%">
<col width="45%">
</colgroup>
<thead><tr class="header">
<th>Filter</th>
<th>Classes</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>absMinIntensity</code>, <code>relMinIntensity</code>
</td>
<td>
<code>features</code>, <code>featureGroups</code>
</td>
<td>Minimum intensity</td>
</tr>
<tr class="even">
<td>
<code>preAbsMinIntensity</code>, <code>preRelMinIntensity</code>
</td>
<td><code>featureGroups</code></td>
<td>Minimum intensity prior to other filtering (see below)</td>
</tr>
<tr class="odd">
<td>
<code>retentionRange</code>, <code>mzRange</code>, <code>mzDefectRange</code>, <code>chromWidthRange</code>
</td>
<td>
<code>features</code>, <code>featureGroups</code>
</td>
<td>Filter by feature properties</td>
</tr>
<tr class="even">
<td>
<code>absMinAnalyses</code>, <code>relMinAnalyses</code>
</td>
<td><code>featureGroups</code></td>
<td>Minimum feature abundance in all analyses</td>
</tr>
<tr class="odd">
<td>
<code>absMinReplicates</code>, <code>relMinReplicates</code>
</td>
<td><code>featureGroups</code></td>
<td>Minimum feature abundance in different replicates</td>
</tr>
<tr class="even">
<td>
<code>absMinFeatures</code>, <code>relMinFeatures</code>
</td>
<td><code>featureGroups</code></td>
<td>Only keep analyses with at least this amount of features</td>
</tr>
<tr class="odd">
<td>
<code>absMinReplicateAbundance</code>, <code>relMinReplicateAbundance</code>
</td>
<td><code>featureGroups</code></td>
<td>Minimum feature abundance in a replicate group</td>
</tr>
<tr class="even">
<td><code>maxReplicateIntRSD</code></td>
<td><code>featureGroups</code></td>
<td>Maximum relative standard deviation of feature intensities in a replicate group.</td>
</tr>
<tr class="odd">
<td><code>blankThreshold</code></td>
<td><code>featureGroups</code></td>
<td>Minimum intensity factor above blank intensity</td>
</tr>
<tr class="even">
<td><code>rGroups</code></td>
<td><code>featureGroups</code></td>
<td>Only keep (features of) these replicate groups</td>
</tr>
</tbody>
</table>
<p>Application of filters to feature data is important for (environmental) non-target analysis. Especially blank and replicate filters (i.e. <code>blankThreshold</code> and <code>absMinReplicateAbundance</code>/<code>relMinReplicateAbundance</code>) are important filters and are highly recommended to always apply for cleaning up your dataset.</p>
<p>All filters are available for feature group data, whereas only a subset is available for feature objects. The main reason is that other filters need grouping of features between analyses. Regardless, in <code>patRoon</code> filtering feature data is less important, and typically only needed when the number of features are extremely large and direct grouping is undesired.</p>
<p>From the table above you can notice that many filters concern both <em>absolute</em> and <em>relative</em> data (i.e. as prefixed with <code>abs</code> and <code>rel</code>). When a relative filter is used the value is scaled between <em>0</em> and <em>1</em>. For instance:</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="co"># remove features not present in at least half of the analyses within a replicate group</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fGroups</span>, relMinReplicateAbundance <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></pre></div>
<p>An advantage of relative filters is that you will not have to worry about the data size involved. For instance, in the above example the filter always takes half of the number of analyses within a replicate group, even when replicate groups have different number of analyses.</p>
<p>Note that multiple filters can be specified at once. Especially for feature group data the order of filtering may impact the final results, this is explained further in the reference manual (i.e. <code><a href="../reference/feature-filtering.html">?`feature-filtering`</a></code>).</p>
<p>Some examples are shown below.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="co"># filter features prior to grouping: remove any features eluting before first 2 minutes</span>
<span class="va">fList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fList</span>, retentionRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">120</span>, <span class="cn">Inf</span><span class="op">)</span><span class="op">)</span>

<span class="co"># common filters for feature groups</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fGroups</span>,
                  absMinIntensity <span class="op">=</span> <span class="fl">500</span>, <span class="co"># remove features &lt;500 intensity</span>
                  relMinReplicateAbundance <span class="op">=</span> <span class="fl">1</span>, <span class="co"># features should be in all analysis of replicate groups</span>
                  maxReplicateIntRSD <span class="op">=</span> <span class="fl">0.75</span>, <span class="co"># remove features with intensity RSD in replicates &gt;75%</span>
                  blankThreshold <span class="op">=</span> <span class="fl">5</span>, <span class="co"># remove features &lt;5x intensity of (average) blank intensity</span>
                  removeBlanks <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># remove blank analyses from object afterwards</span>

<span class="co"># filter by feature properties</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span>mzDefectRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.9</span><span class="op">)</span>,
                  chromWidthRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">120</span><span class="op">)</span><span class="op">)</span>

<span class="co"># remove features not present in at least 3 analyses</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fGroups</span>, absMinAnalyses <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>

<span class="co"># remove features not present in at least 20% of all replicate groups</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fGroups</span>, relMinReplicates <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>

<span class="co"># only keep data present in replicate groups "repl1" and "repl2"</span>
<span class="co"># all other features and analyses will be removed</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fGroups</span>, rGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"repl1"</span>, <span class="st">"repl2"</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="annotation-1" class="section level3">
<h3 class="hasAnchor">
<a href="#annotation-1" class="anchor"></a>Annotation</h3>
<p>There are various filters available for handling annotation data:</p>
<table class="table">
<colgroup>
<col width="36%">
<col width="24%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th>Filter</th>
<th>Classes</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>absMSIntThr</code>, <code>absMSMSIntThr</code>, <code>relMSIntThr</code>, <code>relMSMSIntThr</code>
</td>
<td><code>MSPeakLists</code></td>
<td>Minimum intensity of mass peaks</td>
</tr>
<tr class="even">
<td>
<code>topMSPeaks</code>, <code>topMSMSPeaks</code>
</td>
<td><code>MSPeakLists</code></td>
<td>Only keep most intense mass peaks</td>
</tr>
<tr class="odd">
<td><code>withMSMS</code></td>
<td><code>MSPeakLists</code></td>
<td>Only keep results with MS/MS data</td>
</tr>
<tr class="even">
<td><code>minExplainedPeaks</code></td>
<td>
<code>formulas</code> / <code>compounds</code>
</td>
<td>Minimum number of annotated mass peaks</td>
</tr>
<tr class="odd">
<td>
<code>elements</code>, <code>fragElements</code>, <code>lossElements</code>
</td>
<td>
<code>formulas</code>, <code>compounds</code>
</td>
<td>Restrain elemental composition</td>
</tr>
<tr class="even">
<td><code>topMost</code></td>
<td>
<code>formulas</code>, <code>compounds</code>
</td>
<td>Only keep highest ranked candidates</td>
</tr>
<tr class="odd">
<td>
<code>minScore</code>, <code>minFragScore</code>, <code>minFormulaScore</code>
</td>
<td><code>compounds</code></td>
<td>Minimum compound scorings</td>
</tr>
<tr class="even">
<td><code>scoreLimits</code></td>
<td>
<code>formulas</code>, <code>compounds</code>
</td>
<td>Minimum/Maximum scorings</td>
</tr>
<tr class="odd">
<td><code>OM</code></td>
<td><code>formulas</code></td>
<td>Only keep candidates with likely elemental composition found in organic matter</td>
</tr>
</tbody>
</table>
<p>Several intensity related filters are available to clean-up MS peak list data. For instance, the <code>topMSPeaks</code>/<code>topMSMSPeaks</code> filters provide a simple way to remove noisy data by only retaining a defined number of most intense mass peaks. Note that none of these filters will remove the mass peak of the feature from its MS peak list.</p>
<p>The filters applicable to formula and compound annotation generally concern minimal scoring or chemical properties. The former is useful to remove unlikely candidates, whereas the second is useful to focus on certain study specific chemical properties (e.g. known neutral losses).</p>
<p>Common examples are shown below.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="co"># intensity filtering</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>mslists &lt;-<span class="st"> </span><span class="kw">filter</span>(mslists,</span>
<span id="cb32-3"><a href="#cb32-3"></a>                  <span class="dt">absMSIntThr =</span> <span class="dv">500</span>, <span class="co"># minimum MS mass peak intensity of 500</span></span>
<span id="cb32-4"><a href="#cb32-4"></a>                  <span class="dt">relMSMSIntThr =</span> <span class="fl">0.1</span>) <span class="co"># minimum MS/MS mass peak intensity of 10%</span></span>
<span id="cb32-5"><a href="#cb32-5"></a></span>
<span id="cb32-6"><a href="#cb32-6"></a><span class="co"># only retain 10 most intens mass peaks</span></span>
<span id="cb32-7"><a href="#cb32-7"></a><span class="co"># (feature mass is always retained)</span></span>
<span id="cb32-8"><a href="#cb32-8"></a>mslists &lt;-<span class="st"> </span><span class="kw">filter</span>(mslists, <span class="dt">topMSPeaks =</span> <span class="dv">10</span>)</span>
<span id="cb32-9"><a href="#cb32-9"></a></span>
<span id="cb32-10"><a href="#cb32-10"></a><span class="co"># only keep formulae with 1-10 sulphur or phosphorus elements</span></span>
<span id="cb32-11"><a href="#cb32-11"></a>formulas &lt;-<span class="st"> </span><span class="kw">filter</span>(formulas, <span class="dt">elements =</span> <span class="kw">c</span>(<span class="st">"S1-10"</span>, <span class="st">"P1-10"</span>))</span>
<span id="cb32-12"><a href="#cb32-12"></a></span>
<span id="cb32-13"><a href="#cb32-13"></a><span class="co"># only keep candidates with MS/MS fragments that contain 1-10 carbons and 0-2 oxygens</span></span>
<span id="cb32-14"><a href="#cb32-14"></a>formulas &lt;-<span class="st"> </span><span class="kw">filter</span>(formulas, <span class="dt">fragElements =</span> <span class="st">"C1-10O0-2"</span>)</span>
<span id="cb32-15"><a href="#cb32-15"></a></span>
<span id="cb32-16"><a href="#cb32-16"></a><span class="co"># only keep candidates with CO2 neutral loss</span></span>
<span id="cb32-17"><a href="#cb32-17"></a>formulas &lt;-<span class="st"> </span><span class="kw">filter</span>(formulas, <span class="dt">lossElements =</span> <span class="st">"CO2"</span>)</span>
<span id="cb32-18"><a href="#cb32-18"></a></span>
<span id="cb32-19"><a href="#cb32-19"></a><span class="co"># only keep the 15 highest ranked candidates with at least 1 annotated MS/MS peak</span></span>
<span id="cb32-20"><a href="#cb32-20"></a>compounds &lt;-<span class="st"> </span><span class="kw">filter</span>(compounds, <span class="dt">minExplainedPeaks =</span> <span class="dv">1</span>, <span class="dt">topMost =</span> <span class="dv">15</span>)</span>
<span id="cb32-21"><a href="#cb32-21"></a></span>
<span id="cb32-22"><a href="#cb32-22"></a><span class="co"># minimum in-silico score</span></span>
<span id="cb32-23"><a href="#cb32-23"></a>compounds &lt;-<span class="st"> </span><span class="kw">filter</span>(compounds, <span class="dt">minFragScore =</span> <span class="dv">10</span>)</span>
<span id="cb32-24"><a href="#cb32-24"></a></span>
<span id="cb32-25"><a href="#cb32-25"></a><span class="co"># candidate should be referenced in at least 1 patent</span></span>
<span id="cb32-26"><a href="#cb32-26"></a><span class="co"># (only works if database lists number of patents, e.g. PubChem)</span></span>
<span id="cb32-27"><a href="#cb32-27"></a>compounds &lt;-<span class="st"> </span><span class="kw">filter</span>(compounds,</span>
<span id="cb32-28"><a href="#cb32-28"></a>                    <span class="dt">scoreLimits =</span> <span class="kw">list</span>(<span class="dt">numberPatents =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="ot">Inf</span>))</span></code></pre></div>
</div>
<div id="components" class="section level3">
<h3 class="hasAnchor">
<a href="#components" class="anchor"></a>Components</h3>
<p>Finally several filters are available for components:</p>
<table class="table">
<thead><tr class="header">
<th>Filter</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>size</code></td>
<td>Minimum component size</td>
</tr>
<tr class="even">
<td>
<code>adducts</code>, <code>isotopes</code>
</td>
<td>Filter features by adduct/istopes annotation</td>
</tr>
<tr class="odd">
<td>
<code>rtIncrement</code>, <code>mzIncrement</code>
</td>
<td>Filter homologs by retention/mz increment range</td>
</tr>
</tbody>
</table>
<p>Note that these filters are only applied if the components contain the data the filter works on. For instance, filtering by adducts will <em>not</em> affect components obtained from homologous series.</p>
<p>As before, some typical examples are shown below.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="co"># only keep components with at least 4 features</span>
<span class="va">componInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">componInt</span>, minSize <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>

<span class="co"># remove all features from components are not annotated as an adduct</span>
<span class="va">componRC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">componRC</span>, adducts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># only keep protonated and sodium adducts</span>
<span class="va">componRC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">componRC</span>, adducts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"[M+H]+"</span>, <span class="st">"[M+Na]+"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># remove all features not recognized as isotopes</span>
<span class="va">componRC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">componRC</span>, isotopes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># only keep monoisotopic mass</span>
<span class="va">componRC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">componRC</span>, isotopes <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>

<span class="co"># min/max rt/mz increments for homologs</span>
<span class="va">componNT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">componNT</span>, rtIncrement <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">30</span><span class="op">)</span>,
                   mzIncrement <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></pre></div>
<blockquote>
<p><strong><em>NOTE</em></strong> As mentioned before, components are still in a relative young development phase and results should always be verified!</p>
</blockquote>
</div>
<div id="negation" class="section level3">
<h3 class="hasAnchor">
<a href="#negation" class="anchor"></a>Negation</h3>
<p>All filters support <em>negation</em>: if enabled all specified filters will be executed in an opposite manner. Negation may not be so commonly used, but allows greater flexibility which is sometimes needed for advanced filtering steps. Furthermore, it is also useful to specifically isolate the data that otherwise would have been removed. Some examples are shown below.</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="co"># keep all features/analyses _not_ present from replicate groups "repl1" and "repl2"</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">fGroups</span>, rGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"repl1"</span>, <span class="st">"repl2"</span><span class="op">)</span>, negate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># only retain features with a mass defect outside 0.8-0.9</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span>mzDefectRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.9</span><span class="op">)</span>, negate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># remove candidates with CO2 neutral loss</span>
<span class="va">formulas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">formulas</span>, lossElements <span class="op">=</span> <span class="st">"CO2"</span>, negate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># select 15 worst ranked candidates</span>
<span class="va">compounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">compounds</span>, topMost <span class="op">=</span> <span class="fl">15</span>, negate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># only keep components with &lt;5 features</span>
<span class="va">componInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">filter</a></span><span class="op">(</span><span class="va">componInt</span>, minSize <span class="op">=</span> <span class="fl">5</span>, negate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
</div>
</div>
<div id="subset" class="section level2">
<h2 class="hasAnchor">
<a href="#subset" class="anchor"></a>Subsetting</h2>
<p>The previous section discussed the <code><a href="../reference/generics.html">filter()</a></code> generic function to perform various data cleaning operations. A more generic way to select data is by <em>subsetting</em>: here you can manually specify which parts of an object should be retained. Subsetting is supported for all workflow objects and is performed by the R subset operator (<code>"["</code>). This operator either subsets by one or two arguments, which are referred to as the <code>i</code> and <code>j</code> arguments.</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="15%">
<col width="15%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Class</th>
<th>Argument <code>i</code>
</th>
<th>Argument <code>j</code>
</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>features</code></td>
<td>analyses</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><code>featureGroups</code></td>
<td>analyses</td>
<td>feature groups</td>
<td></td>
</tr>
<tr class="odd">
<td><code>MSPeakLists</code></td>
<td>analyses</td>
<td>feature groups</td>
<td>peak lists for feature groups will be re-averaged when subset on analyses (by default)</td>
</tr>
<tr class="even">
<td><code>formulas</code></td>
<td>feature groups</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><code>compounds</code></td>
<td>feature groups</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><code>components</code></td>
<td>components</td>
<td>feature groups</td>
<td></td>
</tr>
</tbody>
</table>
<p>For objects that support two-dimensional subsetting (e.g. <code>featureGroups</code>, <code>MSPeakLists</code>), either the <code>i</code> or <code>j</code> argument is optional. Furthermore, unlike subsetting a <code>data.frame</code>, the position of <code>i</code> and <code>j</code> does not change when only one argument is specified:</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="va">df</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span> <span class="co"># subset data.frame by first row/column</span>
<span class="va">df</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="co"># subset by first column</span>
<span class="va">df</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span> <span class="co"># subset by first row</span>

<span class="va">fGroups</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span> <span class="co"># subset by first analysis/feature group</span>
<span class="va">fGroups</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="co"># subset by first feature group (i.e. column)</span>
<span class="va">fGroups</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="co"># subset by first analysis (i.e. row)</span></pre></div>
<p>The subset operator allows three types of input:</p>
<ul>
<li>A logical vector: elements are selected if corresponding values are <code>TRUE</code>.</li>
<li>A numeric vector: select elements by numeric index.</li>
<li>A character vector: select elements by their name.</li>
</ul>
<p>When a logical vector is used as input it will be re-cycled if necessary. For instance, the following will select by the first, third, fifth, etc. analysis.</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="va">fGroups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span></pre></div>
<p>In order to select by a <code>character</code> you will need to know the names for each element. These can, for instance, be obtained by the <code><a href="../reference/generics.html">groupNames()</a></code> (feature group names), <code><a href="../reference/generics.html">analyses()</a></code> (analysis names) and <code><a href="../reference/generics.html">names()</a></code> (names for components or feature groups for <code>featureGroups</code> objects) generic functions.</p>
<p>Some more examples of common subsetting operations are shown below.</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="co"># select first three analyses</span>
<span class="va">fList</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>

<span class="co"># select first three analyses and first 500 feature groups</span>
<span class="va">fGroups</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>

<span class="co"># select all feature groups from first component</span>
<span class="va">fGroupsNT</span> <span class="op">&lt;-</span> <span class="va">fGroups</span><span class="op">[</span>, <span class="va">componNT</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">group</span><span class="op">]</span>

<span class="co"># only keep feature groups with formula annotation results</span>
<span class="va">fGroupsForms</span> <span class="op">&lt;-</span> <span class="va">fGroups</span><span class="op">[</span>, <span class="fu"><a href="../reference/generics.html">groupNames</a></span><span class="op">(</span><span class="va">formulas</span><span class="op">)</span><span class="op">]</span>

<span class="co"># only keep feature groups with either formula or compound annotation results</span>
<span class="va">fGroupsAnn</span> <span class="op">&lt;-</span> <span class="va">fGroups</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/sets.html">union</a></span><span class="op">(</span><span class="fu"><a href="../reference/generics.html">groupNames</a></span><span class="op">(</span><span class="va">formulas</span><span class="op">)</span>, <span class="fu"><a href="../reference/generics.html">groupNames</a></span><span class="op">(</span><span class="va">compounds</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>

<span class="co"># select first 15 components</span>
<span class="va">components</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>

<span class="co"># select by name</span>
<span class="va">components</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CMP1"</span>, <span class="st">"CMP5"</span><span class="op">)</span><span class="op">]</span>

<span class="co"># only retain feature groups in components for which compound annotations are</span>
<span class="co"># available</span>
<span class="va">components</span><span class="op">[</span>, <span class="fu"><a href="../reference/generics.html">groupNames</a></span><span class="op">(</span><span class="va">compounds</span><span class="op">)</span><span class="op">]</span></pre></div>
<div id="prioritization-workflow" class="section level3">
<h3 class="hasAnchor">
<a href="#prioritization-workflow" class="anchor"></a>Prioritization workflow</h3>
<p>An important use case of subsetting is prioritization of data. For instance, after statistical analysis only certain feature groups are deemed relevant for the rest of the workflow. A common prioritization workflow is illustrated below:</p>
<div id="htmlwidget-383ddbe17d205db7a9c1" style="width:500px;height:120px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-383ddbe17d205db7a9c1">{"x":{"diagram":"\ndigraph Prioritization {\n  graph [ rankdir = LR ]\n  node [ shape = box,\n         fixedsize = true,\n         width = 2.3,\n         height = 1,\n         fontsize = 18,\n         fillcolor = darkseagreen1,\n         style = filled ]\n\n    \"Object conversion\" -> \"Prioritization\" -> Subsetting\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p>During the first step the workflow object is converted to a suitable format, most often using the <code><a href="../reference/generics.html">as.data.frame()</a></code> function. The converted data is then used as input for the prioritization strategy. Finally, these results are then used to select the data of interest in the original object.</p>
<p>A very simplified example of such a process is shown below.</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="va">featTab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">as.data.frame</a></span><span class="op">(</span><span class="va">fGroups</span>, average <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># prioritization: sort by (averaged) intensity of the "sample" replicate group</span>
<span class="co"># (from high to low) and then obtain the feature group identifiers of the top 5.</span>
<span class="va">featTab</span> <span class="op">&lt;-</span> <span class="va">featTab</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">featTab</span><span class="op">$</span><span class="va">standard</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">groupsOfInterest</span> <span class="op">&lt;-</span> <span class="va">featTab</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>

<span class="co"># subset the original data</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="va">fGroups</span><span class="op">[</span>, <span class="va">groupsOfInterest</span><span class="op">]</span>

<span class="co"># fGroups now only contains the feature groups for which intensity values in the</span>
<span class="co"># "sample" replicate group were in the top 5</span></pre></div>
</div>
</div>
<div id="unOv" class="section level2">
<h2 class="hasAnchor">
<a href="#unOv" class="anchor"></a>Unique and overlapping features</h2>
<p>Often an analysis batch is composed of different sample groups, such as different treatments, influent/effluent etc. In such scenarios it may be highly interesting to evaluate uniqueness or overlap between these samples. Furthermore, extracting overlapping or unique features is a simple but effective prioritization strategy.</p>
<p>The <code><a href="../reference/featureGroups-class.html">overlap()</a></code> and <code><a href="https://rdrr.io/r/base/unique.html">unique()</a></code> functions can be used to extract overlapping and unique features between replicate groups, respectively. Both functions return a subset of the given <code>featureGroups</code> object. An overview of their arguments is given below.</p>
<table class="table">
<colgroup>
<col width="12%">
<col width="23%">
<col width="64%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Function(s)</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>which</code></td>
<td>
<code><a href="https://rdrr.io/r/base/unique.html">unique()</a></code>, <code><a href="../reference/featureGroups-class.html">overlap()</a></code>
</td>
<td>The replicate groups to compare.</td>
</tr>
<tr class="even">
<td><code>relativeTo</code></td>
<td><code><a href="https://rdrr.io/r/base/unique.html">unique()</a></code></td>
<td>Only return unique features compared to these replicate groups (<code>NULL</code> for all). Replicate groups in <code>which</code> are ignored.</td>
</tr>
<tr class="odd">
<td><code>outer</code></td>
<td><code><a href="https://rdrr.io/r/base/unique.html">unique()</a></code></td>
<td>If <code>TRUE</code> then only return features which are <em>also</em> unique among the compared replicates groups.</td>
</tr>
<tr class="even">
<td><code>exclusive</code></td>
<td><code>overlap</code></td>
<td>Only keep features that <em>only</em> overlap between the compared replicate groups.</td>
</tr>
</tbody>
</table>
<p>Some examples:</p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="co"># only keep features uniquely present in replicate group "repl1"</span>
<span class="va">fGroupsUn1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">fGroups</span>, which <span class="op">=</span> <span class="st">"repl1"</span><span class="op">)</span>
<span class="co"># only keep features in repl1/repl2 which are not in repl3</span>
<span class="va">fGroupsUn2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">fGroups</span>, which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"repl1"</span>, <span class="st">"repl2"</span><span class="op">)</span>,
                     relativeTo <span class="op">=</span> <span class="st">"repl3"</span><span class="op">)</span>
<span class="co"># only keep features that are only present in repl1 OR repl2</span>
<span class="va">fGroupsUn3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">fGroups</span>, which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"repl1"</span>, <span class="st">"repl2"</span><span class="op">)</span>,
                     outer <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># only keep features overlapping in repl1/repl2</span>
<span class="va">fGroupsOv1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/featureGroups-class.html">overlap</a></span><span class="op">(</span><span class="va">fGroups</span>, which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"repl1"</span>, <span class="st">"repl2"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># only keep features overlapping in repl1/repl2 AND are not present in any other</span>
<span class="co"># replicate group</span>
<span class="va">fGroupsOv2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/featureGroups-class.html">overlap</a></span><span class="op">(</span><span class="va">fGroups</span>, which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"repl1"</span>, <span class="st">"repl2"</span><span class="op">)</span>,
                      exclusive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<p>In addition, several plotting functions are <a href="#visComp">discussed in the next section</a> that visualize overlap and uniqueness of features.</p>
</div>
<div id="visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h2>
<div id="vis_feat_ann" class="section level3">
<h3 class="hasAnchor">
<a href="#vis_feat_ann" class="anchor"></a>Features and annatation data</h3>
<p>Several generic functions are available to visualize feature and annotation data:</p>
<table class="table">
<colgroup>
<col width="12%">
<col width="39%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>Generic</th>
<th>Classes</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/generics.html">plot()</a></code></td>
<td>
<code>featureGroups</code>, <code>featureGroupsComparison</code>
</td>
<td>Scatter plot for retention and <em>m/z</em> values</td>
</tr>
<tr class="even">
<td><code><a href="../reference/generics.html">plotInt()</a></code></td>
<td><code>featureGroups</code></td>
<td>Intensity profiles across analyses</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>
<code>featureGroups</code>, <code>components</code>
</td>
<td>Plot extracted ion chromatograms (EICs)</td>
</tr>
<tr class="even">
<td><code><a href="../reference/generics.html">plotSpec()</a></code></td>
<td>
<code>MSPeakLists</code>, <code>formulas</code>, <code>compounds</code>, <code>components</code>
</td>
<td>Plots (annotated) spectra</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/generics.html">plotStructure()</a></code></td>
<td><code>compounds</code></td>
<td>Draws candidate structures</td>
</tr>
<tr class="even">
<td><code><a href="../reference/generics.html">plotScores()</a></code></td>
<td>
<code>formulas</code>, <code>compounds</code>
</td>
<td>Barplot for candidate scoring</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/componentsNT-class.html">plotGraph()</a></code></td>
<td><code>componentsNT</code></td>
<td>Draws interactive graphs of linked homologous series</td>
</tr>
</tbody>
</table>
<p>The most common plotting functions are <code><a href="../reference/generics.html">plotEIC()</a></code>, which plots chromatographic data for features, and <code><a href="../reference/generics.html">plotSpec()</a></code>, which will plot (annotated) spectra. An overview of their most important function arguments are shown below.</p>
<table class="table">
<colgroup>
<col width="32%">
<col width="18%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Generic</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>rtWindow</code></td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>Extra time (in s) +/- retention limits of plotted features (useful to zoom out)</td>
</tr>
<tr class="even">
<td><code>mzWindow</code></td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>m/z width of EICs (in Da)</td>
</tr>
<tr class="odd">
<td><code>retMin</code></td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>If <code>TRUE</code> plot retention times in minutes</td>
</tr>
<tr class="even">
<td><code>topMost</code></td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>Only draw this amount of highest intensity features in each group.</td>
</tr>
<tr class="odd">
<td>
<code>showPeakArea</code>, <code>showFGroupRect</code>
</td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>Fill peak areas / draw rectangles around feature groups?</td>
</tr>
<tr class="even">
<td><code>title</code></td>
<td>
<code><a href="../reference/generics.html">plotEIC()</a></code>, <code><a href="../reference/generics.html">plotSpec()</a></code>
</td>
<td>Override plot title</td>
</tr>
<tr class="odd">
<td><code>colourBy</code></td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>Colour individual feature groups (<code>"fGroups"</code>) or replicate groups (<code>"rGroups"</code>). By default nothing is coloured (<code>"none"</code>)</td>
</tr>
<tr class="even">
<td><code>showLegend</code></td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>Display a legend? (only if <code>colourBy!="none"</code>)</td>
</tr>
<tr class="odd">
<td><code>onlyPresent</code></td>
<td><code><a href="../reference/generics.html">plotEIC()</a></code></td>
<td>Only plot EICs for analyses where a feature was detected? Setting to <code>FALSE</code> is useful to inspect if a feature was ‘missed’.</td>
</tr>
<tr class="even">
<td>
<code>xlim</code>, <code>ylim</code>
</td>
<td>
<code><a href="../reference/generics.html">plotEIC()</a></code>, <code><a href="../reference/generics.html">plotSpec()</a></code>
</td>
<td>Override x/y axis ranges, i.e. to manually set plotting range.</td>
</tr>
<tr class="odd">
<td>
<code>groupName</code>, <code>analysis</code>, <code>precursor</code>, <code>index</code>
</td>
<td><code><a href="../reference/generics.html">plotSpec()</a></code></td>
<td>What to plot. See examples below.</td>
</tr>
<tr class="even">
<td><code>MSLevel</code></td>
<td><code><a href="../reference/generics.html">plotSpec()</a></code></td>
<td>Whether to plot an MS or MS/MS spectrum (only <code>MSPeakLists</code>)</td>
</tr>
<tr class="odd">
<td><code>formulas</code></td>
<td><code><a href="../reference/generics.html">plotSpec()</a></code></td>
<td>Whether <code>formula</code> annotation should be added (only <code>compounds</code>)</td>
</tr>
<tr class="even">
<td><code>plotStruct</code></td>
<td><code><a href="../reference/generics.html">plotSpec()</a></code></td>
<td>Whether the structure should be added to the plot (only <code>compounds</code>)</td>
</tr>
</tbody>
</table>
<p>Note that we can use <a href="#subset">subsetting</a> to select which feature data we want to plot, e.g.</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="fu"><a href="../reference/generics.html">plotEIC</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># only plot EICs from first and second analyses.</span>
<span class="fu"><a href="../reference/generics.html">plotEIC</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># only plot all features of first group</span></pre></div>
<p>The <code><a href="../reference/generics.html">plotStructure()</a></code> function will draw a chemical structure for a compound candidate. In addition, this function can draw the maximum common substructure (MCS) of multiple candidates in order to assess common structural features.</p>
<div class="sourceCode" id="cb41"><pre class="downlit">
<span class="co"># structure for first candidate</span>
<span class="fu"><a href="../reference/generics.html">plotStructure</a></span><span class="op">(</span><span class="va">compounds</span>, index <span class="op">=</span> <span class="fl">1</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span><span class="op">)</span>
<span class="co"># MCS for first three candidates</span>
<span class="fu"><a href="../reference/generics.html">plotStructure</a></span><span class="op">(</span><span class="va">compounds</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span><span class="op">)</span></pre></div>
<p>Some other common and less common plotting operations are shown below.</p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="fu"><a href="../reference/generics.html">plot</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span> <span class="co"># simple scatter plot of retention and m/z values</span>

<span class="fu"><a href="../reference/generics.html">plotEIC</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span> <span class="co"># plot EICs for all features</span>
<span class="co"># get overview of all feature groups</span>
<span class="fu"><a href="../reference/generics.html">plotEIC</a></span><span class="op">(</span><span class="va">fGroups</span>,
        colourBy <span class="op">=</span> <span class="st">"fGroup"</span>, <span class="co"># unique colour for each group</span>
        topMost <span class="op">=</span> <span class="fl">1</span>, <span class="co"># only most intense feature in each group</span>
        showPeakArea <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># show integrated areas</span>
        showFGroupRect <span class="op">=</span> <span class="cn">FALSE</span>,
        showLegend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># no legend (too busy for many feature groups)</span>
<span class="fu"><a href="../reference/generics.html">plotEIC</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="co"># only plot all features of first group</span>
        colourBy <span class="op">=</span> <span class="st">"rGroup"</span><span class="op">)</span> <span class="co"># and mark them individually per replicate group</span>
<span class="fu"><a href="../reference/generics.html">plotEIC</a></span><span class="op">(</span><span class="va">components</span>, index <span class="op">=</span> <span class="fl">7</span>, fGroups <span class="op">=</span> <span class="va">fGroups</span><span class="op">)</span> <span class="co"># EICs from a component</span>


<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">mslists</span>, <span class="st">"M120_R328_81"</span><span class="op">)</span> <span class="co"># non-annotated MS spectrum</span>
<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">mslists</span>, <span class="st">"M120_R328_81"</span>, MSLevel <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># non-annotated MS/MS spectrum</span>
<span class="co"># formula annotated spectrum</span>
<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">formulas</span>, precursor <span class="op">=</span> <span class="st">"C6H9N2"</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span>,
         MSPeakLists <span class="op">=</span> <span class="va">mslists</span><span class="op">)</span>
<span class="co"># compound annotated spectrum, with added formula annotations</span>
<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">compounds</span>, index <span class="op">=</span> <span class="fl">1</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span>, MSPeakLists <span class="op">=</span> <span class="va">mslists</span>,
         formulas <span class="op">=</span> <span class="va">formulas</span><span class="op">)</span>
<span class="co"># custom intensity range (e.g. to zoom in)</span>
<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">compounds</span>, index <span class="op">=</span> <span class="fl">1</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span>, MSPeakLists <span class="op">=</span> <span class="va">mslists</span>,
         ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5000</span><span class="op">)</span>, plotStruct <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="../reference/generics.html">plotSpec</a></span><span class="op">(</span><span class="va">components</span>, index <span class="op">=</span> <span class="fl">7</span><span class="op">)</span> <span class="co"># component spectrum</span></pre></div>
<div class="sourceCode" id="cb43"><pre class="downlit">
<span class="co"># Inspect homologous series</span>
<span class="fu"><a href="../reference/componentsNT-class.html">plotGraph</a></span><span class="op">(</span><span class="va">componNT</span><span class="op">)</span></pre></div>
</div>
<div id="visComp" class="section level3">
<h3 class="hasAnchor">
<a href="#visComp" class="anchor"></a>Comparing data</h3>
<p>There are three functions that can be used to visualize overlap and uniqueness between data:</p>
<table class="table">
<thead><tr class="header">
<th>Generic</th>
<th>Classes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>plotVenn</code></td>
<td>
<code>featureGroups</code>, <code>featureGroupsComparison</code>, <code>formulas</code>, <code>compounds</code>
</td>
</tr>
<tr class="even">
<td><code>plotUpSet</code></td>
<td>
<code>featureGroups</code>, <code>featureGroupsComparison</code>, <code>formulas</code>, <code>compounds</code>
</td>
</tr>
<tr class="odd">
<td><code>plotChord</code></td>
<td>
<code>featureGroups</code>, <code>featureGroupsComparison</code>
</td>
</tr>
</tbody>
</table>
<p>The most simple comparison plot is a Venn diagram (i.e. <code><a href="../reference/generics.html">plotVenn()</a></code>). This function is especially useful for two or three-way comparisons. More complex comparisons are better visualized with <a href="https://caleydo.org/tools/upset/">UpSet</a> diagrams (i.e. <code><a href="../reference/generics.html">plotUpSet()</a></code>). Finally, chord diagrams (i.e. <code><a href="../reference/generics.html">plotChord()</a></code>) provide visually pleasing diagrams to assess overlap between data.</p>
<p>These functions can either be used to compare feature data or different objects of the same type. The former is typically used to compare overlap or uniqueness between features in different replicate groups, whereas comparison between objects is useful to visualize differences in algorithmic output. Besides visualization, note that both operations can also be performed to modify or combine objects (see <a href="#unOv">unique and overlapping features</a> and <a href="#consensus">algorithm consensus</a>).</p>
<p>As usual, some examples are shown below.</p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="fu"><a href="../reference/generics.html">plotUpSet</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span> <span class="co"># compare replicate groups</span>
<span class="fu"><a href="../reference/generics.html">plotVenn</a></span><span class="op">(</span><span class="va">fGroups</span>, which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"repl1"</span>, <span class="st">"repl2"</span><span class="op">)</span><span class="op">)</span> <span class="co"># compare some replicate groups</span>
<span class="fu"><a href="../reference/generics.html">plotChord</a></span><span class="op">(</span><span class="va">fGroups</span>, average <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># overlap between replicate groups</span>
<span class="co"># compare with custom made groups</span>
<span class="fu"><a href="../reference/generics.html">plotChord</a></span><span class="op">(</span><span class="va">fGroups</span>, average <span class="op">=</span> <span class="cn">TRUE</span>,
          outer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>repl1 <span class="op">=</span> <span class="st">"grp1"</span>, repl2 <span class="op">=</span> <span class="st">"grp1"</span>, repl3 <span class="op">=</span> <span class="st">"grp2"</span>, repl4 <span class="op">=</span> <span class="st">"grp3"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># compare MetFrag and SIRIUS results</span>
<span class="fu"><a href="../reference/generics.html">plotVenn</a></span><span class="op">(</span><span class="va">compsMF</span>, <span class="va">compsSIR</span>,
         labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MF"</span>, <span class="st">"SIR"</span><span class="op">)</span><span class="op">)</span> <span class="co"># manual labeling</span></pre></div>
</div>
<div id="plotClust" class="section level3">
<h3 class="hasAnchor">
<a href="#plotClust" class="anchor"></a>Hierarchical clustering results</h3>
<p>In <code>patRoon</code> hierarchical clustering is used to generate components based on their intensity profiles (see <a href="#intclust">intensity clustering</a>) and to cluster candidate compounds with similar chemical structure (see <a href="#compclust">compound clustering</a>). The functions below can be used to visualize their results.</p>
<table class="table">
<colgroup>
<col width="17%">
<col width="36%">
<col width="45%">
</colgroup>
<thead><tr class="header">
<th>Generic</th>
<th>Classes</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/generics.html">plot()</a></code></td>
<td>
<code>componentsIntClust</code>, <code>compoundsCluster</code>
</td>
<td>Plots a dendrogram</td>
</tr>
<tr class="even">
<td><code><a href="../reference/generics.html">plotInt()</a></code></td>
<td><code>componentsIntClust</code></td>
<td>Plots normalized intensity profiles in a cluster</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/componentsIntClust-class.html">plotHeatMap()</a></code></td>
<td><code>componentsIntClust</code></td>
<td>Plots an heatmap</td>
</tr>
<tr class="even">
<td><code><a href="../reference/generics.html">plotSilhouettes()</a></code></td>
<td><code>componentsIntClust</code></td>
<td>Plot silhouette information to determine the cluster amount</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/generics.html">plotStructure()</a></code></td>
<td><code>compoundsCluster</code></td>
<td>Plots the maximum common substructure (MCS) of a cluster</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb45"><pre class="downlit">
<span class="fu"><a href="../reference/generics.html">plot</a></span><span class="op">(</span><span class="va">componInt</span><span class="op">)</span> <span class="co"># dendrogram</span>
<span class="fu"><a href="../reference/generics.html">plot</a></span><span class="op">(</span><span class="va">compsClust</span>, groupName <span class="op">=</span> <span class="st">"M120_R328_81"</span><span class="op">)</span> <span class="co"># dendrogram for clustered compounds</span>
<span class="fu"><a href="../reference/generics.html">plotInt</a></span><span class="op">(</span><span class="va">componInt</span>, index <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># intensities of 4th cluster</span></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit">
<span class="fu"><a href="../reference/componentsIntClust-class.html">plotHeatMap</a></span><span class="op">(</span><span class="va">componInt</span><span class="op">)</span> <span class="co"># plot heatmap</span></pre></div>
<div class="sourceCode" id="cb47"><pre class="downlit">
<span class="fu"><a href="../reference/componentsIntClust-class.html">plotHeatMap</a></span><span class="op">(</span><span class="va">componInt</span>, interactive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># interactive heatmap (with zoom-in!)</span>
<span class="fu"><a href="../reference/generics.html">plotSilhouettes</a></span><span class="op">(</span><span class="va">componInt</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="co"># plot silhouettes (e.g. to obtain ideal cluster amount)</span></pre></div>
</div>
<div id="interactive-plotting-of-chromatography-data" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-plotting-of-chromatography-data" class="anchor"></a>Interactive plotting of chromatography data</h3>
<p>The <code><a href="../reference/generics.html">plotEIC()</a></code> function introduced before can be used to visualize chromatography data for one or more features. An interactive alternative is to call the <code><a href="../reference/GUI-utils.html">checkChromatograms()</a></code> function. This function will launch a GUI that allows you to browse through all features and inspect their EICs. Simply pass in the <code>featureGroups</code> object you want to inspect:</p>
<div class="sourceCode" id="cb48"><pre class="downlit">
<span class="fu"><a href="../reference/GUI-utils.html">checkChromatograms</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span></pre></div>
<p>Note that this tool does not work well yet with large number of analyses/features. For this reason, it may be worthwile to launch it with subsets of your data, e.g.</p>
<div class="sourceCode" id="cb49"><pre class="downlit">
<span class="fu"><a href="../reference/GUI-utils.html">checkChromatograms</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">250</span><span class="op">]</span><span class="op">)</span> <span class="co"># only first 3 analyses and their first 500 feature groups</span></pre></div>
<p>Another purpose of the <code><a href="../reference/GUI-utils.html">checkChromatograms()</a></code> function is to remove ‘bad’ features (e.g. those which are probably not really features, but just noise). The workflow for this is:</p>
<ol style="list-style-type: decimal">
<li>Launch <code><a href="../reference/GUI-utils.html">checkChromatograms()</a></code> and remove any unwanted feature groups by disabling the <em>keep</em> checkbox.</li>
<li>Store the result of the <code><a href="../reference/GUI-utils.html">checkChromatograms()</a></code> function to a variable.</li>
<li>Use this variable to subset the original feature groups.</li>
</ol>
<p>To do so:</p>
<div class="sourceCode" id="cb50"><pre class="downlit">
<span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GUI-utils.html">checkChromatograms</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span>
<span class="va">fGroups</span> <span class="op">&lt;-</span> <span class="va">fGroups</span><span class="op">[</span>, <span class="va">keep</span><span class="op">]</span></pre></div>
<p>Note that when you re-run <code><a href="../reference/GUI-utils.html">checkChromatograms()</a></code> you can restore the state of which feature groups should be kept/removed by passing the previous result to the function:</p>
<div class="sourceCode" id="cb51"><pre class="downlit">
<span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GUI-utils.html">checkChromatograms</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span> <span class="co"># select feature groups</span>

<span class="co"># continue at a later stage</span>
<span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GUI-utils.html">checkChromatograms</a></span><span class="op">(</span><span class="va">fGroups</span>, enabledFGroups <span class="op">=</span> <span class="va">keep</span><span class="op">)</span></pre></div>
<blockquote>
<p><strong><em>NOTE</em></strong> Again, <code><a href="../reference/GUI-utils.html">checkChromatograms()</a></code> may be slow when processing larger datasets. The <a href="#report">reporting functionalities</a> provide a good alternative to quickly get an overview of all EIC data.</p>
</blockquote>
</div>
<div id="generating-eics-in-dataanalysis" class="section level3">
<h3 class="hasAnchor">
<a href="#generating-eics-in-dataanalysis" class="anchor"></a>Generating EICs in DataAnalysis</h3>
<p>If you have Bruker data and the DataAnalysis software installed, you can automatically add EIC data in a DataAnalysis session. The <code><a href="../reference/bruker-utils.html">addDAEIC()</a></code> will do this for a single <em>m/z</em> in one analysis, whereas the <code><a href="../reference/bruker-utils.html">addAllDAEICs()</a></code> function adds EICs for all features in a <code>featureGroups</code> object.</p>
<div class="sourceCode" id="cb52"><pre class="downlit">
<span class="co"># add a single EIC with background subtraction</span>
<span class="fu"><a href="../reference/bruker-utils.html">addDAEIC</a></span><span class="op">(</span><span class="st">"mysample"</span>, <span class="st">"~/path/to/sample"</span>, mz <span class="op">=</span> <span class="fl">120.1234</span>, bgsubtr <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co"># add TIC for MS/MS signal of precursor 120.1234 (value of mz is ignored for TICs)</span>
<span class="fu"><a href="../reference/bruker-utils.html">addDAEIC</a></span><span class="op">(</span><span class="st">"mysample"</span>, <span class="st">"~/path/to/sample"</span>, mz <span class="op">=</span> <span class="fl">100</span>, ctype <span class="op">=</span> <span class="st">"TIC"</span>,
         mtype <span class="op">=</span> <span class="st">"MSMS"</span>, fragpath <span class="op">=</span> <span class="st">"120.1234"</span>, name <span class="op">=</span> <span class="st">"MSMS 120"</span><span class="op">)</span>

<span class="fu"><a href="../reference/bruker-utils.html">addAllDAEICs</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span> <span class="co"># add EICs for all features</span>
<span class="fu"><a href="../reference/bruker-utils.html">addAllDAEICs</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span><span class="op">)</span> <span class="co"># as usual, subsetting can be used for partial data</span></pre></div>
</div>
</div>
<div id="report" class="section level2">
<h2 class="hasAnchor">
<a href="#report" class="anchor"></a>Reporting</h2>
<p>The previous sections showed various functionalies to inspect and visualize results. An easy and automated way to do this automatically is by using the <em>reporting</em> functionality of <code>patRoon</code>. The following three reporting functions are available:</p>
<ul>
<li>
<code><a href="../reference/reporting.html">reportCSV()</a></code>: exports workflow data to comma-separated value (csv) files</li>
<li>
<code><a href="../reference/reporting.html">reportPDF()</a></code>: generates simple reports by plotting workflow data in portable document files (PDFs)</li>
<li>
<code><a href="../reference/reporting.html">reportHTML()</a></code>: generates interactive and easily explorable reports</li>
</ul>
<p>There are many different arguments available to configure the reporting process. Some common arguments are listed below; for a complete listing see the reference manual (e.g. <code><a href="../reference/reporting.html">?reporting</a></code>).</p>
<table class="table">
<colgroup>
<col width="29%">
<col width="23%">
<col width="46%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Functions</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>fGroups</code>, <code>formulas</code>, <code>compounds</code>, <code>formulas</code>, <code>components</code>, <code>compsCluster</code>
</td>
<td>All</td>
<td>Objects to plot. Only <code>fGroups</code> is mandatory.</td>
</tr>
<tr class="even">
<td><code>MSPeakLists</code></td>
<td>
<code><a href="../reference/reporting.html">reportPDF()</a></code>, <code><a href="../reference/reporting.html">reportHTML()</a></code>
</td>
<td>The <code>MSPeakLists</code> object that was used to generate annotation data. Only needs to be specified if <code>formulas</code> or <code>compounds</code> are reported.</td>
</tr>
<tr class="odd">
<td><code>path</code></td>
<td>All</td>
<td>Directory path where report files will be stored (<code>"report"</code> by default).</td>
</tr>
<tr class="even">
<td>
<code>formulasTopMost</code>, <code>compoundsTopMost</code>
</td>
<td>
<code><a href="../reference/reporting.html">reportPDF()</a></code>, <code><a href="../reference/reporting.html">reportHTML()</a></code>
</td>
<td>Report no more than this amount of highest ranked candidates.</td>
</tr>
<tr class="odd">
<td><code>EICOnlyPresent</code></td>
<td>
<code><a href="../reference/reporting.html">reportPDF()</a></code>, <code><a href="../reference/reporting.html">reportHTML()</a></code>
</td>
<td>Only plot an EIC for an analysis if a feature was detected.</td>
</tr>
<tr class="even">
<td><code>selfContained</code></td>
<td><code><a href="../reference/reporting.html">reportHTML()</a></code></td>
<td>Outputs to a single and self contained <code>.html</code> file. Handy to share reports, but not recommended for large amounts of data.</td>
</tr>
</tbody>
</table>
<p>Which data will be reported is fully configurable. The only workflow object that must be specified are the feature groups (i.e. with the <code>fGroups</code> argument), all other data (e.g. <code>compounds</code>, <code>components</code>) are optional. This means that reporting can be performed at every stage during the workflow, which, for instance, can be useful to quickly inspect results when testing out various settings to generate workflow data.</p>
<p>When formula or compound results are reported with <code><a href="../reference/reporting.html">reportPDF()</a></code> or <code><a href="../reference/reporting.html">reportHTML()</a></code> then only the top ranked candidates are considered. This limtation is often necessary as reporting many candidates will take considerable time. By default the top 5 for each feature group are reported, however, this number can be changed with the <code>formulasTopMost</code> and <code>compoundsTopMost</code> arguments.</p>
<p>Some typical examples:</p>
<div class="sourceCode" id="cb53"><pre class="downlit">
<span class="fu"><a href="../reference/reporting.html">reportHTML</a></span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span> <span class="co"># simple interactive report with feature data</span>
<span class="co"># generate PDFs with feature and compound annotation data</span>
<span class="fu"><a href="../reference/reporting.html">reportPDF</a></span><span class="op">(</span><span class="va">fGroups</span>, compounds <span class="op">=</span> <span class="va">compounds</span>, MSPeakLists <span class="op">=</span> <span class="va">mslists</span><span class="op">)</span>
<span class="fu"><a href="../reference/reporting.html">reportCSV</a></span><span class="op">(</span><span class="va">fGroups</span>, path <span class="op">=</span> <span class="st">"myReport"</span><span class="op">)</span> <span class="co"># change destination path</span>

<span class="co"># generate report with all workflow types and increase maximum number of</span>
<span class="co"># compound candidates to top 10</span>
<span class="fu"><a href="../reference/reporting.html">reportHTML</a></span><span class="op">(</span><span class="va">fGroups</span>, formulas <span class="op">=</span> <span class="va">formulas</span>, compounds <span class="op">=</span> <span class="va">compounds</span>,
           components <span class="op">=</span> <span class="va">components</span>, MSPeakLists <span class="op">=</span> <span class="va">mslists</span>,
           compsCluster <span class="op">=</span> <span class="va">compsClust</span>,
           compoundsTopMost <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></pre></div>
</div>
</div>
<div id="advanced-usage" class="section level1">
<h1 class="hasAnchor">
<a href="#advanced-usage" class="anchor"></a>Advanced usage</h1>
<div id="adducts" class="section level2">
<h2 class="hasAnchor">
<a href="#adducts" class="anchor"></a>Adducts</h2>
<p>When generating formulae and compound annotations and some other functionalities it is required to specify the adduct species. Behind the scenes, different algorithms typically use different formats. For instance, in order to specify a protonated species…</p>
<ul>
<li>
<code>GenForm</code> either accepts <code>"M+H"</code> and <code>"+H"</code>
</li>
<li>
<code>MetFrag</code> either accepts the numeric code <code>1</code> or <code>"[M+H]+"</code>
</li>
<li>
<code>SIRIUS</code> accepts <code>"[M+H]+"</code>
</li>
</ul>
<p>In addition, most algorithms only accept a limited set of possible adducts, which do not necessarily all overlap with each other. The <code><a href="../reference/adduct-utils.html">GenFormAdducts()</a></code> and <code><a href="../reference/adduct-utils.html">MetFragAdducts()</a></code> functions list the possible adducts for <code>GenForm</code> and <code>MetFrag</code>, respectively.</p>
<p>In order to simplify the situation <code>patRoon</code> internally uses its own format and converts it automatically to the algorithm specific format when necessary. Furthermore, during conversion it checks if the specified adduct format is actually allowed by the algorithm. Adducts in <code>patRoon</code> are stored in the <code>adduct</code> S4 class. Objects from this class specify which elements are added and/or subtracted, the final charge and the number of molecules present in the adduct (e.g. <em>2</em> for a dimer).</p>
<div class="sourceCode" id="cb54"><pre class="downlit">
<span class="fu"><a href="../reference/adduct-class.html">adduct</a></span><span class="op">(</span>add <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span> <span class="co"># [M+H]+</span>
<span class="fu"><a href="../reference/adduct-class.html">adduct</a></span><span class="op">(</span>sub <span class="op">=</span> <span class="st">"H"</span>, charge <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="co"># [M-H]-</span>
<span class="fu"><a href="../reference/adduct-class.html">adduct</a></span><span class="op">(</span>add <span class="op">=</span> <span class="st">"K"</span>, sub <span class="op">=</span> <span class="st">"H2"</span>, charge <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="co"># [M+K-H2]-</span>
<span class="fu"><a href="../reference/adduct-class.html">adduct</a></span><span class="op">(</span>add <span class="op">=</span> <span class="st">"H3"</span>, charge <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="co"># [M+H3]3+</span>
<span class="fu"><a href="../reference/adduct-class.html">adduct</a></span><span class="op">(</span>add <span class="op">=</span> <span class="st">"H"</span>, molMult <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># [2M+H]+</span></pre></div>
<p>A more easy way to generate adduct objects is by using the <code><a href="../reference/adduct-utils.html">as.adduct()</a></code> function:</p>
<div class="sourceCode" id="cb55"><pre class="downlit">
<span class="fu"><a href="../reference/adduct-utils.html">as.adduct</a></span><span class="op">(</span><span class="st">"[M+H]+"</span><span class="op">)</span>
<span class="fu"><a href="../reference/adduct-utils.html">as.adduct</a></span><span class="op">(</span><span class="st">"[M+H2]2+"</span><span class="op">)</span>
<span class="fu"><a href="../reference/adduct-utils.html">as.adduct</a></span><span class="op">(</span><span class="st">"[2M+H]+"</span><span class="op">)</span>
<span class="fu"><a href="../reference/adduct-utils.html">as.adduct</a></span><span class="op">(</span><span class="st">"[M-H]-"</span><span class="op">)</span>
<span class="fu"><a href="../reference/adduct-utils.html">as.adduct</a></span><span class="op">(</span><span class="st">"+H"</span>, format <span class="op">=</span> <span class="st">"genform"</span><span class="op">)</span>
<span class="fu"><a href="../reference/adduct-utils.html">as.adduct</a></span><span class="op">(</span><span class="fl">1</span>, isPositive <span class="op">=</span> <span class="cn">TRUE</span>, format <span class="op">=</span> <span class="st">"metfrag"</span><span class="op">)</span></pre></div>
<p>In fact, the <code>adduct</code> argument to workflow functions such as <code><a href="../reference/formula-generation.html">generateFormulas()</a></code> and <code><a href="../reference/compound-generation.html">generateCompounds()</a></code> is automatically converted to an <code>adduct</code> class with the <code><a href="../reference/adduct-utils.html">as.adduct()</a></code> function if necessary:</p>
<div class="sourceCode" id="cb56"><pre class="downlit">
<span class="va">formulas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">...</span>, adduct <span class="op">=</span> <span class="fu"><a href="../reference/adduct-class.html">adduct</a></span><span class="op">(</span>sub <span class="op">=</span> <span class="st">"H"</span>, charge <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">formulas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">...</span>, adduct <span class="op">=</span> <span class="st">"[M-H]-"</span><span class="op">)</span> <span class="co"># same as above</span></pre></div>
<p>More details can be found in the reference manual (<code><a href="../reference/adduct-class.html">?adduct</a></code> and <code><a href="../reference/adduct-utils.html">?`adduct-utils`</a></code>).</p>
</div>
<div id="fOpt" class="section level2">
<h2 class="hasAnchor">
<a href="#fOpt" class="anchor"></a>Feature parameter optimization</h2>
<p>Many different parameters exist that may affect the output quality of feature finding and grouping. To avoid time consuming manual experimentation, functionality is provided to largely automate the optimization process. The methodology, which uses design of experiments (DoE), is based on the excellent <a href="IPO">Isotopologue Parameter Optimization (IPO)</a> R package. The functionality of this package is directly integrated in patRoon. Some functionality was added or changed, the most important being support for other feature finding and grouping algorithms besides <a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a> and basic optimization support for qualitative parameters. Nevertheless, the core optimization algorithms are largely untouched.</p>
<p>This section will introduce the most important concepts and functionalities. Please see the reference manual for more information (e.g. <code><a href="../reference/feature-optimization.html">?`feature-optimization`</a></code>).</p>
<div id="parameter-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#parameter-sets" class="anchor"></a>Parameter sets</h3>
<p>Before starting an optimization experiment we have to define <em>parameter sets</em>. These sets contain the parameters and (initial) numeric ranges that should be tested. A parameter set is defined as a regular <code>list</code>, and can be easily constructed with the <code><a href="../reference/feature-optimization.html">generateFeatureOptPSet()</a></code> and <code><a href="../reference/feature-optimization.html">generateFGroupsOptPSet()</a></code> functions (for feature finding and feature grouping, respectively).</p>
<div class="sourceCode" id="cb57"><pre class="downlit">
<span class="va">pSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">generateFeatureOptPSet</a></span><span class="op">(</span><span class="st">"openms"</span><span class="op">)</span> <span class="co"># default test set for OpenMS</span>
<span class="va">pSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">generateFeatureOptPSet</a></span><span class="op">(</span><span class="st">"openms"</span>, chromSNR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="co"># add parameter</span>
<span class="co"># of course manually making a list is also possible (e.g. if you don't want to test the default parameters)</span>
<span class="va">pSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>noiseThrInt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>When optimizing with <a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">XCMS</a> a few things have to be considered. First of all, when using the XCMS3 interface (i.e. <code>algorithm="xcms3"</code>) the underlying method that should be used for finding and grouping features and retention alignment should be set. In case these are not set default methods will be used.</p>
<div class="sourceCode" id="cb58"><pre class="downlit">
<span class="va">pSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"centWave"</span>, ppm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span>
<span class="va">pSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>ppm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span> <span class="co"># same: centWave is default</span>

<span class="co"># get defaults, but for different grouping/alignment methods</span>
<span class="va">pSetFG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">generateFGroupsOptPSet</a></span><span class="op">(</span><span class="st">"xcms3"</span>, groupMethod <span class="op">=</span> <span class="st">"nearest"</span>, retAlignMethod <span class="op">=</span> <span class="st">"peakgroups"</span><span class="op">)</span></pre></div>
<p>In addition, when optimizing feature grouping (both XCMS interfaces) we need to set the grouping and retention alignment parameters in two different nested lists: these are <code>groupArgs</code>/<code>retcorArgs</code> (<code>algorithm="xcms"</code>) and <code>groupParams</code>/<code>retAlignParams</code> (<code>algorithm="xcms3"</code>).</p>
<div class="sourceCode" id="cb59"><pre class="downlit">
<span class="va">pSetFG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>groupParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>bw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># xcms3</span>
<span class="va">pSetFG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>retcorArgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>gapInit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">7</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># xcms</span></pre></div>
<p>When a parameter set has been defined it should be used as input for the <code><a href="../reference/feature-optimization.html">optimizeFeatureFinding()</a></code> or <code><a href="../reference/feature-optimization.html">optimizeFeatureGrouping()</a></code> functions.</p>
<div class="sourceCode" id="cb60"><pre class="downlit">
<span class="va">ftOpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">optimizeFeatureFinding</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"openms"</span>, <span class="va">pSet</span><span class="op">)</span>
<span class="va">fgOpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">optimizeFeatureGrouping</a></span><span class="op">(</span><span class="va">fList</span>, <span class="st">"openms"</span>, <span class="va">pSetFG</span><span class="op">)</span> <span class="co"># fList is an existing features object</span></pre></div>
<p>Similar to <code><a href="../reference/feature-finding.html">findFeatures()</a></code>, the first argument to <code><a href="../reference/feature-optimization.html">optimizeFeatureFinding()</a></code> should be the <a href="#anaInfo">analysis information</a>. Note that it is not uncommon to perform the optimization with only a subset of (representative) analyses (i.e. to reduce processing time).</p>
<div class="sourceCode" id="cb61"><pre class="downlit">
<span class="va">ftOpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">optimizeFeatureFinding</a></span><span class="op">(</span><span class="va">anaInfo</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="op">]</span>, <span class="st">"openms"</span>, <span class="va">pSet</span><span class="op">)</span> <span class="co"># only use first two analyses</span></pre></div>
<p>From the parameter set a design of experiment will be automatically created. Obviously, the more parameters are specified, the longer such an experiment will take. After an experiment has finished, the optimization algorithm will start a new experiment where numeric ranges for each parameter are increased or decreased in order to more accurately find optimum values. Hence, the numeric ranges specified in the parameter set are only <em>initial</em> ranges, and will be changed in subsequent experiments. After each experiment iteration the results will be evaluated and a new experiment will be started as long as better results were obtained during the last experiment (although there is a hard limit defined by the <code>maxIterations</code> argument).</p>
<p>For some parameters it is recommended or even necessary to set hard limits on the numeric ranges that are allowed to be tested. For instance, setting a minimum feature intensity threshold is highly recommended to avoid excessive processing time and potentially suboptimal results due to excessive amounts of resulting features. Configuring absolute parameter ranges is done by setting the <code>paramRanges</code> argument.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1"></a><span class="co"># set minimum intensity threshold (but no max)</span></span>
<span id="cb62-2"><a href="#cb62-2"></a>ftOpt &lt;-<span class="st"> </span><span class="kw">optimizeFeatureFinding</span>(anaInfo, <span class="st">"openms"</span>,</span>
<span id="cb62-3"><a href="#cb62-3"></a>                                <span class="kw">list</span>(<span class="dt">noiseThrInt =</span> <span class="kw">c</span>(<span class="dv">1000</span>, <span class="dv">5000</span>), <span class="co"># initial testing range</span></span>
<span id="cb62-4"><a href="#cb62-4"></a>                                <span class="dt">paramRanges =</span> <span class="kw">list</span>(<span class="dt">noiseThrInt =</span> <span class="kw">c</span>(<span class="dv">500</span>, <span class="ot">Inf</span>))) <span class="co"># never test below 500</span></span></code></pre></div>
<p>Depending on the used algorithm, several default absolute limits are imposed. These may be obtained with the <code><a href="../reference/feature-optimization.html">getDefFeaturesOptParamRanges()</a></code> and <code><a href="../reference/feature-optimization.html">getDefFGroupsOptParamRanges()</a></code> functions.</p>
<p>The common operation is to optimize numeric parameters. However, parameters that are not numeric (i.e. <em>qualitative</em>) need a different approach. In this case you will need to define multiple parameter sets, where each set defines a different qualitative value.</p>
<div class="sourceCode" id="cb63"><pre class="downlit">
<span class="va">ftOpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">optimizeFeatureFinding</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"openms"</span>,
                                <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>chromFWHM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">8</span><span class="op">)</span>, isotopeFilteringModel <span class="op">=</span> <span class="st">"metabolites (5% RMS)"</span><span class="op">)</span>,
                                <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>chromFWHM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">8</span><span class="op">)</span>, isotopeFilteringModel <span class="op">=</span> <span class="st">"metabolites (2% RMS)"</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>In the above example there are two parameter sets: both define the numeric <code>chromFWHM</code> parameter, whereas the qualitative <code>isotopeFilteringModel</code> parameter has a different value for each. Note that we had to specify the <code>chromFWHM</code> twice, this can be remediated by using the <code>templateParams</code> argument:</p>
<div class="sourceCode" id="cb64"><pre class="downlit">
<span class="va">ftOpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/feature-optimization.html">optimizeFeatureFinding</a></span><span class="op">(</span><span class="va">anaInfo</span>, <span class="st">"openms"</span>,
                                <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>isotopeFilteringModel <span class="op">=</span> <span class="st">"metabolites (5% RMS)"</span><span class="op">)</span>,
                                <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>isotopeFilteringModel <span class="op">=</span> <span class="st">"metabolites (2% RMS)"</span><span class="op">)</span>,
                                templateParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>chromFWHM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>As its name suggests, the <code>templateParams</code> argument serves as a template parameter set, and its values are essentially combined with each given parameter set. Note that current support for optimizing qualitative parameters is relatively basic and time consuming. This is because tests are essentially repeated for each parameter set (e.g. in the above example the <code>chromFWHM</code> parameter is optimized twice, each time with a different value for <code>isotopeFilteringModel</code>).</p>
</div>
<div id="processing-optmization-results" class="section level3">
<h3 class="hasAnchor">
<a href="#processing-optmization-results" class="anchor"></a>Processing optmization results</h3>
<p>The results of an optimization process are stored in objects from the S4 <code>optimizationResult</code> class. Several methods are defined to inspect and visualize these results.</p>
<p>The <code><a href="../reference/optimizationResult-class.html">optimizedParameters()</a></code> function is used to inspect the best parameter settings. Similarly, the <code><a href="../reference/optimizationResult-class.html">optimizedObject()</a></code> function can be used to obtain the object that was created with these settings (i.e. a <code>features</code> or <code>featureGroups</code> object).</p>
<div class="sourceCode" id="cb65"><pre class="downlit">
<span class="fu"><a href="../reference/optimizationResult-class.html">optimizedParameters</a></span><span class="op">(</span><span class="va">ftOpt</span><span class="op">)</span> <span class="co"># best settings for whole experiment</span>
<span class="fu"><a href="../reference/optimizationResult-class.html">optimizedObject</a></span><span class="op">(</span><span class="va">ftOpt</span><span class="op">)</span> <span class="co"># features object with best settings for whole experiment</span></pre></div>
<p>Results can also be obtained for specific parameter sets/iterations.</p>
<div class="sourceCode" id="cb66"><pre class="downlit">
<span class="fu"><a href="../reference/optimizationResult-class.html">optimizedParameters</a></span><span class="op">(</span><span class="va">ftOpt</span>, <span class="fl">1</span><span class="op">)</span> <span class="co"># best settings for first parameter set</span>
<span class="fu"><a href="../reference/optimizationResult-class.html">optimizedParameters</a></span><span class="op">(</span><span class="va">ftOpt</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span> <span class="co"># best settings for first parameter set and experiment iteration</span>
<span class="fu"><a href="../reference/optimizationResult-class.html">optimizedObject</a></span><span class="op">(</span><span class="va">ftOpt</span>, <span class="fl">1</span><span class="op">)</span> <span class="co"># features object with best settings for first parameter set</span></pre></div>
<p>The <code><a href="../reference/generics.html">plot()</a></code> function can be used to visualize optimization results. This function will plot graphs for results of all tested parameter pairs. The graphs can be contour, image or perspective plots (as specified by the <code>type</code> argument).</p>
<div class="sourceCode" id="cb67"><pre class="downlit">
<span class="fu"><a href="../reference/generics.html">plot</a></span><span class="op">(</span><span class="va">ftOpt</span>, paramSet <span class="op">=</span> <span class="fl">1</span>, DoEIteration <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># contour plots for first param set/experiment</span>
<span class="fu"><a href="../reference/generics.html">plot</a></span><span class="op">(</span><span class="va">ftOpt</span>, paramSet <span class="op">=</span> <span class="fl">1</span>, DoEIteration <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"persp"</span><span class="op">)</span> <span class="co"># pretty perspective plots</span></pre></div>
<p>Please refer to the reference manual for more methods to inspect optimization results (e.g. <code><a href="../reference/optimizationResult-class.html">?optimizationResult</a></code>).</p>
</div>
</div>
<div id="exporting-and-converting-feature-data" class="section level2">
<h2 class="hasAnchor">
<a href="#exporting-and-converting-feature-data" class="anchor"></a>Exporting and converting feature data</h2>
<p>The feature group data obtained during the workflow can be exported to various formats with the <code><a href="../reference/featureGroups-class.html">export()</a></code> generic function. There are currently three formats supported: <code>"brukerpa"</code> (Bruker ProfileAnalysis), <code>"brukertasq"</code> (Bruker TASQ) and <code>"mzmine"</code> (mzMine). The former exports a ‘bucket table’ which can be loaded in ProfileAnalysis, the second and third export a target list that can be processed with TASQ and mzMine, respectively.</p>
<p>The <code><a href="../reference/getXCMSSet.html">getXCMSSet()</a></code> function converts a <code>features</code> or <code>featureGroups</code> object to an <code>xcmsSet</code> object which can be used for further processing with <a href="https://www.bioconductor.org/packages/release/bioc/html/xcms.html">xcms</a>. Similarly, the <code>getXCMSnExp()</code> function can be used for conversion to an XCMS3 style <code>XCMSnExp</code> object.</p>
<p>Some examples for these functions are shown below.</p>
<div class="sourceCode" id="cb68"><pre class="downlit">
<span class="fu"><a href="../reference/featureGroups-class.html">export</a></span><span class="op">(</span><span class="va">fGroups</span>, <span class="st">"brukertasq"</span>, out <span class="op">=</span> <span class="st">"my_targets.csv"</span><span class="op">)</span>

<span class="co"># convert features to xcmsSet.</span>
<span class="co"># NOTE: exportedData should only be FALSE when the analysis data files cannot be</span>
<span class="co"># loaded by XCMS (e.g. when obtained with DataAnalysis)</span>
<span class="va">xset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getXCMSSet.html">getXCMSSet</a></span><span class="op">(</span><span class="va">fList</span>, exportedData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">xsetg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getXCMSSet.html">getXCMSSet</a></span><span class="op">(</span><span class="va">fGroups</span>, exportedData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># get grouped xcmsSet</span>

<span class="co"># using the new XCMS3 interface</span>
<span class="co"># NOTE: for XCMS3 data currently always has to be exported</span>
<span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu">getXCMSnExp</span><span class="op">(</span><span class="va">fList</span><span class="op">)</span>
<span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu">getXCMSnExp</span><span class="op">(</span><span class="va">fGroups</span><span class="op">)</span></pre></div>
</div>
<div id="consensus" class="section level2">
<h2 class="hasAnchor">
<a href="#consensus" class="anchor"></a>Algorithm consensus</h2>
<p>With <code>patRoon</code> you have the option to choose between several algorithms for most workflow steps. Each algorithm is typically characterized by its efficiency, robustness, and may be optimized towards certain data properties. Comparing their output is therefore advantegeuous in order to design an optimum workflow. The <code><a href="../reference/generics.html">consensus()</a></code> generic function will compare different results from different algorithms and returns a <em>consensus</em>, which may be based on minimal overlap, uniqueness or simply a combination of all results from involved objects. The output from the <code><a href="../reference/generics.html">consensus()</a></code> function is of similar type as the input types and is therefore compatible to any ‘regular’ further data processing operations (e.g. input for other workflow steps or plotting). Note that a consensus can also be made from objects generated by the same algorithm, for instance, to compare or combine results obtained with different parameters (e.g. different databases used for compound annotation).</p>
<p>The <code><a href="../reference/generics.html">consensus()</a></code> generic is defined for most workflow objects. Some of its common function arguments are listed below.</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="40%">
<col width="49%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Classes</th>
<th>Remarks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>obj</code>, <code>...</code>
</td>
<td>All</td>
<td>Two or more objects (of the same type) that should be compared to generate the consensus.</td>
</tr>
<tr class="even">
<td>
<code>compThreshold</code>, <code>relAbundance</code>, <code>absAbundance</code>, <code>formThreshold</code>
</td>
<td>
<code>compounds</code>, <code>formulas</code>, <code>featureGroupsComparison</code>
</td>
<td>The minimum overlap (relative/absolute) for a result (feature, candidate) to be kept.</td>
</tr>
<tr class="odd">
<td><code>uniqueFrom</code></td>
<td>
<code>compounds</code>, <code>formulas</code>, <code>featureGroupsComparison</code>
</td>
<td>Only keep <em>unique</em> results from specified objects.</td>
</tr>
<tr class="even">
<td><code>uniqueOuter</code></td>
<td>
<code>compounds</code>, <code>formulas</code>, <code>featureGroupsComparison</code>
</td>
<td>Should be combined with <code>uniqueFrom</code>. If <code>TRUE</code> then only results are kept which are <em>also</em> unique between the objects specified with <code>uniqueFrom</code>.</td>
</tr>
</tbody>
</table>
<p>Note that current support for generating a consensus between <code>components</code> objects is very simplistic; here results are not compared, but the consensus simply consists a combination of all the components from each object.</p>
<p>Generating a consensus for feature groups involves first generating a <code>featureGroupsComparison</code> object. This step is necessary since (small) deviations between retention times and/or mass values reported by different feature finding/grouping algorithms complicates a direct comparison. The comparison objects are made by the <code><a href="../reference/featureGroups-compare.html">comparison()</a></code> function, and its results can be visualized by the plotting functions discussed <a href="#visComp">in the previous chapter</a>.</p>
<p>Some examples are shown below</p>
<div class="sourceCode" id="cb69"><pre class="downlit">
<span class="va">compoundsCons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">consensus</a></span><span class="op">(</span><span class="va">compoundsMF</span>, <span class="va">compoundsSIR</span><span class="op">)</span> <span class="co"># combine MetFrag/SIRIUS results</span>
<span class="va">compoundsCons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">consensus</a></span><span class="op">(</span><span class="va">compoundsMF</span>, <span class="va">compoundsSIR</span>,
                           compThreshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># only keep results that overlap</span>

<span class="va">fGroupComp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/featureGroups-compare.html">comparison</a></span><span class="op">(</span><span class="va">fGroupsXCMS</span>, <span class="va">fGroupsOpenMS</span>, <span class="va">fGroupsEnviPick</span>,
                         groupAlgo <span class="op">=</span> <span class="st">"openms"</span><span class="op">)</span>
<span class="fu"><a href="../reference/generics.html">plotVenn</a></span><span class="op">(</span><span class="va">fGroupComp</span><span class="op">)</span> <span class="co"># visualize overlap/uniqueness</span>
<span class="va">fGroupsCons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">consensus</a></span><span class="op">(</span><span class="va">fGroupComp</span>,
                         uniqueFrom <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="co"># only keep results unique in OpenMS+XCMS</span>
<span class="va">fGroupsCons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">consensus</a></span><span class="op">(</span><span class="va">fGroupComp</span>,
                         uniqueFrom <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,
                         uniqueOuter <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># as above, but also exclude any overlap between OpenMS/XCMS</span></pre></div>
</div>
<div id="compclust" class="section level2">
<h2 class="hasAnchor">
<a href="#compclust" class="anchor"></a>Compound clustering</h2>
<p>When large databases such as <a href="https://pubchem.ncbi.nlm.nih.gov/">PubChem</a> or <a href="http://www.chemspider.com/">ChemSpider</a> are used for compound annotation, it is common to find <em>many</em> candidate structures for even a single feature. While choosing the right scoring settings can significantly improve their ranking, it is still very much possible that many candidates of potential interest remain. In this situation it might help to perform <em>compound clustering</em>. During this process, all candidates for a feature are clustered hierarchically on basis of similar chemical structure. From the resulting cluster the <em>maximum common substructure</em> (MCS) can be derived, which represents the largest possible substructure that ‘fit’ in all candidates. By visual inspection of the MCS it may be possible to identify likely common structural properties of a feature.</p>
<p>In order to perform compound clustering the <code><a href="../reference/compounds-cluster.html">makeHCluster()</a></code> generic function should be used. This function heavily relies on chemical fingerprinting functionality provided by <a href="https://github.com/CDK-R/cdkr">rcdk</a>.</p>
<div class="sourceCode" id="cb70"><pre class="downlit">
<span class="va">compounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compound-generation.html">generateCompounds</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="co"># get our compounds</span>
<span class="va">compsClust</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compounds-cluster.html">makeHCluster</a></span><span class="op">(</span><span class="va">compounds</span><span class="op">)</span></pre></div>
<p>This function accepts several arguments to fine tune the clustering process:</p>
<ul>
<li>
<code>method</code>: the clustering method (e.g. <code>"complete"</code> (default), <code>"ward.D2"</code>), see <code><a href="https://rdrr.io/r/stats/hclust.html">?hclust</a></code> for options</li>
<li>
<code>fpType</code>: finger printing type (<code>"extended"</code> by default), see <code>?get.fingerprint</code>
</li>
<li>
<code>fpSimMethod</code>: similarity method for generating the distance method (<code>"tanimoto"</code> by default), see <code>?fp.sim.matrix</code>
</li>
</ul>
<p>For all arguments see the reference manual (<code>?makeHClust</code>).</p>
<p>The resulting object is of type <code>compoundsCluster</code>. Several methods are defined to modify and inspect these results:</p>
<div class="sourceCode" id="cb71"><pre class="downlit">
<span class="co"># plot MCS of first cluster from candidates of M109_R116_61</span>
<span class="fu"><a href="../reference/generics.html">plotStructure</a></span><span class="op">(</span><span class="va">compsClust</span>, groupName <span class="op">=</span> <span class="st">"M109_R116_61"</span>, <span class="fl">1</span><span class="op">)</span>

<span class="co"># plot dendrogram</span>
<span class="fu"><a href="../reference/generics.html">plot</a></span><span class="op">(</span><span class="va">compsClust</span>, groupName <span class="op">=</span> <span class="st">"M109_R116_61"</span><span class="op">)</span>

<span class="co"># re-assign clusters for a feature group</span>
<span class="va">compsClust</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">treeCut</a></span><span class="op">(</span><span class="va">compsClust</span>, k <span class="op">=</span> <span class="fl">5</span>, groupName <span class="op">=</span> <span class="st">"M109_R116_61"</span><span class="op">)</span>
<span class="co"># ditto, but automatic cluster determination</span>
<span class="va">compsClust</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">treeCutDynamic</a></span><span class="op">(</span><span class="va">compsClust</span>, minModuleSize <span class="op">=</span> <span class="fl">3</span>, groupName <span class="op">=</span> <span class="st">"M109_R116_61"</span><span class="op">)</span></pre></div>
<p>For a complete overview see the reference manual (<code><a href="../reference/compoundsCluster-class.html">?compoundsCluster</a></code>).</p>
</div>
<div id="basic-quantitative-and-regression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-quantitative-and-regression-analysis" class="anchor"></a>Basic quantitative and regression analysis</h2>
<p>While <code>patRoon</code> is currently primarily focused on qualitative analyses, some <em>basic</em> quantitative analysis can also be performed, for instance, to estimate concentrations of features. In fact, other types of data that may be useful for regression analysis can be set such as sample dilution factor or sampling time. The latter may, for instance, be used to isolate features with increasing or decreasing intensity. Regardless of what kind of regression analysis is performed, here we simply refer the values to be calculated as <em>concentrations</em>. In order to use this functionality, an extra column (<code>conc</code>) should be added to the <a href="#anaInfo">analysis information</a>, for instance:</p>
<div class="sourceCode" id="cb72"><pre class="downlit">
<span class="co"># obtain analysis information as usual, but add some concentrations.</span>
<span class="co"># The blanks are set to NA, whereas the standards are set to increasing levels.</span>
<span class="va">anaInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/analysis-information.html">generateAnalysisInfo</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="fu">patRoonData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/patRoonData/man/patRoonData.html">exampleDataPath</a></span><span class="op">(</span><span class="op">)</span>,
                                groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"solvent"</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"standard"</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,
                                blanks <span class="op">=</span> <span class="st">"solvent"</span>,
                                concs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>For analyses with known concentrations (e.g. standards) the <code>conc</code> column should be set; for all others the value should be set to <code>NA</code>.</p>
<p>The <code><a href="../reference/generics.html">as.data.table()</a></code> function (or <code><a href="../reference/generics.html">as.data.frame()</a></code>) can then be used to calculate regression data and estimate concentrations:</p>
<div class="sourceCode" id="cb73"><pre class="downlit">
<span class="co"># use areas for quantitation and make sure that feature data is reported</span>
<span class="co"># (otherwise no concentrations are calculated)</span>
<span class="co"># (only relevant columns are shown for clarity)</span>
<span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">fGroups</span>, areas <span class="op">=</span> <span class="cn">TRUE</span>, features <span class="op">=</span> <span class="cn">TRUE</span>, regression <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<p>Calculated concentrations are stored in the <code>conc_reg</code> column, alongside while other regression data (i.e. <code>RSQ</code>, <code>intercept</code>, <code>slope</code> columns). To perform basic trend analysis the <code>RSQ</code> (i.e. R squared) can be used:</p>
<div class="sourceCode" id="cb74"><pre class="downlit">
<span class="va">fGroupsTab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generics.html">as.data.table</a></span><span class="op">(</span><span class="va">fGroups</span>, areas <span class="op">=</span> <span class="cn">TRUE</span>, features <span class="op">=</span> <span class="cn">FALSE</span>, regression <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co"># subset fGroups with reasonable correlation</span>
<span class="va">increasingFGroups</span> <span class="op">&lt;-</span> <span class="va">fGroups</span><span class="op">[</span>, <span class="va">fGroupsTab</span><span class="op">[</span><span class="va">RSQ</span> <span class="op">&gt;=</span> <span class="fl">0.8</span>, <span class="va">group</span><span class="op">]</span><span class="op">]</span></pre></div>
</div>
<div id="caching" class="section level2">
<h2 class="hasAnchor">
<a href="#caching" class="anchor"></a>Caching</h2>
<p>In <code>patRoon</code> lengthy processing operations such as finding features and generating annotation data is <em>cached</em>. This means that when you run such a calculation again (without changing any parameters), the data is simply loaded from the cache data instead of re-generating it. This in turn is very useful, for instance, if you have closed your R session and want to continue with data processing at a later stage.</p>
<p>The cache data is stored in a <a href="https://www.sqlite.org/index.html">sqlite</a> database file. This file is stored by default under the name <code>cache.sqlite</code> in the current working directory (for this reason it is very important to always restore your working directory!). However, the name and location can be changed by setting a global package option:</p>
<div class="sourceCode" id="cb75"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.cache.fileName <span class="op">=</span> <span class="st">"~/myCacheFile.sqlite"</span><span class="op">)</span></pre></div>
<p>For instance, this might be useful if you want to use a shared cache file between projects.</p>
<p>After a while you may see that your cache file can get quite large. This is especially true when testing different parameters to optimize your workflow. Furthermore, you may want to clear the cache after you have updated <code>patRoon</code> and want to make sure that the latest code is used to generate the data. At any point you can simply remove the cache file. A more fine tuned approach which doesn’t wipe all your cached data is by using the <code><a href="../reference/clearCache.html">clearCache()</a></code> function. With this function you can selectively remove parts of the cache file. The function has two arguments: <code>what</code>, which specifies what should be removed, and <code>file</code> which specifies the path to the cache file. The latter only needs to be specified if you want to manage a different cache file.</p>
<p>In order to figure what is in the cache you can run <code><a href="../reference/clearCache.html">clearCache()</a></code> without any arguments:</p>
<div class="sourceCode" id="cb76"><pre class="downlit">
<span class="fu"><a href="../reference/clearCache.html">clearCache</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>#&gt; No cache file found, nothing to do ...</code></pre>
<p>Using this output you can re-run the function again, for instance:</p>
<div class="sourceCode" id="cb78"><pre class="downlit">
<span class="fu"><a href="../reference/clearCache.html">clearCache</a></span><span class="op">(</span><span class="st">"featuresOpenMS"</span><span class="op">)</span>
<span class="fu"><a href="../reference/clearCache.html">clearCache</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"featureGroupsOpenMS"</span>, <span class="st">"formulasGenForm"</span><span class="op">)</span><span class="op">)</span> <span class="co"># clear multiple</span>
<span class="fu"><a href="../reference/clearCache.html">clearCache</a></span><span class="op">(</span><span class="st">"OpenMS"</span><span class="op">)</span> <span class="co"># clear all with OpenMS in name (ie partial matched)</span>
<span class="fu"><a href="../reference/clearCache.html">clearCache</a></span><span class="op">(</span><span class="st">"all"</span><span class="op">)</span> <span class="co"># same as simply removing the file</span></pre></div>
</div>
<div id="parallelization" class="section level2">
<h2 class="hasAnchor">
<a href="#parallelization" class="anchor"></a>Parallelization</h2>
<p><code>patRoon</code> relies on several external (command-line) utilities to generate workflow data such as <code>MetFrag CL</code>, <code>OpenMS</code> and <code>SIRIUS</code>. When these tools are used for obtaining results for a large number of features it may take a long time before processing is finished. In order to speed this up these commands are executed in <em>parallel</em>. Running several commands simultenously is especially advantageous on multi core CPUs.</p>
<p>The number of commands to run in parallel is by default equal to the number of physical cores. You may want to change this amount, for instance, when you don’t want to dedicate all of your computers resources to <code>patRoon</code>. This can either be done by changing a global package option, which affects all operations, or when executing a single workflow step.</p>
<div class="sourceCode" id="cb79"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>patRoon.maxProcAmount <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="co"># execute max 3 commands in parallel</span>
<span class="va">formulas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/formula-generation.html">generateFormulas</a></span><span class="op">(</span><span class="va">...</span>, maxProcAmount <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="co"># ditto, but only for this function</span></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Rick Helmus.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
